---
title: "real_data_results_analysis"
output: html_document
---


```{r}
library(IRanges)
library(GenomicRanges)
library(data.table)
library(stringr)
library(foreach)
library(doParallel)
registerDoParallel(detectCores()-2)
library(Gviz)
library(ggplot2)
library(gridExtra)
```


```{r }
load("method_set_list.Rdata")
load("data_set_list.Rdata")

figures_dir <- "./results/figures/"
tables_dir <- "./results/tables/"
```

```{r add significance values to data_set_lists}
# for(i in 1:length(data_set_list)){
#   data_set <- data_set_list[[i]]
#   temp <- DMRscaler::run_MWW(data_set$g1, data_set$g2, data_set$B)
#   data_set_list[[i]]$locs$pval <- temp$p_val
# }
# save(data_set_list, file="data_set_list.Rdata")
```

```{r check method_set parameters }
for(i in 1:length(method_set_list)){
  print(method_set_list[[i]]$function_call )
}
```


```{r add granges for called dmrs}
filenames <- list.files(path = "./results/real_dataset_results/", pattern = "*result.csv", full.names = T)


data_set_results <- list()

for(data_set_id in 1:length(data_set_list)){
  for(method_id in 1:length(method_set_list)){
    filename <- filenames[grep(paste("data_set_",data_set_id,
                                     "__method_set_",method_id,
                                     sep=""), filenames)]
    if(length(filename)==0){next}
    result <- fread(filename)
    # if(method_set_list[[method_id]]$method=="dmrscaler"){
    #   result <- result[grep("64",result$layer),]
    # }
    
    locs <- data_set_list[[data_set_id]]$locs
    chr_seqlengths <- numeric(length=length(unique(locs$chr)))
    names(chr_seqlengths) <- unique(locs$chr)
    for(chr in unique(locs$chr)){
      chr_seqlengths[chr] <- max(locs$pos[which(locs$chr==chr)])+1
    }
    chr_seqinfo <- Seqinfo(names(chr_seqlengths), seqlengths = chr_seqlengths)
    
    data_set_results[[basename(filename)]]$grange <- GRanges(
      seqnames = result$chr,
      ranges = IRanges(start = result$start,
                       end = result$stop,
                       pval_region = result$pval_region),
      strand = rep("*",nrow(result)),
      seqinfo = chr_seqinfo
    )
    if(grepl("dmrscaler",names(method_set_list)[method_id] )){
      data_set_results[[basename(filename)]]$grange$layer <- result$layer
    }
    ### need to set a cutoff to region p-value for TP,FP, FN, etc
    which <- which(data_set_results[[basename(filename)]]$grange$pval_region < 0.05)
    data_set_results[[basename(filename)]]$grange <- data_set_results[[basename(filename)]]$grange[which]
    
    data_set_results[[basename(filename)]]$locs_grs <- GRanges(seqnames=locs$chr, ranges = IRanges(start=locs$pos, width=1))
    data_set_results[[basename(filename)]]$method <- method_set_list[[method_id]]$method
    data_set_results[[basename(filename)]]$method_id <- names(method_set_list)[method_id]
    data_set_results[[basename(filename)]]$data_set <- data_set_list[[data_set_id]]$data_set_name
    
    
  }
}
```

```{r write DMR width percentiles}
library("stringr")

quants_list <- list()
quants <- data.frame(percentile = numeric(), value = numeric(), method = character(), pars = character(), chrs = character())
ds_index <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", ds_index, sep = ""), names(data_set_results))
for(rs_index in result_set_ids){
  result_set <- data_set_results[[rs_index]] 
  if(grepl("dmrscaler", result_set$method)){
    if(grepl("_1", result_set$method_id)){
      for(l in 1:length(unique(result_set$grange$layer))){
          which <- which(seqnames(result_set$grange) == "chrX" &
                           result_set$grange$layer==unique(result_set$grange$layer)[l])
          temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
          quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                             method = "dmrscaler",
                                             pars = paste("dmrscaler",unique(result_set$grange$layer)[l]),
                                             chrs="chrX"))
          which <- which(seqnames(result_set$grange) != "chrX" & 
                           seqnames(result_set$grange) != "chrY"  &
                           result_set$grange$layer==unique(result_set$grange$layer)[l])
          temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
          quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                             method = "dmrscaler",
                                             pars =paste("dmrscaler",unique(result_set$grange$layer)[l]),
                                             chrs="Autosome"))
      }
    }
  } else {
    which <- which(seqnames(result_set$grange) == "chrX")
    temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
    quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                       method = result_set$method, pars = result_set$method_id,
                                       chrs="chrX"))
    which <- which(seqnames(result_set$grange) != "chrX" & seqnames(result_set$grange) != "chrY")
    temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
    quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                       method = result_set$method, pars = result_set$method_id,
                                       chrs="Autosome"))
  }
}
quants_list[[ names(data_set_list)[ds_index] ]] <- quants


data_set_ids <- grep("Sex", names(data_set_list), invert = T)
for(ds_index in data_set_ids){
  quants <- data.frame(percentile = numeric(), value = numeric(),
                       method = character(), pars = character(), chrs = character())
  result_set_ids <- grep(paste("data_set_", ds_index, sep = ""), names(data_set_results))
  for(rs_index in result_set_ids){
    result_set <- data_set_results[[rs_index]] 
    if(grepl("dmrscaler", result_set$method)){
      if(grepl("_1", result_set$method_id)){
        for(l in 1:length(unique(result_set$grange$layer))){
            which <- which(result_set$grange$layer==unique(result_set$grange$layer)[l])
            temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
            quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                               method = "dmrscaler",
                                               pars = paste("dmrscaler",unique(result_set$grange$layer)[l])))
        }
      }
    } else {
      temp <- quantile(log10(end(result_set$grange) - start(result_set$grange)), seq(0.01,1,by=0.01), type=1)
      quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                         method = result_set$method, pars = result_set$method_id))
    }
  }
  quants_list[[ names(data_set_list)[ds_index] ]] <- quants
}


for(i in 1:length(quants_list)){
  quants <- quants_list[[i]]
  gg<-ggplot(quants, aes(x=percentile, y=value, color=pars))+
    geom_line()+
    ylab("-log10(Width)")+
    theme(text=element_text(size=18))
  if(grepl("Sex",names(quants_list)[i])){
    gg <- gg + facet_grid(rows = vars(chrs),cols = vars(method) )
  } else{
    gg <- gg + facet_grid(cols = vars(method) )
  }
  plot(gg)

}


for(i in 1:length(quants_list)){
  filename <-  paste(tables_dir,"quantile_table_", names(quants_list)[i], ".csv", sep="")
  write.csv(file=filename, quants_list[[i]])
}
# 
# library(gridExtra)
# library(grid)
# filename=paste(results_dir,"/_w_combp_Width_Percentile_Plot.png", sep = "")
# #png(filename = filename, width = 10, height = 6, units = "in", res = 240)
# grid.newpage()
# grid.arrange(gg, nrow=1)
# #dev.off()  
# 
# quants$width <- 10^quants$value
# #filename=paste(results_dir, "_w_combp_Width_Percentile.csv", sep = "")
# #write.csv(file = filename, quants)


```

```{r}

gtf_df <- rtracklayer::import("../../STABLE_DATA/HUMAN_GENOME/Homo_sapiens.GRCh37.87.chr.gtf")
gtf_df <- as.data.frame(gtf_df)
gtf_df$seqnames<-paste("chr", gtf_df$seqnames, sep = "")

genes<-as.data.frame(cbind(chromosome=gtf_df$seqnames, start=gtf_df$start, end=gtf_df$end ))
genes<-cbind(genes, width=as.numeric(as.character(gtf_df$width)), strand=gtf_df$strand, feature=gtf_df$type, gene=gtf_df$gene_id, exon=gtf_df$exon_id, transcript=gtf_df$transcript_id, symbol=gtf_df$gene_name)

genes$start<-as.numeric(as.character(genes$start))
genes$end<-as.numeric(as.character(genes$end))


```

```{r, fig.height=9}
data_set_regions_df <- data.frame("data_set"=c("Sotos"),
                                  "chr"=c("chr5"),
                                  "start"=c(140090846),
                                  "stop"=c(140984057))

for(region_id in 1:nrow(data_set_regions_df)){
  data_set <- data_set_list[[data_set_regions_df$data_set[region_id] ]] 
  g1 <- data_set$g1
  g2 <- data_set$g2
  
  group<-character(length = ncol(data_set$B))
  group[which(is.element(colnames(data_set$B),g1))] <- data_set$g1g2_labels$g1
  group[which(is.element(colnames(data_set$B),g2))] <- data_set$g1g2_labels$g2

  group <- as.factor(group)
  color=c("blue", "seagreen3")
  linewide=c(2, 2 )
  
  chr <- data_set_regions_df$chr
  start <- data_set_regions_df$start
  stop <- data_set_regions_df$stop
  
  
  genes_exons<-genes[which((genes$chr==chr) & (genes$start>=start) & (genes$end<=stop)),]
  genes_exons<-droplevels(genes_exons)
  genes_exons<-genes_exons[complete.cases(genes_exons),]
  
  
  gen<-"hg19"
  
  which<-which(data_set$locs$chr==chr & data_set$locs$pos>=start & data_set$locs$pos<=stop )
  B_sub<- data_set$B[which,]
  B_norm<-B_sub - rowMeans(B_sub[,g1])
  locs <- data_set$locs[which,]
  
  delta_means <- rowMeans(B_sub[,g1])-rowMeans(B_sub[,g2])
  
  
  irange<-IRanges(start=locs$pos, width=1)
  grange<-GRanges(seqnames=chr, ranges =  ranges(irange), mcols = B_sub )
  
  manhattan_range <- GRanges(seqnames=chr, ranges =  ranges(irange) )
  manhattan_track <- DataTrack(range = manhattan_range, data = -log10(locs$pval),
                               lwd=1.5, col.histogram=c("black"), type=c("hist"), ylim=c(0,max(-log10(locs$pval))))

 
  raw_beta_track<-DataTrack(range=grange, data=t(B_sub), groups=group,
                            genome=gen, name="Beta", col=color, type=c("a","g", "confint"),
                            lwd=linewide, fontcolor.title="grey20", col.axis="black",
                            fontcolor.legend="grey20", legend = TRUE, lwd.grid=0.8,
                            col.grid="grey65",v=0 )
  displayPars(raw_beta_track) <- list( cex.legend=1, cex.axis=1, cex.title=1 )

  norm_beta_track<-DataTrack(range=grange, data=delta_means,  genome=gen,
                             group ="delta", name="Delta Beta", col="black",
                             type=c("a", "g"), fontcolor.title="grey20",col.axis="black",
                             fontcolor.legend="grey20",legend = TRUE, baseline=0,
                             col.baseline="black", lty.baseline=3)
  displayPars(norm_beta_track) <- list( cex.legend=1, cex.axis=1, cex.title=1 , lwd.grid=1, col.grid="grey80",v=0,h=0)
  
  
  grtrack <- GeneRegionTrack(genes_exons, genome=gen, chromosome=chr,
                             name="Gene Model", transcriptAnnotation="symbol",
                             collapseTranscripts = "meta")
  displayPars(grtrack) <- list( background.panel="white", col="grey20", fontsize=25, 
                                fontsize.group=25, fill = "grey50", lwd=1, 
                                fontcolor.group="grey20", fontcolor.title="grey20",
                                stackHeight = 0.95, shape="box")
  if( length(which)==0 ){
    grtrack <- itrack
  }
  
  dmr_tracks <- list()
  
  data_set_id <- grep(data_set$data_set_name, names(data_set_list))
  result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
  for(rs_index in result_set_ids){
    result_set <- data_set_results[[rs_index]] 
    if(grepl("dmrscaler", result_set$method)){
      if(grepl("_2", result_set$method_id)){
        for(l in 1:length(unique(result_set$grange$layer))){
          which_in_layer <- which(result_set$grange$layer==unique(result_set$grange$layer)[l]
                                  & seqnames(result_set$grange)==chr )
          dmr_tracks[[length(dmr_tracks)+1]] <- AnnotationTrack(start = start(result_set$grange[which_in_layer]) , 
                                                                  end =  end(result_set$grange[which_in_layer]),
                                                                  chromosome = chr,
                                                                  strand = "*", genome = "hg19",
                                                                  name=unique(result_set$grange$layer)[l])
          displayPars(dmr_tracks[[length(dmr_tracks)]]) <- list(cex.legend=0.7, cex.axis=0.7, cex.title=0.7, rotation.title=0, col="grey20", fill="grey70", fontcolor.title="grey20", stackHeight = 1, shape="box")
        }
      }
    } else {
      which_in_anno <- which(seqnames(result_set$grange)==chr)
      dmr_tracks[[length(dmr_tracks)+1]] <- AnnotationTrack(start = start(result_set$grange[which_in_anno]) , 
                                                              end =  end(result_set$grange[which_in_anno]),
                                                              chromosome = chr,
                                                              strand = "*", genome = "hg19",
                                                              name=result_set$method_id)
      displayPars(dmr_tracks[[length(dmr_tracks)]]) <- list(cex.legend=0.7, cex.axis=0.7, cex.title=0.7, rotation.title=0, col="grey20", fill="grey70", fontcolor.title="grey20", stackHeight = 1, shape="box")
      
    }
  }
  
  gtrack<-GenomeAxisTrack()
  displayPars(gtrack) <- list(cex=1, cex.id=1, fontcolor="grey20", col="grey20")
  itrack<-IdeogramTrack(genome=gen,chromosome = chr)
  displayPars(itrack) <- list(cex=1,cex.bands=1, fontcolor="grey20")
  track_list <- c(list(itrack, gtrack, raw_beta_track, norm_beta_track, grtrack),dmr_tracks )
  plotTracks(track_list,
             from = start-1, to=stop+1, # sizes =c(0.5,0.75,3,2,3,0.35,0.35,0.35,0.35,0.35),
             background.title = "white"
            )
  
  
  
  
  
  
  
  
  gglist <- list()
  gglist[["signif_adj_plot"]] <- ggplot(locs, aes(x=1:length(pval), y=-log10(pval) )) +
    geom_segment(aes(xend=1:length(-log10(pval)), y=0, yend=-log10(pval), alpha=0.5)) +
    geom_point(aes(color= -log10(pval))) + 
    scale_color_gradient2(low="grey60", mid="grey30", high = "red", midpoint = -log10(0.05) )+
    xlab("CG index over region") + ylab("-log10(pval)")+
    ggtitle(paste("Adjacent_CG_plot_", chr,"_" ,start, "_", stop))+
    ylab("-log10pval") +
    scale_x_continuous(expand = c(0, 0))+
    theme_bw()+
    theme(
      plot.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.text.x = element_blank(), axis.title.x = element_blank(),
      text = element_text(size=12, color = "black"),
      axis.text = element_text(size=12, color = "black"),
      legend.position = "none"
    )

  plot(gglist[["signif_adj_plot"]])  

  df_betas <- cbind("index" = 1:nrow(locs), locs, "mean"= rowMeans(B_sub[,g1]), "group"=data_set$g1g2_labels$g1 )
  df_betas <- rbind(df_betas, cbind("index" = 1:nrow(locs), locs, "mean" = rowMeans(B_sub[,g2]), "group"=data_set$g1g2_labels$g2))
                    
  gglist[["beta_raw_adj_plot"]] <- ggplot(df_betas)+
    geom_line(aes(x=index, y=mean, group=names, alpha=0.5))+
    geom_point(aes(x=index, y=mean, color=group))+
    scale_color_manual(values=color)+
    scale_x_continuous(expand = c(0, 0))+
  ylab("Beta") +
  theme_bw()+
  theme(
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_blank(), axis.title.x = element_blank(),
    text = element_text(size=12, color = "black"),
    axis.text = element_text(size=12, color = "black"),
    legend.position = "none"
  )
  
  plot(gglist[["beta_raw_adj_plot"]])
    
  df_betas <- cbind("index" = 1:nrow(locs), locs, "mean"= rowMeans(B_norm[,g1]), "group"=data_set$g1g2_labels$g1 )
  df_betas <- rbind(df_betas, cbind("index" = 1:nrow(locs), locs, "mean" = rowMeans(B_norm[, g2]), "group"=data_set$g1g2_labels$g2))
  
  gglist[["beta_norm_adj_plot"]] <- ggplot(df_betas)+
    geom_line(aes(x=index, y=mean, group=names, alpha=0.5))+
    geom_point(aes(x=index, y=mean, color=group))+
    scale_color_manual(values=color)+
    scale_x_continuous(expand = c(0, 0))+
  ylab("Normalized Beta") +
  theme_bw()+
  theme(
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_blank(), axis.title.x = element_blank(),
    text = element_text(size=12, color = "black"),
    axis.text = element_text(size=12, color = "black"),
    legend.position = "none"
  )
  
  plot(gglist[["beta_norm_adj_plot"]]) 
  
  
  
  ## is CG j in a DMR a level i
  in_dmr_list <- list()
  locs_gr <- GRanges(seqnames=locs$chr, IRanges(start=locs$pos, width=1))
  for(rs_index in result_set_ids){
    result_set <- data_set_results[[rs_index]]
    if(grepl("dmrscaler", result_set$method)){
      if(grepl("_2", result_set$method_id)){
        in_dmr_list[[result_set$method]]$locs <- locs
        in_dmr_list[[result_set$method]]$locs$index <- 1:nrow(locs)
        in_dmr_list[[result_set$method]]$locs$in_dmr <- 0 
        for(l in 1:length(unique(result_set$grange$layer))){
          which_in_layer <- which(result_set$grange$layer==unique(result_set$grange$layer)[l])
          in_dmr_list[[result_set$method]]$locs$in_dmr <- in_dmr_list[[result_set$method]]$locs$in_dmr + 
            countOverlaps(locs_gr, result_set$grange[which_in_layer])
        }
        in_dmr_list[[result_set$method]]$layer_names <- unique(result_set$grange$layer)
      }
    } else {
      in_dmr_list[[result_set$method_id]]$locs <- locs
        in_dmr_list[[result_set$method_id]]$locs$index <- 1:nrow(locs)
      in_dmr_list[[result_set$method_id]]$locs$in_dmr <- countOverlaps(locs_gr, result_set$grange)
    }
  }
  
  
  for(in_dmr_plot_id in 1:length(in_dmr_list)){
    plot_name <- paste("in_dmr_plot_", names(in_dmr_list)[in_dmr_plot_id], sep="") 
    gglist[[plot_name]] <- ggplot(in_dmr_list[[in_dmr_plot_id]]$locs)+
      geom_bar(aes(x=index, y=in_dmr, color=in_dmr), fill="grey70",  stat="identity")+
      scale_color_gradient(low="grey50", high = "grey50")+
      xlab("CG index over region") +
      theme_bw()+
      scale_x_continuous(expand = c(0, 0))+
      theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        text = element_text(size=12, color = "black"),
        axis.text = element_text(size=12, color = "black"),
        legend.position = "none"
      )
    if(grepl("dmrscaler", names(in_dmr_list)[in_dmr_plot_id]) ){
      gglist[[plot_name]] <- gglist[[plot_name]] + 
        scale_y_continuous(breaks = seq(0,length(in_dmr_list$dmrscaler$layer_names),1) ,
                           labels= c("none", rev(in_dmr_list$dmrscaler$layer_names)))+
      ylab("Layer") 
    } else {
      gglist[[plot_name]] <- gglist[[plot_name]] + 
        scale_y_continuous(breaks = seq(0,1),
                           labels= c("0", "1"))+
        ylab(gsub("in_dmr_plot_","", plot_name))+
        theme(
          axis.title.y = element_text(angle = 0, hjust=1),
          axis.title.x=element_blank(),
          axis.text.x=element_blank(),
          axis.ticks.x=element_blank()
        )
    }
  }
  gtable <- ggplot_gtable(ggplot_build(gglist[[1]]))
  for(i in 2:length(gglist)){
    gtable <- rbind(gtable, ggplot_gtable(ggplot_build(gglist[[i]])) )
  }
  panels <- gtable$layout$t[grepl("panel", gtable$layout$name)]
  gtable$heights[panels] <- unit(c(rep(4,4), rep(1,6)),"null")
  grid.newpage()
  grid.draw(gtable)
  
  # filename=paste(figures_dir"Adjacent_CG_plots_", chr,"_" ,start, "_", stop, ".png", sep = "")  
  # png(filename = filename, width = 2.75, height = 4, units = "in", res = 320)
  

  # dev.off()
  
  
  
  
    
}


```

```{r}

```

```{r  make rare disease analysis results tables}

library("gt")

results_table_list <- list()

for(ds_id in grep("Sex", names(data_set_list), invert = T)){
  results_table <- data.frame("method"=character(),
                              "num_dmrs"=numeric(), 
                              "mean_width"=numeric(),
                              "median_width"=numeric(),
                              "percent_total_width"=numeric())
  result_set_ids <- grep(paste("data_set_", ds_id, sep = ""), names(data_set_results))
  for(rs_index in result_set_ids){
    result_set <- data_set_results[[rs_index]] 
    if(grepl("dmrscaler", result_set$method)){
      if(grepl("_2", result_set$method_id)){
        for(l in 1:length(unique(result_set$grange$layer))){
          which <- which(result_set$grange$layer == unique(result_set$grange$layer)[l])
          results_table <- rbind(results_table,
                                 data.frame("method" = result_set$method,
                                            "num_dmrs" = length(result_set$grange[which]),
                                            "mean_width" = mean(width(result_set$grange[which])),
                                            "median_width" = median(width(result_set$grange[which])), 
                                            "percent_total_width" = sum(width(result_set$grange[which])) / 
                                              sum(seqlengths(result_set$grange))
                           ))
          rownames(results_table)[nrow(results_table)] <- unique(result_set$grange$layer)[l] 
        }
      }
    }else{
      results_table <- rbind(results_table,
                             data.frame("method" = result_set$method,
                                        "num_dmrs" = length(result_set$grange),
                                        "mean_width" = mean(width(result_set$grange)),
                                        "median_width" = median(width(result_set$grange)), 
                                        "percent_total_width" = sum(width(result_set$grange)) / 
                                          sum(seqlengths(result_set$grange))
                       ))
      rownames(results_table)[nrow(results_table)] <- result_set$method_id
    }
  
  }
  
  results_table$method_params <- rownames(results_table)
  
  clean_results_table <- gt(results_table, groupname_col = "method", rowname_col = "method_params") 
  clean_results_table <- tab_header(clean_results_table, title = paste(names(data_set_list)[ds_id],"Analysis DMR Summary Table"))
  clean_results_table <- tab_stubhead(clean_results_table , "method")
  
  clean_results_table <- cols_label(clean_results_table,
                                    num_dmrs = html(" #\n<br>DMRs "),
                                    mean_width = html(" mean\n<br>DMR width "),
                                    median_width = html(" median\n<br>DMR width "),
                                    percent_total_width = html(" % total\n<br>width "))
  
  
  clean_results_table <- fmt_number(clean_results_table, columns = c(percent_total_width),
                                    n_sigfig = 2, scale_by = 100, pattern = "{x}%")
  clean_results_table <- fmt_number(clean_results_table, columns = c(mean_width),
                                    rows = (mean_width >= 1e6),
                                    decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
  clean_results_table <- fmt_number(clean_results_table, columns = c(mean_width),
                                    rows = (mean_width > 1000) & (mean_width < 1e6),
                                    decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
  clean_results_table <- fmt_number(clean_results_table, columns = c(mean_width),
                                    rows = (mean_width <= 1000),
                                    decimals = 0, pattern = "{x} bp")
  
  clean_results_table <- fmt_number(clean_results_table, columns = c(median_width),
                                    rows = (median_width >= 1e6),
                                    decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
  clean_results_table <- fmt_number(clean_results_table, columns = c(median_width),
                                    rows = (median_width > 1000) & (median_width < 1e6),
                                    decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
  clean_results_table <- fmt_number(clean_results_table, columns = c(median_width),
                                    rows = (median_width <= 1000),
                                    decimals = 0, pattern = "{x} bp")
  
  results_table_list[[names(data_set_list)[ds_id]]] <- clean_results_table
  
}


for(i in 1:length(results_table_list)){
  print(results_table_list[i])
}

clean_results_table

```


```{r make XY sex analysis results table}

library("gt")

results_table <- data.frame("method"=character(),
                            "chrX_num_dmrs"=numeric(), 
                            "chrX_mean_width"=numeric(),
                            "chrX_median_width"=numeric(),
                            "chrX_percent_total_width"=numeric(),
                            "autosome_num_dmrs"=numeric(), 
                            "autosome_avg_width"=numeric(),
                            "autosome_median_width"=numeric(),                            
                            "autosome_percent_total_width"=numeric())


data_set_id <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
for(rs_index in result_set_ids){
  result_set <- data_set_results[[rs_index]] 
  if(grepl("dmrscaler", result_set$method)){
    if(grepl("_2", result_set$method_id)){
      for(l in 1:length(unique(result_set$grange$layer))){
        which_x <- which(result_set$grange@seqnames=="chrX" &
                           result_set$grange$layer == unique(result_set$grange$layer)[l])
        which_not_x <- which(result_set$grange@seqnames != "chrX" & result_set$grange@seqnames != "chrY" &
                           result_set$grange$layer == unique(result_set$grange$layer)[l])
        results_table <- rbind(results_table,
                               data.frame("method" = result_set$method,
                                          "chrX_num_dmrs" = length(which_x),
                                          "chrX_mean_width" = mean(width(result_set$grange[which_x,])),
                                          "chrX_median_width" = median(width(result_set$grange[which_x,])), 
                                          "chrX_percent_total_width" = sum(width(result_set$grange[which_x,])) / 
                                            seqlengths(result_set$grange)["chrX"],
                                          "autosome_num_dmrs" = length(which_not_x),
                                          "autosome_mean_width" = mean(width(result_set$grange[which_not_x,])),
                                          "autosome_median_width"=median(width(result_set$grange[which_not_x,])),
                                          "autosome_percent_total_width" = sum(width(result_set$grange[which_not_x,])) / 
                                            sum(seqlengths(result_set$grange)[!grepl("[XY]",names(seqlengths(result_set$grange)))])
                         ))
        rownames(results_table)[nrow(results_table)] <- unique(result_set$grange$layer)[l] 
      }
    }
  }else{
    which_x <- which(result_set$grange@seqnames=="chrX")
    which_not_x <- which(result_set$grange@seqnames != "chrX" & result_set$grange@seqnames != "chrY")
    results_table <- rbind(results_table,
                           data.frame("method" = result_set$method,
                                      "chrX_num_dmrs" = length(which_x),
                                      "chrX_mean_width" = mean(width(result_set$grange[which_x,])),
                                      "chrX_median_width" = median(width(result_set$grange[which_x,])), 
                                      "chrX_percent_total_width" = sum(width(result_set$grange[which_x,])) / 
                                        seqlengths(result_set$grange)["chrX"],
                                      "autosome_num_dmrs" = length(which_not_x),
                                      "autosome_mean_width" = mean(width(result_set$grange[which_not_x,])),
                                      "autosome_median_width"=median(width(result_set$grange[which_not_x,])),
                                      "autosome_percent_total_width" = sum(width(result_set$grange[which_not_x,])) / 
                                        sum(seqlengths(result_set$grange)[!grepl("[XY]",names(seqlengths(result_set$grange)))])
                     ))
    rownames(results_table)[nrow(results_table)] <- result_set$method_id
  }

}


results_table$method_params <- rownames(results_table)

clean_results_table <- gt(results_table, groupname_col = "method", rowname_col = "method_params") 
clean_results_table <- tab_header(clean_results_table, title = "Sex Analysis DMR Summary Table")
clean_results_table <- tab_spanner(clean_results_table, label="chrX",
                                   columns = c("chrX_num_dmrs","chrX_mean_width","chrX_median_width", "chrX_percent_total_width") )
clean_results_table <- tab_spanner(clean_results_table, label="Autosomes",
                                   columns = c("autosome_num_dmrs","autosome_mean_width","autosome_median_width", "autosome_percent_total_width") )
clean_results_table <- tab_stubhead(clean_results_table , "method")

clean_results_table <- cols_label(clean_results_table,
                                  chrX_num_dmrs = html(" #\n<br>DMRs "), chrX_mean_width = html(" mean\n<br>DMR width "),
                                  chrX_median_width = html(" median\n<br>DMR width "), chrX_percent_total_width = html(" % total\n<br>width "),
                                  autosome_num_dmrs = html(" #\n<br>DMRs "), autosome_mean_width = html(" mean\n<br>DMR width "),
                                  autosome_median_width = html(" median\n<br>DMR width "), autosome_percent_total_width = html(" % total\n<br>width " ))


clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_percent_total_width, autosome_percent_total_width),
                                  n_sigfig = 2, scale_by = 100, pattern = "{x}%")

clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_mean_width),
                                  rows = (autosome_mean_width > 1000) & (autosome_mean_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_mean_width),
                                  rows = (autosome_mean_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")

clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_mean_width),
                                  rows = (chrX_mean_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_mean_width),
                                  rows = (chrX_mean_width > 1000) & (chrX_mean_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_mean_width),
                                  rows = (chrX_mean_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")

clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_median_width),
                                  rows = (chrX_median_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_median_width),
                                  rows = (chrX_median_width > 1000) & (chrX_median_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_median_width),
                                  rows = (chrX_median_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")

clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_median_width),
                                  rows = (autosome_median_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_median_width),
                                  rows = (autosome_median_width > 1000) & (autosome_median_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_median_width),
                                  rows = (autosome_median_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")


clean_results_table

#gtsave(clean_results_table, filename = paste( "Sex_Analysis_DMR_Summary_Table_2.png",sep=""))

```

```{r make XY analysis circos network plot}

library(networkD3)
library(rjson)

data_set_id <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
method_set_id <- grep("dmrscaler_1", names(method_set_list))
method_set_ids <- grep(paste("method_set_", method_set_id, sep = ""), names(data_set_results))
result_set <- data_set_results[[ intersect(result_set_ids, method_set_ids ) ]] 



full_node_set <- list()
for(j in 1:length(unique(result_set$grange$layer))){
  #which <- which(built_layers[[j]]$chr =="chrX")
  #which <- 1:length(built_layers[[j]]$chr)
  gr_nodes <- list()
  which <- which(result_set$grange$layer == unique(result_set$grange$layer)[j]) 
  for(i in 1:length(which)){
    temp_name <- paste(as.character(seqnames(result_set$grange[ which[i],])), start(result_set$grange[which[i],]), end(result_set$grange[which[i],]), sep = "_")
   # temp_name <- paste(format(start(x_base[i,]),big.mark=","), format(end(x_base[i,]),big.mark=","), sep = " - ")
    temp_child <- list("name"= temp_name, "alt_name"= temp_name, "grange"=result_set$grange[which[i],], "children"=list() )
    gr_nodes[[i]] <-  list("name"= temp_name, "alt_name"= temp_name, "grange"=result_set$grange[which[i],], "children"=list())
  }
  full_node_set[[j]]<-gr_nodes
}

temp_full_node_set <- full_node_set
#full_node_set <- temp_full_node_set

for(j in 2:length(full_node_set)){
  print(j)
  
  lower <- full_node_set[[j-1]]
  upper <- full_node_set[[j]]
  
  lower_gr <- lower[[1]]$grange
  for(i in 2:length(lower)){
    lower_gr <- c(lower_gr, lower[[i]]$grange)
  }
  upper_gr <- upper[[1]]$grange
  if(length(upper) >= 2){
    for(i in 2:length(upper)){
      upper_gr <- c(upper_gr, upper[[i]]$grange)
    }
  }
  
  hits <- GenomicRanges::findOverlaps(lower_gr,upper_gr,type = "within");hits
  for( i in 1:length(hits@from)){
    upper[[ hits@to[i] ]]$children <- c(upper[[ hits@to[i] ]]$children, list(lower[[ hits@from[i] ]]))
    if(upper[[ hits@to[i] ]]$name == lower[[ hits@from[i] ]]$name |
       upper[[ hits@to[i] ]]$alt_name == lower[[ hits@from[i] ]]$name  |
       upper[[ hits@to[i] ]]$name == lower[[ hits@from[i] ]]$alt_name
       ){  ### avoid redundant child-parent pairs
      upper[[ hits@to[i] ]]$name <- "..."
    }
  }
  full_node_set[[j]]<-upper
}

#dmr_tree <- full_node_set[[3]][[1]]
#dmr_tree <- list("name"="root","children"=full_node_set[[3]])
dmr_tree <- list("name"="root","children"=full_node_set[[length(full_node_set)]])

dmr_tree_clean <- list()

node <- dmr_tree



strip_grs <- function(node){
  node <- list("name" = node$name, "children" = node$children)
  if(length(node$children)==0){return(node)}
  for(i in 1:length(node$children)){
    node$children[[i]]<-strip_grs(node$children[[i]])
  }
  return(node)
}
dmr_tree_clean <- strip_grs(dmr_tree)


library("webshot") ###  library to fix an obnoxious problem

temp <- radialNetwork(List = dmr_tree_clean, fontSize = 16, fontFamily = "bold", nodeStroke = "black", linkColour = "grey60", opacity = 1, margin=0, height = 1600, width = 1600)
# saveNetwork(temp,paste("All_chrom_dmrs_radialNetwork.html",sep = "") ,
#             selfcontained = T)
# webshot(paste("All_chrom_dmrs_radialNetwork.html",sep = ""),
#         paste("All_all_layer_chrom_dmrs_radialNetwork.png",sep = ""),
#         zoom = 2)
# 
# 
# temp <- diagonalNetwork(List = dmr_tree_clean[[2]][[1]], fontSize = 12, fontFamily = "bold", nodeStroke = "black", linkColour = "black", opacity = 1,  margin=0, height = 400, width = 150)
# saveNetwork(temp,paste(figures_dir, "FINAL_X2_chrom_dmrs_diagonalNetwork.html", sep="") , selfcontained = T )
# webshot(paste("FINAL_X2_chrom_dmrs_diagonalNetwork.html", sep=""),
#         paste("FINAL_X2_chrom_dmrs_diagonalNetwork.png", sep=""),
#         vheight = 400, vwidth = 150 , zoom =10 )


```



```{r make XY analysis DMR called width histograms}
library("gt")


data_set_id <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
for(rs_index in result_set_ids){
  result_set <- data_set_results[[rs_index]] 
  if(grepl("dmrscaler", result_set$method)){
    if(grepl("_1", result_set$method_id)){
      for(l in 1:length(unique(result_set$grange$layer))){
        which_x <- which(result_set$grange@seqnames=="chrX" &
                           result_set$grange$layer == unique(result_set$grange$layer)[l])
        which_not_x <- which(result_set$grange@seqnames != "chrX" & result_set$grange@seqnames != "chrY" &
                           result_set$grange$layer == unique(result_set$grange$layer)[l])

        gglist <- list()
        gglist[["X_chromosome"]] <- ggplot(data.frame(widths=width(result_set$grange[which_x])), aes(x=log10(widths)) )+
          ggtitle(label = "X chromosome DMR widths",
                  subtitle = paste(unique(result_set$grange$layer)[l], sep=""))+
          geom_histogram(bins=50)+
          xlim(0,8.5)+
          theme(text=element_text(size=18))
        gglist[["Autosomes"]] <- ggplot(data.frame(widths=width(result_set$grange[which_not_x]) ), aes(x=log10(widths)) )+
          ggtitle(label = "Autosome DMR widths",
                  subtitle = paste(unique(result_set$grange$layer)[l], sep=""))+
          geom_histogram(bins=50)+
          xlim(0,8.5)+
          theme(text=element_text(size=18))
        filename=paste(figures_dir, result_set$data_set ,"_DMR_widths__",unique(result_set$grange$layer)[l], ".png", sep = "")
        png(filename = filename, width = 10, height = 6, units = "in", res = 150)
        grid.newpage()
        grid.arrange( gglist[["X_chromosome"]],  gglist[["Autosomes"]], nrow=1)
        dev.off()
      }
    }
  }else{
    gglist <- list()
    which_x <- which(result_set$grange@seqnames=="chrX")
    which_not_x <- which(result_set$grange@seqnames != "chrX" & result_set$grange@seqnames != "chrY")
    gglist <- list()
    gglist[["X_chromosome"]] <- ggplot(data.frame(widths=width(result_set$grange[which_x]) ), aes(x=log10(widths)) )+
      ggtitle(label = "X chromosome DMR widths",
              subtitle = paste(result_set$method_id, sep=""))+
      geom_histogram(bins=50)+
      xlim(0,8.5)+
      theme(text=element_text(size=18))
    gglist[["Autosomes"]] <- ggplot(data.frame(widths=width(result_set$grange[which_not_x]) ), aes(x=log10(widths)) )+
      ggtitle(label = "Autosome DMR widths",
              subtitle = paste(result_set$method_id, sep=""))+
      geom_histogram(bins=50)+
      xlim(0,8.5)+
      theme(text=element_text(size=18))
    filename=paste(figures_dir, result_set$data_set ,"_DMR_widths__",result_set$method_id, ".png", sep = "")
    png(filename = filename, width = 10, height = 6, units = "in", res = 150)
    grid.newpage()
    grid.arrange( gglist[["X_chromosome"]],  gglist[["Autosomes"]], nrow=1)
    dev.off()
  }
}

```













```{r funcs for kws overlap analysis}

frac_dbeta_same_dir <- function(gr12, B1, atom1, case1, B2, case2, atom2, sigcut){
    num_sig <- numeric(length = length(gr12))
    frac_shared_sig <- numeric(length = length(gr12))
    frac_sig_same_dir <- numeric(length = length(gr12))
    for(i in 1:length(gr12)){
      #print(i)
      chr <- as.character(gr12@seqnames)[i]
      start <- start(gr12)[i]
      end <- end(gr12)[i]
      control1 <- (1:ncol(B1))[-case1]
      control2 <- (1:ncol(B2))[-case2]
      
      cgs <- which(atom1$chr == chr & atom1$pos <= end & atom1$pos >= start)
      
      ##hyper (caseBeta > controlBeta)
      if(length(cgs)==1){ 
        hyper1 <- mean(B1[cgs,case1]) > mean(B1[cgs,control1])
        hyper2 <- mean(B2[cgs,case2]) > mean(B2[cgs,control2]);
      }
      else{
        hyper1 <- ( rowMeans(B1[cgs,case1]) > rowMeans(B1[cgs,control1]) ) 
        hyper2 <- ( rowMeans(B2[cgs,case2]) > rowMeans(B2[cgs,control2]) ) 
      }
      sig1 <- (atom1$pval[cgs] < sigcut) ## 1 corresponds to FDR = 0.1 based on scoring func
      sig2 <- (atom2$pval[cgs] < sigcut) ## 1 corresponds to FDR = 0.1 based on scoring func
      num_sig[i] <- sum(sig1 | sig2)
      frac_shared_sig[i] <- sum(sig1 & sig2) / sum(sig1 | sig2)
      frac_sig_same_dir[i] <- sum((hyper1 == hyper2) & sig1 & sig2 )  / sum(sig1 & sig2)
      if(is.na(frac_sig_same_dir[i])){frac_sig_same_dir[i] = 0}
    }
    return( data.frame(num_sig = num_sig, frac_shared_sig = frac_shared_sig, frac_sig_same_dir = frac_sig_same_dir ))
}

frac_A_in_B <- function(gr1,gr2){
  which <- findOverlaps(gr1,gr2)@from
  frac_a_in_b <- rep(0,length(gr1))
  
  temp_int_gr <- intersect(gr1, gr2)
  temp_ovs <- findOverlaps(gr1, temp_int_gr)
  
  froms <- temp_ovs@from
  tos <- temp_ovs@to
  
  gr1 <- as.data.frame(gr1)
  gr2 <- as.data.frame(gr2)
  temp_int_gr <- as.data.frame(temp_int_gr)
  for(i in 1:length(unique(froms))){
    frac_a_in_b[froms[i]] <- 
      sum(temp_int_gr$width[tos[which(froms==(unique(froms)[i]) )] ]) / 
      sum(gr1$width[froms[which(froms==unique(froms)[i]) ]]  )
      
  }
  
  return(frac_a_in_b)
}
```

```{r kat6a-weaver-sotos overlap analysis}
genes <- gtf_df[which(gtf_df$gene_biotype =="protein_coding" ),]
genes <- genes[which(genes$gene_source=="ensembl_havana"),]
genes <- genes[which(genes$type=="gene" ),]
genes <- genes[,!(colSums(!is.na(genes))==0)]
gene_ranges <- GRanges(seqnames = genes$seqnames, IRanges(start = genes$start, width = genes$width))
gene_ranges$gene_names <- genes$gene_name


geneset_overlaps = data.frame(layer=numeric(),
                              overlap_between = character(),
                              gene = character())
range_overlaps <- data.frame(layer=numeric(), datasets=character(),
                             seqnames = character(), start = numeric(), width = numeric(),
                             overlapping_genes = character(),
                             gene_overlap_count = numeric(),
                             num_cg_sig = numeric(),
                             frac_shared_sig = numeric(),
                             frac_sig_same_dir = numeric(),
                             frac_shared_kat6a = numeric(),
                             frac_shared_sotos = numeric(),
                             frac_shared_weaver = numeric() ) 
syndrome_overlap_list <- list()
for(i in 1:length(method_set_list)){
  syndrome_overlap_list[[ names(method_set_list)[i] ]] <- list(geneset_overlaps = geneset_overlaps,
                                                      range_overlaps = range_overlaps) 
}

for(dsr_index in 1:length(data_set_results)){
  print(dsr_index)
  temp_gr <- data_set_results[[dsr_index]]$grange
  temp_gr$overlapping_genes <- ""
  temp_gr$gene_overlap_count <- ""
  
  temp <- findOverlaps(temp_gr, gene_ranges)
  froms <- unique(temp@from)
  temp_genes <- gene_ranges$gene_names[temp@to]
  temp_gr <- as.data.frame(temp_gr)  ### NOTE: much faster if convert to df first -> grange ops are slow in loop
  
  for(i in 1:length(froms)){
    temp_gr$overlapping_genes[froms[i]] <- paste(temp_genes[which(temp@from==froms[i])], collapse = ",")
  }
  temp_gr <- GRanges(temp_gr)
  temp_gr$gene_overlap_count <- countOverlaps(temp_gr, gene_ranges)
  data_set_results[[dsr_index]]$grange <- temp_gr
}



data_set_ids <- grep("Sex", names(data_set_list), invert = T)
for(ds_index_1 in data_set_ids){
  print(paste("ds_index_1:",ds_index_1))
  for(ds_index_2 in data_set_ids){
    print(paste("ds_index_2:",ds_index_2))
    result_set_ids <- grep(paste("data_set_", ds_index_1, sep = ""), names(data_set_results))
    for(rs_index_1 in result_set_ids){
      result_set_1 <- data_set_results[[rs_index_1]] 
      method_set <- gsub("data_set.*method_set_", "", names(data_set_results)[rs_index_1])
      method_set <- gsub("_result.csv","",method_set)
      rs_index_2 <- grep(paste("data_set_", ds_index_2,"__method_set_",
                               method_set, sep = ""), names(data_set_results))
      gr_k <- data_set_results[[grep(paste("data_set_", grep("kat6a", names(data_set_list), ignore.case = T), 
                         "__method_set_", method_set, sep = ""), names(data_set_results))]]$grange
      gr_s <- data_set_results[[grep(paste("data_set_", grep("sotos", names(data_set_list), ignore.case = T), 
                         "__method_set_", method_set, sep = ""), names(data_set_results))]]$grange
      gr_w <- data_set_results[[grep(paste("data_set_", grep("weaver", names(data_set_list), ignore.case = T), 
                         "__method_set_", method_set, sep = ""), names(data_set_results))]]$grange
      result_set_2 <- data_set_results[[rs_index_2]]
      
      if(grepl("dmrscaler", result_set_1$method)){
        if(grepl("_1", result_set_1$method_id)){
          for(l in 1:length(unique(result_set_1$grange$layer))){
            temp_gr_k <- gr_k[which(gr_k$layer==unique(result_set_1$grange$layer)[l])] 
            temp_gr_s <- gr_s[which(gr_s$layer==unique(result_set_1$grange$layer)[l])] 
            temp_gr_w <- gr_w[which(gr_w$layer==unique(result_set_1$grange$layer)[l])]
            
            
            print(paste("layer:",l))
            temp_geneset_overlaps <- syndrome_overlap_list[[result_set_1$method_id]]$geneset_overlaps
            which_1 <- which(result_set_1$grange$layer == unique(result_set_1$grange$layer)[l])
            which_2 <- which(result_set_2$grange$layer == unique(result_set_1$grange$layer)[l])
            temp_geneset_overlaps <- rbind(temp_overlap_list$geneset_overlaps,
              data.frame( layer =  unique(result_set_1$grange$layer)[l],
                 overlap_between = paste(sort(c(result_set_1$data_set,
                                                result_set_2$data_set)),
                                         collapse="-"),
                 gene = intersect(
                   unique(gene_ranges$gene_names[findOverlaps(result_set_1$grange[which_1], gene_ranges)@to]),
                   unique(gene_ranges$gene_names[findOverlaps(result_set_2$grange[which_2], gene_ranges)@to])
                 )
              )
            )
            syndrome_overlap_list[[result_set_1$method_id]]$geneset_overlaps <- temp_geneset_overlaps
            
            
            
            temp_gr <- intersect(result_set_1$grange[which_1], result_set_2$grange[which_2])
            temp_frac_dbsd <- frac_dbeta_same_dir(gr12 = temp_gr,
                                B1 = data_set_list[[result_set_1$data_set]]$B,
                                atom1 = data_set_list[[result_set_1$data_set]]$locs,
                                case1 = which(is.element( colnames(data_set_list[[result_set_1$data_set]]$B),
                                                       data_set_list[[result_set_1$data_set]]$g2)),
                                B2= data_set_list[[result_set_2$data_set]]$B,
                                atom2 = data_set_list[[result_set_2$data_set]]$locs,
                                case2 = which(is.element( colnames(data_set_list[[result_set_2$data_set]]$B),
                                                       data_set_list[[result_set_2$data_set]]$g2)),
                                sigcut=0.05
                                )
            temp_fo <- findOverlaps(temp_gr, gene_ranges)
            froms <- unique(temp_fo@from)
            temp_genes <- gene_ranges$gene_names[temp_fo@to]
            temp_gr <- as.data.frame(temp_gr)
            temp_gr$overlapping_genes <- ""
            temp_gr$overlap <- ""
            for(i in 1:length(froms)){
              temp_gr$overlapping_genes[froms[i]] <- paste(temp_genes[which(temp_fo@from==froms[i])],
                                                           collapse = ",")
            }
            temp_gr <- GRanges(temp_gr)
            temp_gr$gene_overlap_count <- countOverlaps(temp_gr, gene_ranges)
            
            
            temp_range_overlap <- data.frame(layer=unique(result_set_1$grange$layer)[l],
               datasets=paste(sort(c(result_set_1$data_set,
                                     result_set_2$data_set)),collapse="-"),
               seqnames=as.character(seqnames(temp_gr)),start=start(temp_gr), width=width(temp_gr),
               overlapping_genes=temp_gr$overlapping_genes, gene_overlap_count=temp_gr$gene_overlap_count,
               num_cg_sig= temp_frac_dbsd$num_sig,
               frac_shared_sig= temp_frac_dbsd$frac_shared_sig,
               frac_sig_same_dir= temp_frac_dbsd$frac_sig_same_dir,
               frac_shared_kat6a = frac_A_in_B(temp_gr,temp_gr_k),
               frac_shared_sotos = frac_A_in_B(temp_gr,temp_gr_s),
               frac_shared_weaver = frac_A_in_B(temp_gr,temp_gr_w))
            
            
            syndrome_overlap_list[[result_set_1$method_id]]$range_overlaps <- temp_range_overlap
          }
        }
      } else {
        print(paste("method:",result_set_1$method_id))
        temp_geneset_overlaps <- syndrome_overlap_list[[result_set_1$method_id]]$geneset_overlaps
        temp_geneset_overlaps <- rbind(temp_overlap_list$geneset_overlaps,
          data.frame( layer =  result_set_1$method_id,
             overlap_between = paste(sort(c(result_set_1$data_set,
                                            result_set_2$data_set)),
                                     collapse="-"),
             gene = intersect(
               unique(gene_ranges$gene_names[findOverlaps(result_set_1$grange, gene_ranges)@to]),
               unique(gene_ranges$gene_names[findOverlaps(result_set_2$grange, gene_ranges)@to])
             )
          )
        )
        syndrome_overlap_list[[result_set_1$method_id]]$geneset_overlaps <- temp_geneset_overlaps
        
        
        
        temp_gr <- intersect(result_set_1$grange, result_set_2$grange)
        temp_frac_dbsd <- frac_dbeta_same_dir(gr12 = temp_gr,
                            B1 = data_set_list[[result_set_1$data_set]]$B,
                            atom1 = data_set_list[[result_set_1$data_set]]$locs,
                            case1 = which(is.element( colnames(data_set_list[[result_set_1$data_set]]$B),
                                                   data_set_list[[result_set_1$data_set]]$g2)),
                            B2= data_set_list[[result_set_2$data_set]]$B,
                            atom2 = data_set_list[[result_set_2$data_set]]$locs,
                            case2 = which(is.element( colnames(data_set_list[[result_set_2$data_set]]$B),
                                                   data_set_list[[result_set_2$data_set]]$g2)),
                            sigcut=0.05
                            )
        temp_fo <- findOverlaps(temp_gr, gene_ranges)
        froms <- unique(temp_fo@from)
        temp_genes <- gene_ranges$gene_names[temp_fo@to]
        temp_gr <- as.data.frame(temp_gr)
        temp_gr$overlapping_genes <- ""
        temp_gr$overlap <- ""
        for(i in 1:length(froms)){
          temp_gr$overlapping_genes[froms[i]] <- paste(temp_genes[which(temp_fo@from==froms[i])],
                                                       collapse = ",")
        }
        temp_gr <- GRanges(temp_gr)
        temp_gr$gene_overlap_count <- countOverlaps(temp_gr, gene_ranges)
        
        
        temp_range_overlap <- data.frame(layer=unique(result_set_1$grange$layer)[l],
           datasets=paste(sort(c(result_set_1$data_set,
                                 result_set_2$data_set)),collapse="-"),
           seqnames=as.character(seqnames(temp_gr)),start=start(temp_gr), width=width(temp_gr),
           overlapping_genes=temp_gr$overlapping_genes, gene_overlap_count=temp_gr$gene_overlap_count,
           num_cg_sig= temp_frac_dbsd$num_sig,
           frac_shared_sig= temp_frac_dbsd$frac_shared_sig,
           frac_sig_same_dir= temp_frac_dbsd$frac_sig_same_dir,
           frac_shared_kat6a = frac_A_in_B(temp_gr,gr_k),
           frac_shared_sotos = frac_A_in_B(temp_gr,gr_s),
           frac_shared_weaver = frac_A_in_B(temp_gr,gr_w))
        
        
        syndrome_overlap_list[[result_set_1$method_id]]$range_overlaps <- temp_range_overlap
    
        
        
        
      }
    }
  }
}











#kws_dat_cp <- kws_dat
kws_dat <- kws_dat_cp

## overlap K-W, K-S, W-S, K-W-S
## feature overlap, bp overlap
## match layers 4-4, 8-8, 16-16, etc
## % same direction (hypo or hyper)
## genes overlapped
## OR+-CI cgs, OR+-CI gene sets, OR+-CI region
results_dir = "/home/lbondhus/Desktop/PROJECTS/dmrscaler_real_data/results/OVERLAPS/"
  
geneset_overlaps = data.frame(layer=numeric(),
                              overlap_between = character(),
                              gene = character())
range_overlaps <- data.frame(layer=numeric(), datasets=character(),
                             seqnames = character(), start = numeric(), width = numeric(),
                             overlapping_genes = character(),
                             gene_overlap_count = numeric(),
                             num_cg_sig = numeric(),
                             frac_shared_sig = numeric(),
                             frac_sig_same_dir = numeric(),
                             frac_shared_kat6a = numeric(),
                             frac_shared_sotos = numeric(),
                             frac_shared_weaver = numeric() )  
for(layer_num in 1:length(kws_dat$KAT6A$built_layers)){
test <- foreach(layer_num=1:length(kws_dat$KAT6A$built_layers), .combine=rbind) %dopar% {
  geneset_overlaps = data.frame(layer=numeric(), overlap_between = character(), gene = character())
  range_overlaps <- data.frame(layer=numeric(), datasets=character(),
                             seqnames = character(), start = numeric(), width = numeric(),
                             overlapping_genes = character(), gene_overlap_count = numeric(),
                             num_cg_sig = numeric(), frac_shared_sig = numeric(), frac_sig_same_dir = numeric(),
                             frac_shared_kat6a = numeric(), frac_shared_sotos = numeric(),frac_shared_weaver = numeric() )  
  ## foreach
  
  print(layer_num)
  gr_k <- GRanges(seqnames = kws_dat$KAT6A$built_layers[[layer_num]]$chr,
                     IRanges(start = kws_dat$KAT6A$built_layers[[layer_num]]$start_pos, 
                             end = kws_dat$KAT6A$built_layers[[layer_num]]$stop_pos))
  gr_s <- GRanges(seqnames = kws_dat$Sotos$built_layers[[layer_num]]$chr,
                     IRanges(start = kws_dat$Sotos$built_layers[[layer_num]]$start_pos, 
                             end = kws_dat$Sotos$built_layers[[layer_num]]$stop_pos))
  gr_w <- GRanges(seqnames = kws_dat$Weaver$built_layers[[layer_num]]$chr,
                     IRanges(start = kws_dat$Weaver$built_layers[[layer_num]]$start_pos, 
                             end = kws_dat$Weaver$built_layers[[layer_num]]$stop_pos))
  for(i in 1:length(gr_k)){
    temp <- findOverlaps(gr_k[i], gene_ranges)
    gr_k$overlapping_genes[i] <- paste(genes[temp@to,]$gene_name, collapse = "," )
    gr_k$gene_overlap_count[i] <- countOverlaps(gr_k[i], gene_ranges)
  }
  for(i in 1:length(gr_s)){
    temp <- findOverlaps(gr_s[i], gene_ranges)
    gr_s$overlapping_genes[i] <- paste(genes[temp@to,]$gene_name, collapse = "," )
    gr_s$gene_overlap_count[i] <- countOverlaps(gr_s[i], gene_ranges)
  }
  for(i in 1:length(gr_w)){
    temp <- findOverlaps(gr_w[i], gene_ranges)
     gr_w$overlapping_genes[i] <- paste(genes[temp@to,]$gene_name, collapse = "," )
     gr_w$gene_overlap_count[i] <- countOverlaps(gr_w[i], gene_ranges)
  }
  temp <- frac_dbeta_same_dir( gr12 = gr_k,
                               B1 = kws_dat$KAT6A$B, atom1 = kws_dat$KAT6A$atomic_layer,
                               case1 = kws_dat$KAT6A$patient_index,
                               B2 = kws_dat$KAT6A$B, atom2 = kws_dat$KAT6A$atomic_layer,
                               case2 = kws_dat$KAT6A$patient_index)
  gr_k$num_cg_sig <- temp$num_sig
  gr_temp <- gr_k
  temp <- data.frame(layer=layer_num, datasets="KAT6A",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=NA, frac_sig_same_dir=NA,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w)
                     )
  range_overlaps <- rbind(range_overlaps,temp)
  
  temp <- frac_dbeta_same_dir( gr12 = gr_s,
                               B1 = kws_dat$Sotos$B, atom1 = kws_dat$Sotos$atomic_layer,
                               case1 = kws_dat$Sotos$patient_index,
                               B2 = kws_dat$Sotos$B, atom2 = kws_dat$Sotos$atomic_layer,
                               case2 = kws_dat$Sotos$patient_index)
  gr_s$num_cg_sig <- temp$num_sig
  gr_temp <- gr_s
  temp <- data.frame(layer=layer_num, datasets="Sotos",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=NA, frac_sig_same_dir=NA,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w))
  range_overlaps <- rbind(range_overlaps,temp)
  
  temp <- frac_dbeta_same_dir( gr12 = gr_w,
                               B1 = kws_dat$Weaver$B, atom1 = kws_dat$Weaver$atomic_layer,
                               case1 = kws_dat$Weaver$patient_index,
                               B2 = kws_dat$Weaver$B, atom2 = kws_dat$Weaver$atomic_layer,
                               case2 = kws_dat$Weaver$patient_index)
  gr_w$num_cg_sig <- temp$num_sig
  gr_temp <- gr_w
  temp <- data.frame(layer=layer_num, datasets="Weaver",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=NA, frac_sig_same_dir=NA,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w))
  range_overlaps <- rbind(range_overlaps,temp)
  
  
  gr_ks <- intersect(gr_k, gr_s)
  gr_kw <- intersect(gr_k, gr_w)
  gr_sw <- intersect(gr_s, gr_w)
  gr_ksw <- intersect(gr_ks,gr_w)


    
  for(i in 1:length(gr_ks)){
    temp <- findOverlaps(gr_ks[i],gene_ranges)
    gr_ks$overlapping_genes[i] <-  paste(genes[temp@to,]$gene_name, collapse = "," )
    gr_ks$gene_overlap_count[i] <-  countOverlaps(gr_ks[i], gene_ranges)
  }
  temp <- frac_dbeta_same_dir( gr12 = gr_ks,
                               B1 = kws_dat$KAT6A$B, atom1 = kws_dat$KAT6A$atomic_layer,
                               case1 = kws_dat$KAT6A$patient_index,
                               B2 = kws_dat$Sotos$B, atom2 = kws_dat$Sotos$atomic_layer,
                               case2 = kws_dat$Sotos$patient_index)
  gr_ks$num_cg_sig <- temp$num_sig
  gr_ks$frac_shared_sig <- temp$frac_shared_sig
  gr_ks$frac_sig_same_dir <- temp$frac_sig_same_dir
  gr_temp <- gr_ks
  temp <- data.frame(layer=layer_num, datasets="KAT6A-Sotos",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=gr_temp$frac_shared_sig, frac_sig_same_dir=gr_temp$frac_sig_same_dir,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w))
  range_overlaps <- rbind(range_overlaps,temp)
  
  
  for(i in 1:length(gr_kw)){
    temp <- findOverlaps(gr_kw[i],gene_ranges)
    gr_kw$overlapping_genes[i] <-  paste(genes[temp@to,]$gene_name, collapse = "," )
    gr_kw$gene_overlap_count[i] <-  countOverlaps(gr_kw[i], gene_ranges)
  }
  temp <- frac_dbeta_same_dir( gr12 = gr_kw,
                               B1 = kws_dat$KAT6A$B, atom1 = kws_dat$KAT6A$atomic_layer,
                               case1 = kws_dat$KAT6A$patient_index,
                               B2 = kws_dat$Weaver$B, atom2 = kws_dat$Weaver$atomic_layer,
                               case2 = kws_dat$Weaver$patient_index)
  gr_kw$num_cg_sig <- temp$num_sig
  gr_kw$frac_shared_sig <- temp$frac_shared_sig
  gr_kw$frac_sig_same_dir <- temp$frac_sig_same_dir
  gr_temp <- gr_kw
  temp <- data.frame(layer=layer_num, datasets="KAT6A-Weaver",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=gr_temp$frac_shared_sig, frac_sig_same_dir=gr_temp$frac_sig_same_dir,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w))
  range_overlaps <- rbind(range_overlaps,temp)
  
  for(i in 1:length(gr_sw)){
    temp <- findOverlaps(gr_sw[i],gene_ranges)
    gr_sw$overlapping_genes[i] <-  paste(genes[temp@to,]$gene_name, collapse = "," )
    gr_sw$gene_overlap_count[i] <-  countOverlaps(gr_sw[i], gene_ranges)
  }
  temp <- frac_dbeta_same_dir( gr12 = gr_sw,
                               B1 = kws_dat$Sotos$B, atom1 = kws_dat$Sotos$atomic_layer,
                               case1 = kws_dat$Sotos$patient_index,
                               B2 = kws_dat$Weaver$B, atom2 = kws_dat$Weaver$atomic_layer,
                               case2 = kws_dat$Weaver$patient_index)
  gr_sw$num_cg_sig <- temp$num_sig
  gr_sw$frac_shared_sig <- temp$frac_shared_sig
  gr_sw$frac_sig_same_dir <- temp$frac_sig_same_dir
  gr_temp <- gr_sw
  temp <- data.frame(layer=layer_num, datasets="Sotos-Weaver",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=gr_temp$frac_shared_sig, frac_sig_same_dir=gr_temp$frac_sig_same_dir,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w))
  range_overlaps <- rbind(range_overlaps,temp)
  
  ## gene set overlaps
  
  geneset_ks <- intersect(unique(genes[(findOverlaps(gr_k, gene_ranges)@to),]$gene_name),
                          unique(genes[(findOverlaps(gr_s, gene_ranges)@to),]$gene_name))
  geneset_kw <- intersect(unique(genes[(findOverlaps(gr_k, gene_ranges)@to),]$gene_name),
                          unique(genes[(findOverlaps(gr_w, gene_ranges)@to),]$gene_name))
  geneset_sw <- intersect(unique(genes[(findOverlaps(gr_s, gene_ranges)@to),]$gene_name),
                          unique(genes[(findOverlaps(gr_w, gene_ranges)@to),]$gene_name))
  geneset_ksw <-  intersect(intersect(unique(genes[(findOverlaps(gr_k, gene_ranges)@to),]$gene_name),
                                      unique(genes[(findOverlaps(gr_s, gene_ranges)@to),]$gene_name)),
                            unique(genes[(findOverlaps(gr_w, gene_ranges)@to),]$gene_name))

  geneset_overlaps <- rbind(geneset_overlaps, cbind(layer = layer_num, overlap_between = "KAT6A & Sotos", gene = geneset_ks) )
  geneset_overlaps <- rbind(geneset_overlaps, cbind(layer = layer_num, overlap_between = "KAT6A & Weaver", gene =  geneset_kw) )
  geneset_overlaps <- rbind(geneset_overlaps, cbind(layer = layer_num, overlap_between = "Sotos & Weaver", gene = geneset_sw) )
  geneset_overlaps <- rbind(geneset_overlaps, cbind(layer = layer_num, overlap_between = "KAT6A & Sotos & Weaver", gene = geneset_ksw) )
  
  


  range_overlaps
}

write.csv(test, file=paste(results_dir,"range_overlaps2.csv",sep = ""), row.names = F )
write.csv(geneset_overlaps, file=paste(results_dir,"geneset_overlaps2.csv",sep = ""), row.names = F )
```

```{r syndrome overlaps}

##calculate odds ratios
library("epitools")
library(ggplot2)
library(dplyr)
##

hist(atom_KAT6A$scoring_values)
hist(atom_Sotos$scoring_values)
hist(atom_Weaver$scoring_values)


cg_grs <- GRanges(seqnames = atomic_layer$chr,
                  ranges = IRanges( start = atomic_layer$pos, width = 1))

syndrome_cg_overlaps <- matrix(nrow = 7, ncol = 5)
rownames(syndrome_cg_overlaps) <- c("cg_k","cg_s","cg_w","cg_ks","cg_kw","cg_sw","cg_ksw")
colnames(syndrome_cg_overlaps) <- paste("L", 1:5, sep = "_")
syndrome_cg_overlaps_expect <- syndrome_cg_overlaps
syndrome_cg_overlaps_or <- syndrome_cg_overlaps
syndrome_cg_overlaps_or_ci <- syndrome_cg_overlaps
for(i in 1:5){
  ds <- "KAT6A"
  temp_k <- GRanges(seqnames = kws_dat[[ds]]$built_layers[[i]]$chr ,
                  ranges = IRanges( start = kws_dat[[ds]]$built_layers[[i]]$start_pos,
                                    end = kws_dat[[ds]]$built_layers[[i]]$stop_pos))
  ds <- "Sotos"
  temp_s <- GRanges(seqnames = kws_dat[[ds]]$built_layers[[i]]$chr ,
                  ranges = IRanges( start = kws_dat[[ds]]$built_layers[[i]]$start_pos,
                                    end = kws_dat[[ds]]$built_layers[[i]]$stop_pos))
  ds <- "Weaver"
  temp_w <- GRanges(seqnames = kws_dat[[ds]]$built_layers[[i]]$chr ,
                  ranges = IRanges( start = kws_dat[[ds]]$built_layers[[i]]$start_pos,
                                    end = kws_dat[[ds]]$built_layers[[i]]$stop_pos))
  
  cg_k <- unique(findOverlaps(cg_grs, temp_k)@from)
  cg_s <- unique(findOverlaps(cg_grs, temp_s)@from)
  cg_w <- unique(findOverlaps(cg_grs, temp_w)@from)
  cg_ks <- intersect(cg_k,cg_s)
  cg_kw <- intersect(cg_k,cg_w)
  cg_sw <- intersect(cg_s,cg_w)
  cg_ksw <- intersect(cg_k, cg_sw)
  syndrome_cg_overlaps["cg_k",i] <- length(cg_k)
  syndrome_cg_overlaps["cg_s",i] <- length(cg_s)
  syndrome_cg_overlaps["cg_w",i] <- length(cg_w)
  syndrome_cg_overlaps["cg_ks",i] <- length(cg_ks)
  syndrome_cg_overlaps["cg_kw",i] <- length(cg_kw)
  syndrome_cg_overlaps["cg_sw",i] <- length(cg_sw)
  syndrome_cg_overlaps["cg_ksw",i] <- length(cg_ksw)
  
  
  syndrome_cg_overlaps_expect["cg_ks",i] <- length(cg_k)*length(cg_s) / nrow(atomic_layer)
  syndrome_cg_overlaps_expect["cg_kw",i] <- length(cg_k)*length(cg_w) / nrow(atomic_layer)
  syndrome_cg_overlaps_expect["cg_sw",i] <- length(cg_s)*length(cg_w) / nrow(atomic_layer)
  syndrome_cg_overlaps_expect["cg_ksw",i] <- length(cg_kw)*length(cg_s) / nrow(atomic_layer)
  
  syn_or <- function(cg1, cg2, nt){
    cg12 <- length(intersect(cg1,cg2))
    cg_n1_y2 <- length(cg2) - cg12
    cg_n2_y1 <- length(cg1) - cg12
    cg_n12 <- nt - cg12 - cg_n1_y2 - cg_n2_y1
    return( round((cg12 * cg_n12) / (cg_n1_y2 * cg_n2_y1),2))
  }
  syn_or_CI95 <- function(cg1, cg2, nt){
    cg12 <- length(intersect(cg1,cg2))
    cg_n1_y2 <- length(cg2) - cg12
    cg_n2_y1 <- length(cg1) - cg12
    cg_n12 <- nt - cg12 - cg_n1_y2 - cg_n2_y1
    log_or <- log((cg12 * cg_n12) / (cg_n1_y2 * cg_n2_y1))
    se_log_oe <-  sqrt(1/cg12 + 1/cg_n1_y2 + 1/cg_n2_y1 + 1/cg_n12)
    left <-  log_or - 1.96*se_log_oe
    right <-  log_or + 1.96*se_log_oe
    return( paste(round(exp(left),2), "-",round(exp(right),2),sep = "")   )
  }
  
  syndrome_cg_overlaps_or["cg_ks",i] <- syn_or(cg_k,cg_s,nrow(atomic_layer))
  syndrome_cg_overlaps_or["cg_kw",i] <- syn_or(cg_k,cg_w,nrow(atomic_layer))
  syndrome_cg_overlaps_or["cg_sw",i] <- syn_or(cg_s,cg_w,nrow(atomic_layer))
  syndrome_cg_overlaps_or["cg_ksw",i] <- syn_or(cg_k,cg_sw,nrow(atomic_layer))
  
    syndrome_cg_overlaps_or_ci["cg_ks",i] <- syn_or_CI95(cg_k,cg_s,nrow(atomic_layer))
  syndrome_cg_overlaps_or_ci["cg_kw",i] <- syn_or_CI95(cg_k,cg_w,nrow(atomic_layer))
  syndrome_cg_overlaps_or_ci["cg_sw",i] <- syn_or_CI95(cg_s,cg_w,nrow(atomic_layer))
  syndrome_cg_overlaps_or_ci["cg_ksw",i] <- syn_or_CI95(cg_k,cg_sw,nrow(atomic_layer))
}


write.csv(syndrome_cg_overlaps,
          file = paste(results_dir,"syndrome_cg_overlaps.csv", sep = ""))
write.csv(syndrome_cg_overlaps_expect,
          file = paste(results_dir,"syndrome_cg_overlaps_expect.csv", sep = ""))
write.csv(syndrome_cg_overlaps/syndrome_cg_overlaps_expect,
          file = paste(results_dir,"syndrome_cg_overlaps_obs_over_exp.csv", sep = ""))
write.csv(syndrome_cg_overlaps_or,
          file = paste(results_dir,"syndrome_cg_overlaps_or.csv", sep = ""))
write.csv(syndrome_cg_overlaps_or_ci,
          file = paste(results_dir,"syndrome_cg_overlaps_or_ci.csv", sep = ""))


syndrome_cg_overlaps_or_ci
or_w_ci <- matrix(paste(syndrome_cg_overlaps_or, " (", syndrome_cg_overlaps_or_ci, ")", sep = ""), nrow=7,ncol=5)
#or_w_ci <- matrix(nrow = 7, ncol = 5)
rownames(or_w_ci) <- c("cg_k","cg_s","cg_w","cg_ks","cg_kw","cg_sw","cg_ksw")
colnames(or_w_ci) <- paste("L", 1:5, sep = "_")
write.csv(or_w_ci,
          file = paste(results_dir,"syndrome_cg_overlaps_or_w_ci.csv", sep = ""))

```


