---
title: "real_data_results_analysis"
output: html_document
---


```{r}
library(IRanges)
library(GenomicRanges)
library(data.table)
library(stringr)
library(foreach)
library(doParallel)
registerDoParallel(detectCores())
library(Gviz)
library(ggplot2)
```


```{r }
load("method_set_list.Rdata")
load("data_set_list.Rdata")

```

```{r add significance values to data_set_lists}
# for(i in 1:length(data_set_list)){
#   data_set <- data_set_list[[i]]
#   temp <- DMRscaler::run_MWW(data_set$g1, data_set$g2, data_set$B)
#   data_set_list[[i]]$locs$pval <- temp$p_val
# }
# save(data_set_list, file="data_set_list.Rdata")
```

```{r check method_set parameters }
for(i in 1:length(method_set_list)){
  print(method_set_list[[i]]$function_call )
}
```


```{r add granges for called dmrs}
filenames <- list.files(path = "./results/real_dataset_results/", pattern = "*result.csv", full.names = T)


data_set_results <- list()

for(data_set_id in 1:length(data_set_list)){
  for(method_id in 1:length(method_set_list)){
    filename <- filenames[grep(paste("data_set_",data_set_id,
                                     "__method_set_",method_id,
                                     sep=""), filenames)]
    result <- fread(filename)
    # if(method_set_list[[method_id]]$method=="dmrscaler"){
    #   result <- result[grep("64",result$layer),]
    # }
    
    locs <- data_set_list[[data_set_id]]$locs
    chr_seqlengths <- numeric(length=length(unique(locs$chr)))
    names(chr_seqlengths) <- unique(locs$chr)
    for(chr in unique(locs$chr)){
      chr_seqlengths[chr] <- max(locs$pos[which(locs$chr==chr)])+1
    }
    chr_seqinfo <- Seqinfo(names(chr_seqlengths), seqlengths = chr_seqlengths)
    
    data_set_results[[basename(filename)]]$grange <- GRanges(
      seqnames = result$chr,
      ranges = IRanges(start = result$start,
                       end = result$stop,
                       pval_region = result$pval_region),
      strand = rep("*",nrow(result)),
      seqinfo = chr_seqinfo
    )
    if(grepl("dmrscaler",names(method_set_list)[method_id] )){
      data_set_results[[basename(filename)]]$grange$layer <- result$layer
    }
    ### need to set a cutoff to region p-value for TP,FP, FN, etc
    which <- which(data_set_results[[basename(filename)]]$grange$pval_region < 0.05)
    data_set_results[[basename(filename)]]$grange <- data_set_results[[basename(filename)]]$grange[which]
    
    data_set_results[[basename(filename)]]$locs_grs <- GRanges(seqnames=locs$chr, ranges = IRanges(start=locs$pos, width=1))
    data_set_results[[basename(filename)]]$method <- method_set_list[[method_id]]$method
    data_set_results[[basename(filename)]]$method_id <- names(method_set_list)[method_id]
    data_set_results[[basename(filename)]]$data_set <- data_set_list[[data_set_id]]$data_set_name
    
    
  }
}
```

```{r write DMR width percentiles}
library("stringr")

quants_list <- list()
quants <- data.frame(percentile = numeric(), value = numeric(), method = character(), pars = character(), chrs = character())
ds_index <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", ds_index, sep = ""), names(data_set_results))
for(rs_index in result_set_ids){
  result_set <- data_set_results[[rs_index]] 
  if(grepl("dmrscaler", result_set$method)){
    if(grepl("_1", result_set$method_id)){
      for(l in 1:length(unique(result_set$grange$layer))){
          which <- which(seqnames(result_set$grange) == "chrX" &
                           result_set$grange$layer==unique(result_set$grange$layer)[l])
          temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
          quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                             method = "dmrscaler",
                                             pars = paste("dmrscaler",unique(result_set$grange$layer)[l]),
                                             chrs="chrX"))
          which <- which(seqnames(result_set$grange) != "chrX" & 
                           seqnames(result_set$grange) != "chrY"  &
                           result_set$grange$layer==unique(result_set$grange$layer)[l])
          temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
          quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                             method = "dmrscaler",
                                             pars =paste("dmrscaler",unique(result_set$grange$layer)[l]),
                                             chrs="Autosome"))
      }
    }
  } else {
    which <- which(seqnames(result_set$grange) == "chrX")
    temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
    quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                       method = result_set$method, pars = result_set$method_id,
                                       chrs="chrX"))
    which <- which(seqnames(result_set$grange) != "chrX" & seqnames(result_set$grange) != "chrY")
    temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
    quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                       method = result_set$method, pars = result_set$method_id,
                                       chrs="Autosome"))
  }
}
quants_list[[ names(data_set_list)[ds_index] ]] <- quants


data_set_ids <- grep("Sex", names(data_set_list), invert = T)
for(ds_index in data_set_ids){
  quants <- data.frame(percentile = numeric(), value = numeric(),
                       method = character(), pars = character(), chrs = character())
  result_set_ids <- grep(paste("data_set_", ds_index, sep = ""), names(data_set_results))
  for(rs_index in result_set_ids){
    result_set <- data_set_results[[rs_index]] 
    if(grepl("dmrscaler", result_set$method)){
      if(grepl("_1", result_set$method_id)){
        for(l in 1:length(unique(result_set$grange$layer))){
            which <- which(result_set$grange$layer==unique(result_set$grange$layer)[l])
            temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
            quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                               method = "dmrscaler",
                                               pars = paste("dmrscaler",unique(result_set$grange$layer)[l])))
        }
      }
    } else {
      temp <- quantile(log10(end(result_set$grange) - start(result_set$grange)), seq(0.01,1,by=0.01), type=1)
      quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                         method = result_set$method, pars = result_set$method_id))
    }
  }
  quants_list[[ names(data_set_list)[ds_index] ]] <- quants
}


for(i in 1:length(quants_list)){
  quants <- quants_list[[i]]
  gg<-ggplot(quants, aes(x=percentile, y=value, color=pars))+
    geom_line()+
    ylab("-log10(Width)")+
    theme(text=element_text(size=18))
  if(grepl("Sex",names(quants_list)[i])){
    gg <- gg + facet_grid(rows = vars(chrs),cols = vars(method) )
  } else{
    gg <- gg + facet_grid(cols = vars(method) )
  }
  plot(gg)

}

library(gridExtra)
library(grid)
filename=paste(results_dir,"/_w_combp_Width_Percentile_Plot.png", sep = "")
#png(filename = filename, width = 10, height = 6, units = "in", res = 240)
grid.newpage()
grid.arrange(gg, nrow=1)
#dev.off()  

quants$width <- 10^quants$value
#filename=paste(results_dir, "_w_combp_Width_Percentile.csv", sep = "")
#write.csv(file = filename, quants)





```

```{r}

gtf_df <- rtracklayer::import("../../STABLE_DATA/HUMAN_GENOME/Homo_sapiens.GRCh37.87.chr.gtf")
gtf_df <- as.data.frame(gtf_df)
gtf_df$seqnames<-paste("chr", gtf_df$seqnames, sep = "")

genes<-as.data.frame(cbind(chromosome=gtf_df$seqnames, start=gtf_df$start, end=gtf_df$end ))
genes<-cbind(genes, width=as.numeric(as.character(gtf_df$width)), strand=gtf_df$strand, feature=gtf_df$type, gene=gtf_df$gene_id, exon=gtf_df$exon_id, transcript=gtf_df$transcript_id, symbol=gtf_df$gene_name)

genes$start<-as.numeric(as.character(genes$start))
genes$end<-as.numeric(as.character(genes$end))


```

```{r}
data_set_regions_df <- data.frame("data_set"=c("Sotos"),
                                  "chr"=c("chr5"),
                                  "start"=c(140090846),
                                  "stop"=c(140984057))

for(i in 1:nrow(data_set_regions_df)){
  data_set <- data_set_list[[data_set_regions_df$data_set[i] ]] 
  g1 <- data_set$g1
  g2 <- data_set$g2
  
  group<-character(length = ncol(data_set$B))
  group[which(is.element(colnames(data_set$B),g1))] <- data_set$g1g2_labels$g1
  group[which(is.element(colnames(data_set$B),g2))] <- data_set$g1g2_labels$g2

  group <- as.factor(group)
  color=c("blue", "seagreen3")
  linewide=c(2, 2 )
  
  chr <- data_set_regions_df$chr
  start <- data_set_regions_df$start
  stop <- data_set_regions_df$stop
  
  
  genes_exons<-genes[which((genes$chr==chr) & (genes$start>=start) & (genes$end<=stop)),]
  genes_exons<-droplevels(genes_exons)
  genes_exons<-genes_exons[complete.cases(genes_exons),]
  
  
  gen<-"hg19"
  
  which<-which(data_set$locs$chr==chr & data_set$locs$pos>=start & data_set$locs$pos<=stop )
  B_sub<- data_set$B[which,]
  B_norm<-B_sub - rowMeans(B_sub[,g1])
  locs <- data_set$locs[which,]
  
  delta_means <- rowMeans(B_sub[,g1])-rowMeans(B_sub[,g2])
  
  
  irange<-IRanges(start=locs$pos, width=1)
  grange<-GRanges(seqnames=chr, ranges =  ranges(irange), mcols = B_sub )
  
  manhattan_range <- GRanges(seqnames=chr, ranges =  ranges(irange) )
  manhattan_track <- DataTrack(range = manhattan_range, data = -log10(locs$pval),
                               lwd=1.5, col.histogram=c("black"), type=c("hist"), ylim=c(0,max(-log10(locs$pval))))

 
  raw_beta_track<-DataTrack(range=grange, data=t(B_sub), groups=group,
                            genome=gen, name="Beta", col=color, type=c("a","g", "confint"),
                            lwd=linewide, fontcolor.title="grey20", col.axis="black",
                            fontcolor.legend="grey20", legend = TRUE, lwd.grid=0.8,
                            col.grid="grey65",v=0 )
  displayPars(raw_beta_track) <- list( cex.legend=1, cex.axis=1, cex.title=1 )

  norm_beta_track<-DataTrack(range=grange, data=delta_means,  genome=gen,
                             group ="delta", name="Delta Beta", col="black",
                             type=c("a", "g"), fontcolor.title="grey20",col.axis="black",
                             fontcolor.legend="grey20",legend = TRUE, baseline=0,
                             col.baseline="black", lty.baseline=3)
  displayPars(norm_beta_track) <- list( cex.legend=1, cex.axis=1, cex.title=1 , lwd.grid=1, col.grid="grey80",v=0,h=0)
  
  
  grtrack <- GeneRegionTrack(genes_exons, genome=gen, chromosome=chr,
                             name="Gene Model", transcriptAnnotation="symbol",
                             collapseTranscripts = "meta")
  displayPars(grtrack) <- list( background.panel="white", col="grey20", fontsize=25, 
                                fontsize.group=25, fill = "grey50", lwd=1, 
                                fontcolor.group="grey20", fontcolor.title="grey20",
                                stackHeight = 0.95, shape="box")
  if( length(which)==0 ){
    grtrack <- itrack
  }
  
  dmr_tracks <- list()
  
  data_set_id <- grep(data_set$data_set_name, names(data_set_list))
  result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
  for(rs_index in result_set_ids){
    result_set <- data_set_results[[rs_index]] 
    if(grepl("dmrscaler", result_set$method)){
      if(grepl("_2", result_set$method_id)){
        for(l in 1:length(unique(result_set$grange$layer))){
          which_in_layer <- which(result_set$grange$layer==unique(result_set$grange$layer)[l]
                                  & seqnames(result_set$grange)==chr )
          dmr_tracks[[length(dmr_tracks)+1]] <- AnnotationTrack(start = start(result_set$grange[which_in_layer]) , 
                                                                  end =  end(result_set$grange[which_in_layer]),
                                                                  chromosome = chr,
                                                                  strand = "*", genome = "hg19",
                                                                  name=unique(result_set$grange$layer)[l])
          displayPars(dmr_tracks[[length(dmr_tracks)]]) <- list(cex.legend=0.7, cex.axis=0.7, cex.title=0.7, rotation.title=0, col="grey20", fill="grey70", fontcolor.title="grey20", stackHeight = 1, shape="box")
        }
      }
    } else {
      which_in_anno <- which(seqnames(result_set$grange)==chr)
      dmr_tracks[[length(dmr_tracks)+1]] <- AnnotationTrack(start = start(result_set$grange[which_in_anno]) , 
                                                              end =  end(result_set$grange[which_in_anno]),
                                                              chromosome = chr,
                                                              strand = "*", genome = "hg19",
                                                              name=result_set$method_id)
      displayPars(dmr_tracks[[length(dmr_tracks)]]) <- list(cex.legend=0.7, cex.axis=0.7, cex.title=0.7, rotation.title=0, col="grey20", fill="grey70", fontcolor.title="grey20", stackHeight = 1, shape="box")
      
    }
  }
  
  gtrack<-GenomeAxisTrack()
  displayPars(gtrack) <- list(cex=1, cex.id=1, fontcolor="grey20", col="grey20")
  itrack<-IdeogramTrack(genome=gen,chromosome = chr)
  displayPars(itrack) <- list(cex=1,cex.bands=1, fontcolor="grey20")
  track_list <- c(list(itrack, gtrack, raw_beta_track, norm_beta_track, grtrack),dmr_tracks )
  plotTracks(track_list,
             from = start-1, to=stop+1, # sizes =c(0.5,0.75,3,2,3,0.35,0.35,0.35,0.35,0.35),
             background.title = "white"
            )

  
}


```

```{r}

```

```{r}

```


```{r make XY sex analysis results table}

library("gt")

results_table <- data.frame("method"=character(),
                            "chrX_num_dmrs"=numeric(), 
                            "chrX_mean_width"=numeric(),
                            "chrX_median_width"=numeric(),
                            "chrX_percent_total_width"=numeric(),
                            "autosome_num_dmrs"=numeric(), 
                            "autosome_avg_width"=numeric(),
                            "autosome_median_width"=numeric(),                            
                            "autosome_percent_total_width"=numeric())


data_set_id <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
for(rs_index in result_set_ids){
  result_set <- data_set_results[[rs_index]] 
  if(grepl("dmrscaler", result_set$method)){
    if(grepl("_1", result_set$method_id)){
      for(l in 1:length(unique(result_set$grange$layer))){
        which_x <- which(result_set$grange@seqnames=="chrX" &
                           result_set$grange$layer == unique(result_set$grange$layer)[l])
        which_not_x <- which(result_set$grange@seqnames != "chrX" & result_set$grange@seqnames != "chrY" &
                           result_set$grange$layer == unique(result_set$grange$layer)[l])
        results_table <- rbind(results_table,
                               data.frame("method" = result_set$method,
                                          "chrX_num_dmrs" = length(which_x),
                                          "chrX_mean_width" = mean(width(result_set$grange[which_x,])),
                                          "chrX_median_width" = median(width(result_set$grange[which_x,])), 
                                          "chrX_percent_total_width" = sum(width(result_set$grange[which_x,])) / 
                                            seqlengths(result_set$grange)["chrX"],
                                          "autosome_num_dmrs" = length(which_not_x),
                                          "autosome_mean_width" = mean(width(result_set$grange[which_not_x,])),
                                          "autosome_median_width"=median(width(result_set$grange[which_not_x,])),
                                          "autosome_percent_total_width" = sum(width(result_set$grange[which_not_x,])) / 
                                            sum(seqlengths(result_set$grange)[!grepl("[XY]",names(seqlengths(result_set$grange)))])
                         ))
        rownames(results_table)[nrow(results_table)] <- unique(result_set$grange$layer)[l] 
      }
    }
  }else{
    which_x <- which(result_set$grange@seqnames=="chrX")
    which_not_x <- which(result_set$grange@seqnames != "chrX" & result_set$grange@seqnames != "chrY")
    results_table <- rbind(results_table,
                           data.frame("method" = result_set$method,
                                      "chrX_num_dmrs" = length(which_x),
                                      "chrX_mean_width" = mean(width(result_set$grange[which_x,])),
                                      "chrX_median_width" = median(width(result_set$grange[which_x,])), 
                                      "chrX_percent_total_width" = sum(width(result_set$grange[which_x,])) / 
                                        seqlengths(result_set$grange)["chrX"],
                                      "autosome_num_dmrs" = length(which_not_x),
                                      "autosome_mean_width" = mean(width(result_set$grange[which_not_x,])),
                                      "autosome_median_width"=median(width(result_set$grange[which_not_x,])),
                                      "autosome_percent_total_width" = sum(width(result_set$grange[which_not_x,])) / 
                                        sum(seqlengths(result_set$grange)[!grepl("[XY]",names(seqlengths(result_set$grange)))])
                     ))
    rownames(results_table)[nrow(results_table)] <- result_set$method_id
  }

}


results_table$method_params <- rownames(results_table)

clean_results_table <- gt(results_table, groupname_col = "method", rowname_col = "method_params") 
clean_results_table <- tab_header(clean_results_table, title = "Sex Analysis DMR Summary Table")
clean_results_table <- tab_spanner(clean_results_table, label="chrX",
                                   columns = c("chrX_num_dmrs","chrX_mean_width","chrX_median_width", "chrX_percent_total_width") )
clean_results_table <- tab_spanner(clean_results_table, label="Autosomes",
                                   columns = c("autosome_num_dmrs","autosome_mean_width","autosome_median_width", "autosome_percent_total_width") )
clean_results_table <- tab_stubhead(clean_results_table , "method")

clean_results_table <- cols_label(clean_results_table,
                                  chrX_num_dmrs = html(" #\n<br>DMRs "), chrX_mean_width = html(" mean\n<br>DMR width "),
                                  chrX_median_width = html(" median\n<br>DMR width "), chrX_percent_total_width = html(" % total\n<br>width "),
                                  autosome_num_dmrs = html(" #\n<br>DMRs "), autosome_mean_width = html(" mean\n<br>DMR width "),
                                  autosome_median_width = html(" median\n<br>DMR width "), autosome_percent_total_width = html(" % total\n<br>width " ))


clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_percent_total_width, autosome_percent_total_width),
                                  n_sigfig = 2, scale_by = 100, pattern = "{x}%")

clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_mean_width),
                                  rows = (autosome_mean_width > 1000) & (autosome_mean_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_mean_width),
                                  rows = (autosome_mean_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")

clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_mean_width),
                                  rows = (chrX_mean_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_mean_width),
                                  rows = (chrX_mean_width > 1000) & (chrX_mean_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_mean_width),
                                  rows = (chrX_mean_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")

clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_median_width),
                                  rows = (chrX_median_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_median_width),
                                  rows = (chrX_median_width > 1000) & (chrX_median_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_median_width),
                                  rows = (chrX_median_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")

clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_median_width),
                                  rows = (autosome_median_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_median_width),
                                  rows = (autosome_median_width > 1000) & (autosome_median_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_median_width),
                                  rows = (autosome_median_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")


clean_results_table

gtsave(clean_results_table, filename = paste( "Sex_Analysis_DMR_Summary_Table_2.png",sep=""))

```

```{r make XY analysis circos network plot}

library(networkD3)
library(rjson)

data_set_id <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
method_set_id <- grep("dmrscaler_1", names(method_set_list))
method_set_ids <- grep(paste("method_set_", method_set_id, sep = ""), names(data_set_results))
result_set <- data_set_results[[ intersect(result_set_ids, method_set_ids ) ]] 



full_node_set <- list()
for(j in 1:length(unique(result_set$grange$layer))){
  #which <- which(built_layers[[j]]$chr =="chrX")
  #which <- 1:length(built_layers[[j]]$chr)
  gr_nodes <- list()
  which <- which(result_set$grange$layer == unique(result_set$grange$layer)[j]) 
  for(i in 1:length(which)){
    temp_name <- paste(as.character(seqnames(result_set$grange[ which[i],])), start(result_set$grange[which[i],]), end(result_set$grange[which[i],]), sep = "_")
   # temp_name <- paste(format(start(x_base[i,]),big.mark=","), format(end(x_base[i,]),big.mark=","), sep = " - ")
    temp_child <- list("name"= temp_name, "alt_name"= temp_name, "grange"=result_set$grange[which[i],], "children"=list() )
    gr_nodes[[i]] <-  list("name"= temp_name, "alt_name"= temp_name, "grange"=result_set$grange[which[i],], "children"=list())
  }
  full_node_set[[j]]<-gr_nodes
}

temp_full_node_set <- full_node_set
#full_node_set <- temp_full_node_set

for(j in 2:length(full_node_set)){
  print(j)
  
  lower <- full_node_set[[j-1]]
  upper <- full_node_set[[j]]
  
  lower_gr <- lower[[1]]$grange
  for(i in 2:length(lower)){
    lower_gr <- c(lower_gr, lower[[i]]$grange)
  }
  upper_gr <- upper[[1]]$grange
  if(length(upper) >= 2){
    for(i in 2:length(upper)){
      upper_gr <- c(upper_gr, upper[[i]]$grange)
    }
  }
  
  hits <- GenomicRanges::findOverlaps(lower_gr,upper_gr,type = "within");hits
  for( i in 1:length(hits@from)){
    upper[[ hits@to[i] ]]$children <- c(upper[[ hits@to[i] ]]$children, list(lower[[ hits@from[i] ]]))
    if(upper[[ hits@to[i] ]]$name == lower[[ hits@from[i] ]]$name |
       upper[[ hits@to[i] ]]$alt_name == lower[[ hits@from[i] ]]$name  |
       upper[[ hits@to[i] ]]$name == lower[[ hits@from[i] ]]$alt_name
       ){  ### avoid redundant child-parent pairs
      upper[[ hits@to[i] ]]$name <- "..."
    }
  }
  full_node_set[[j]]<-upper
}

#dmr_tree <- full_node_set[[3]][[1]]
#dmr_tree <- list("name"="root","children"=full_node_set[[3]])
dmr_tree <- list("name"="root","children"=full_node_set[[length(full_node_set)]])

dmr_tree_clean <- list()

node <- dmr_tree



strip_grs <- function(node){
  node <- list("name" = node$name, "children" = node$children)
  if(length(node$children)==0){return(node)}
  for(i in 1:length(node$children)){
    node$children[[i]]<-strip_grs(node$children[[i]])
  }
  return(node)
}
dmr_tree_clean <- strip_grs(dmr_tree)


library("webshot") ###  library to fix an obnoxious problem

temp <- radialNetwork(List = dmr_tree_clean, fontSize = 16, fontFamily = "bold", nodeStroke = "black", linkColour = "grey60", opacity = 1, margin=0, height = 1600, width = 1600)
saveNetwork(temp,paste("All_chrom_dmrs_radialNetwork.html",sep = "") ,
            selfcontained = T)
webshot(paste("All_chrom_dmrs_radialNetwork.html",sep = ""),
        paste("All_all_layer_chrom_dmrs_radialNetwork.png",sep = ""),
        zoom = 2)


temp <- diagonalNetwork(List = dmr_tree_clean[[2]][[1]], fontSize = 12, fontFamily = "bold", nodeStroke = "black", linkColour = "black", opacity = 1,  margin=0, height = 400, width = 150)
saveNetwork(temp,paste(figures_dir, "FINAL_X2_chrom_dmrs_diagonalNetwork.html", sep="") , selfcontained = T )
webshot(paste("FINAL_X2_chrom_dmrs_diagonalNetwork.html", sep=""),
        paste("FINAL_X2_chrom_dmrs_diagonalNetwork.png", sep=""),
        vheight = 400, vwidth = 150 , zoom =10 )


```




