---
title: "real_data_figures"
author: "Leroy_Bondhus"
date: "September 19, 2020"
output: html_document
---

```{r}
library("tidyverse")
library("ggplot2")
library("gridExtra")


library("Gviz")
library(circlize)

library(HilbertCurve)
```


```{r setup for KAT6A analysis}
load("/home/leroy/Desktop/PROJECTS/dmrscaler_real_data/intermediate_data/KAT6A_GRset.funnorm.Rdata")
load("/home/leroy/Desktop/PROJECTS/dmrscaler_real_data/intermediate_data/KAT6A_dmrscaler_built_layers.Rdata")
#load("/home/leroy/Desktop/PROJECTS/dmrscaler_real_data/intermediate_data/450k_KAT6A_dmrscaler_built_layers.Rdata")

B<-getBeta(GRset.funnorm)
M<-getM(GRset.funnorm)
pdat<-pData(GRset.funnorm)

locs <- getLocations(GRset.funnorm)
locs <- data.frame("names"=locs@ranges@NAMES, "pos"=locs@ranges@start, "chr" = rep(locs@seqnames@values, locs@seqnames@lengths))


patient_index<-grep("KAT",pdat$Sample_Name)
control_index<-grep("CONTROL", pdat$Sample_Name)                      

group<-as.character(seq(from=1, to=ncol(B), by =1 ))

group[patient_index] <- "KAT6A"
group[control_index] <- "control"

group <- as.factor(group)
color=c("blue", "seagreen3")
linewide=c(2, 2)

g1 <- colnames(B)[patient_index]
g2 <- colnames(B)[control_index]

kws_dat <- list("KAT6A" = list(B=B,locs=locs,patient_index=patient_index,control_index=control_index,
                               atomic_layer=atomic_layer, built_layers=built_layers))

```

```{r setup for Weaver analysis}
load("/home/leroy/Desktop/PROJECTS/dmrscaler_real_data/intermediate_data/Weaver_GRset.funnorm.Rdata")
load("/home/leroy/Desktop/PROJECTS/dmrscaler_real_data/intermediate_data/Weaver_dmrscaler_built_layers.Rdata")
#load("/home/leroy/Desktop/PROJECTS/dmrscaler_real_data/intermediate_data/850k_Weaver_dmrscaler_built_layers.Rdata")


B<-getBeta(GRset.funnorm)
M<-getM(GRset.funnorm)
pdat<-pData(GRset.funnorm)

locs <- getLocations(GRset.funnorm)
locs <- data.frame("names"=locs@ranges@NAMES, "pos"=locs@ranges@start, "chr" = rep(locs@seqnames@values, locs@seqnames@lengths))

control_index <- grep("Control",pdat$Sample_Group)
patient_index <- grep("Weaver", pdat$Sample_Group)

group<-as.character(seq(from=1, to=ncol(B), by =1 ))
group[patient_index] <- "Weaver"
group[control_index] <- "control"

group <- as.factor(group)
color=c("blue", "seagreen3")
linewide=c(2, 2 )

g1 <- colnames(B)[patient_index]
g2 <- colnames(B)[control_index]


kws_dat[["Weaver"]] <-list(B=B, locs=locs, patient_index=patient_index, control_index=control_index,
                           atomic_layer=atomic_layer, built_layers=built_layers )

```

```{r setup for Sotos analysis}
load("/home/leroy/Desktop/PROJECTS/dmrscaler_real_data/intermediate_data/Sotos_GRset.funnorm.Rdata")
load("/home/leroy/Desktop/PROJECTS/dmrscaler_real_data/intermediate_data/Sotos_dmrscaler_built_layers.Rdata")
#load("/home/leroy/Desktop/PROJECTS/dmrscaler_real_data/intermediate_data/850k_Sotos_dmrscaler_built_layers.Rdata")


B<-getBeta(GRset.funnorm)
M<-getM(GRset.funnorm)
pdat<-pData(GRset.funnorm)

locs <- getLocations(GRset.funnorm)
locs <- data.frame("names"=locs@ranges@NAMES, "pos"=locs@ranges@start, "chr" = rep(locs@seqnames@values, locs@seqnames@lengths))

control_index <- grep("Control",pdat$Sample_Group)
patient_index <- grep("Sotos", pdat$Sample_Group)

group<-as.character(seq(from=1, to=ncol(B), by =1 ))
group[patient_index] <- "Sotos"
group[control_index] <- "control"

group <- as.factor(group)
color=c("blue", "seagreen3")
linewide=c(2, 2 )

g1 <- colnames(B)[patient_index]
g2 <- colnames(B)[control_index]

kws_dat<-list()

kws_dat[["Sotos"]] <-list(B=B, locs=locs, patient_index=patient_index, control_index=control_index,
                          atomic_layer=atomic_layer, built_layers=built_layers )

```


```{r bumphunter and dmrcate}
results_dir <-paste("/home/leroy/Desktop/PROJECTS/dmrscaler_real_data/results2/KAT6A/")

##write csv

for(i in 1:5){
  write.csv(built_layers[[i]], file = paste(results_dir, "dmrscaler_layer",i,".csv"))
}

library(DMRcate)
#g1<-pdat$Basename[which(trimws(pdat$Sex)=="female")][1:8]
#g2<-pdat$Basename[which(trimws(pdat$Sex)=="male")][1:8]

B_mod<-B[,c(g1,g2)]
                    ##   BUMPHUNTER  ##
design<-rep(-1,length(colnames(B_mod)))
design[which(is.element(colnames(B_mod),g1))]<-1
design<-cbind(rep(1,length(colnames(B_mod) ) ), design )

bumps_default <- bumphunter(B_mod,as.matrix(design), cutoff=0.1 , chr = locs$chr,B=250, pos=locs$pos)
## save output to csv
write.csv(bumps_default$table, paste(results_dir,"bumps_default.csv", sep = ""))



bumps_1Mb <- bumphunter(B_mod,as.matrix(design),chr = locs$chr, pos=locs$pos, cutoff=0.1, maxGap=1e6, B=250, smoothFunction=loessByCluster )
## save output to csv
write.csv(bumps_1Mb$table, paste(results_dir,"bumps_1Mb.csv", sep = ""))


# 
# temp_names <- paste("called", bumps4$table$chr, bumps4$table$start, sep = "_")
# bh_called_grs <- GRanges(
#   seqnames = bumps4$table$chr,
#   ranges = IRanges(start = bumps4$table$start, end = bumps4$table$end, names = temp_names),
#   strand = rep("*", nrow(bumps4$table))
# )

                ## DMRcate  ##

M_mod<-M[,c(g1,g2)]
colnames(design) <- c("(Intercept)","(Intercept)")
myannotation <- cpg.annotate("array", M_mod, what="M", arraytype = "450K",analysis.type="differential", design=design, coef=2, fdr = 0.1)
dmrcate_dmrs_default <- dmrcate(myannotation, lambda=1000, C=2)
## save output to csv
write.csv(dmrcate_dmrs_default$results, paste(results_dir,"dmrcate_dmrs_default.csv", sep = ""))
dmrcate_dmrs_default  <- extractRanges(dmrcate_dmrs_default , genome = "hg19")

dmrcate_dmrs_1M <- dmrcate(myannotation, lambda=1e6, C=2000)
## save output to csv
write.csv(dmrcate_dmrs_1M$results, paste(results_dir,"dmrcate_dmrs_1M.csv", sep = ""))


dmrcate_dmrs_1M  <- extractRanges(dmrcate_dmrs_1M , genome = "hg19")




       ## combp  ##
# clean up a bit
rm(M)
  ## Set up bed file
locs <- atomic_layer
combp_input_bed <- data.frame(chrom=locs$chr,start=locs$pos,end=locs$pos+1,pval=10^-atomic_layer$scoring_values)
combp_input_bed <- combp_input_bed[order(as.character(combp_input_bed$chrom)),]
colnames(combp_input_bed)[1] <- "#chrom"
data.table::fwrite(combp_input_bed, file = "./intermediate_data2/combp_temp_input.bed", row.names = F,col.names = T, sep = "\t")
rm(combp_input_bed)
## Pass bed file into comb-p
Sys.time()
system("comb-p pipeline -c 4 --dist 1000 --step 100 --seed 1e-3 -p ./intermediate_data2/test --region-filter-p 0.1 ./intermediate_data2/combp_temp_input.bed ")
Sys.time() 
## Parse comb-p resuults into the CALLED_TP_FP and the MAPPING VALUES tables
print("done with 1st combp")
combp_results_1k_100 <- data.table::fread("./intermediate_data2/test.regions-t.bed")
write.csv(combp_results_1k_100, paste(results_dir,"combp_results_1k_100.csv", sep = ""))

Sys.time()
print("start with 10k,1k")
system("comb-p pipeline -c 4 --dist 10000 --step 1000 --seed 1e-3 -p ./intermediate_data2/test --region-filter-p 0.1 ./intermediate_data2/combp_temp_input.bed ")
Sys.time() 
## Parse comb-p resuults into the CALLED_TP_FP and the MAPPING VALUES tables
combp_results_10k_1k <- data.table::fread("./intermediate_data2/test.regions-t.bed")
write.csv(combp_results_10k_1k, paste(results_dir,"combp_results_10k_1k.csv", sep = ""))
print("done with 1st combp")


Sys.time()
print("start with 100k 10k")
system("comb-p pipeline -c 4 --dist 100000 --step 10000 --seed 1e-3 -p ./intermediate_data2/test --region-filter-p 0.1 ./intermediate_data2/combp_temp_input.bed ")
Sys.time() 
## Parse comb-p resuults into the CALLED_TP_FP and the MAPPING VALUES tables
combp_results_100k_10k <- data.table::fread("./intermediate_data2/test.regions-t.bed")
write.csv(combp_results_100k_10k, paste(results_dir,"combp_results_100k_10k.csv", sep = ""))
print("done with 2nd combp")

save.image()
```

```{r write DMR width percentiles}
library("stringr")

temp <- str_split(dmrcate_dmrs_1M$coord, "[:-]", simplify = T)
dmrcate_dmrs_1M$width <- as.numeric(temp[,3])-as.numeric(temp[,2])
dmrcate_dmrs_1M$chr <- temp[,1]


temp <- str_split(dmrcate_dmrs_default$coord, "[:-]", simplify = T)
dmrcate_dmrs_default$width <- as.numeric(temp[,3])-as.numeric(temp[,2])
dmrcate_dmrs_default$chr <- temp[,1]

quants <- data.frame(percentile = numeric(), value = numeric(), method = character(), pars = character())
for(i in 1:length(built_layers)){
  temp <- quantile(log10(built_layers[[i]]$stop_pos - built_layers[[i]]$start_pos), seq(0.01,1,by=0.01), type=1)
  quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                     method = "DMRscaler", pars = paste("DMRscaler_layer",i,sep = "")) )
}
temp <- quantile(log10(bumps_default$end - bumps_default$start), seq(0.01,1,by=0.01), type=1)
quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                      method = "bumphunter",pars="bumphunter_default"))
temp <- quantile(log10(bumps_1Mb$end - bumps_1Mb$start), seq(0.01,1,by=0.01), type=1)
quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                     method = "bumphunter",pars="bumphunter:maxWidth=1Mb"))

temp <- quantile(log10(combp_results_1k_100$end - combp_results_1k_100$start), seq(0.01,1,by=0.01), type=1)
quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                      method = "combp",pars="combp:dist=1kb,step=100bp"))
temp <- quantile(log10(combp_results_1M_5k$end - combp_results_1M_5k$start), seq(0.01,1,by=0.01), type=1)
quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                     method = "combp",pars="combp:dist=1Mb,step=5kb"))

temp <- quantile(log10(dmrcate_dmrs_default$width), seq(0.01,1,by=0.01), type=1)
quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                     method = "DMRcate",pars="dmrcate_default"))
temp <- quantile(log10(dmrcate_dmrs_1M$width), seq(0.01,1,by=0.01), type=1)
quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                     method = "DMRcate",pars="dmrcate:lambda=1Mb,C=2000"))


gg<-ggplot(quants, aes(x=percentile, y=value, color=pars))+
  geom_line()+
  ylab("-log10(Width)")+
  facet_grid(cols = vars(method) )+
  theme(text=element_text(size=18))
gg

library(gridExtra)
library(grid)
filename=paste(results_dir,"/_w_combp_Width_Percentile_Plot.png", sep = "")
png(filename = filename, width = 10, height = 6, units = "in", res = 240)
grid.newpage()
grid.arrange(gg, nrow=1)
dev.off()  

quants$width <- 10^quants$value
filename=paste(results_dir, "_w_combp_Width_Percentile.csv", sep = "")
write.csv(file = filename, quants)

```


```{r Gene Annotations}

gtf_df <- rtracklayer::import("/home/lbondhus/Desktop/STABLE_DATA/HUMAN_GENOME/Homo_sapiens.GRCh37.87.chr.gtf")
gtf_df <- as.data.frame(gtf_df)
gtf_df$seqnames<-paste("chr", gtf_df$seqnames, sep = "")

genes<-as.data.frame(cbind(chromosome=gtf_df$seqnames, start=gtf_df$start, end=gtf_df$end ))
genes<-cbind(genes, width=as.numeric(as.character(gtf_df$width)), strand=gtf_df$strand, feature=gtf_df$type, gene=gtf_df$gene_id, exon=gtf_df$exon_id, transcript=gtf_df$transcript_id, symbol=gtf_df$gene_name)

genes$start<-as.numeric(as.character(genes$start))
genes$end<-as.numeric(as.character(genes$end))

```


```{r build_plotting_funcs}
B<-getBeta(GRset.funnorm)
normB <- B-rowMeans(B[,control_index])
deltaMeans <- rowMeans(B[,patient_index])-rowMeans(B[,control_index]) 
deltaMeans <- data.frame(names = names(deltaMeans), deltaMeans=deltaMeans)

## is CG j in a DMR a level i
locs$dmr_layers <- 0
pos_gr <- GRanges(seqnames = locs$chr, ranges = IRanges(start = locs$pos, width = 1))
for(i in 1:length(built_layers)){
    bl_gr <- GRanges(seqnames = built_layers[[i]]$chr, ranges = IRanges(start = built_layers[[i]]$start_pos, end = built_layers[[i]]$stop_pos))
    locs$dmr_layers <- locs$dmr_layers + GenomicRanges::countOverlaps(pos_gr, bl_gr)
    
}
#chr16 chr2
#61777508 80.1e6 81825714 140166213 27127501
#63968522 84.7e6 84541711 140811253 27291695
# chr2 81825714 84541711



HOX <- data.frame(cluster=c("A","B","C","D", "N"), 
          chr=c("chr7","chr17","chr12", "chr2", "chr5"),
          start_pos=c(27.12e6, 46.60e6, 54.3e6, 176.9e6, 140.11e6),
          stop_pos=c(27.3e6, 46.72e6, 54.5e6, 177.07e6, 140.83e6))
  
#for(i in c(30,213,221)){
#for(i in 1:nrow(built_layers[[6]]) ){
#for(i in   which(built_layers[[6]]$unsigned_bin_sig > 0.99995 &  built_layers[[6]]$numCG > 50) ){
 # temp <- built_layers[[6]][i,]
for(i in 1:nrow(HOX)){
  temp<-HOX[i,]
  
  chr <- as.character(temp$chr)
  start <- temp$start_pos
  stop <- temp$stop_pos
#  start <- temp$start_pos - floor((temp$stop_pos - temp$start_pos)/10)
#  stop <- temp$stop_pos + floor((temp$stop_pos - temp$start_pos)/10)
  
  genes_exons<-genes[which((genes$chr==chr) & (genes$start>=start) & (genes$end<=stop)),]
## PCDH organization
  genes_exons<- genes_exons[c(grep("HOX",genes_exons$symbol),grep("EVX",genes_exons$symbol)),]
    ##
  genes_exons<-droplevels(genes_exons)
  genes_exons<-genes_exons[complete.cases(genes_exons),]
  
  cex_size=7
  cex_small=5
  gen<-"hg19"
  
  
  which<-which(locs$chr==chr & locs$pos>=start & locs$pos<=stop )
  Bt<-B[which,]
  Bt_norm<-normB[which,]
  
  irange<-IRanges(start=locs[which,"pos"], width=1)
  grange<-GRanges(seqnames=chr, ranges =  ranges(irange), mcols = Bt )
  
  manhattan_range <- GRanges(seqnames=chr, ranges =  ranges(irange), mcols = atomic_layer[which,] )
  manhattan_track <- DataTrack(range = manhattan_range, data = atomic_layer[which,]$scoring_value, lwd=1.5, col.histogram=c("black"), type=c("hist"), ylim=c(0,max(atomic_layer$scoring_values)))

#  raw_beta_track<-DataTrack(range=grange, data=t(Bt), groups=group, genome=gen, name="Beta", col=color, type=c("a","g"), lwd=linewide, fontcolor.title="grey20", col.axis="black", fontcolor.legend="grey20", legend = FALSE, lwd.grid=0.8, col.grid="grey65", ylim=c(0,1))
 
  raw_beta_track<-DataTrack(range=grange, data=t(Bt), groups=group, genome=gen, name="Beta", col=color, type=c("a","g", "confint"), lwd=linewide, fontcolor.title="grey20", col.axis="black", fontcolor.legend="grey20", legend = TRUE, lwd.grid=0.8, col.grid="grey65",v=0 )
  displayPars(raw_beta_track) <- list( cex.legend=1, cex.axis=1, cex.title=1 )

  norm_beta_track<-DataTrack(range=grange, data=deltaMeans[which,],  genome=gen, group ="delta", name="Delta Beta", col="black", type=c("a", "g"), fontcolor.title="grey20", col.axis="black", fontcolor.legend="grey20", legend = TRUE, baseline=0, col.baseline="black", lty.baseline=3)
  displayPars(norm_beta_track) <- list( cex.legend=1, cex.axis=1, cex.title=1 , lwd.grid=1, col.grid="grey80",v=0,h=0)
  
 # raw_beta_track<-DataTrack(range=grange, data=cMeans, genome=gen, name="Beta", col="black", type=c("b", "g"), fontcolor.title="grey20", col.axis="black", fontcolor.legend="grey20", ylim=c(0,1))
 # displayPars(raw_beta_track) <- list( cex.legend=1, cex.axis=1, cex.title=1 )
  
 # norm_beta_track<-DataTrack(range=grange, data=t(Bt_norm), groups=group, genome=gen, name="Delta Beta", col=color, lwd=linewide, type=c("a", "g" ,"confint"), fontcolor.title="grey20", col.axis="black", fontcolor.legend="grey20")
 # displayPars(norm_beta_track) <- list( cex.legend=1, cex.axis=1, cex.title=1 , lwd.grid=1, col.grid="grey80",v=0)
  
  
  grtrack <- GeneRegionTrack(genes_exons, genome=gen, chromosome=chr, name="Gene Model", transcriptAnnotation="symbol",collapseTranscripts = "meta")
  displayPars(grtrack) <- list( background.panel="white", col="grey20", fontsize=25, fontsize.group=25, fill = "grey50", lwd=1, fontcolor.group="grey20", fontcolor.title="grey20", stackHeight = 0.95, shape="box")
  if( length(which)==0 ){
    grtrack <- itrack
  }
  
  layer_track <- list()
for(j in 1:length(built_layers)){
  layer_track[[j]] <- AnnotationTrack(start = built_layers[[j]]$start_pos, 
                                      end =  built_layers[[j]]$stop_pos,
                                      chromosome = built_layers[[j]]$chr,
                                      strand = "*", genome = "hg19", name=paste("layer", j, sep = "_"))
  displayPars(layer_track[[j]]) <- list(cex.legend=0.7, cex.axis=0.7, cex.title=0.7, rotation.title=0, col="grey20", fill="grey70", fontcolor.title="grey20", stackHeight = 1, shape="box")
}
gtrack<-GenomeAxisTrack()
displayPars(gtrack) <- list(cex=1, cex.id=1, fontcolor="grey20", col="grey20")
itrack<-IdeogramTrack(genome=gen,chromosome = chr)
displayPars(itrack) <- list(cex=1,cex.bands=1, fontcolor="grey20")

  
  Region_name <- paste(chr, start, stop, sep="_")
  filename=paste("/home/lbondhus/Desktop/PROJECTS/dmrscaler_real_data/figures/WEAVER/3_", Region_name , ".png", sep = "")
 # filename=paste("/home/lbondhus/Desktop/PROJECTS/dmrscaler_real_data/figures/KAT6A/", "HOX" , HOX$cluster[i], ".png", sep = "")
    png(file=filename, width = 12, height = 8, units = "in", res = 450) ####### start writing 
  
  #plotTracks(list(itrack, gtrack, raw_beta_track, norm_beta_track, manhattan_track, grtrack,
  plotTracks(list(itrack, gtrack, raw_beta_track, norm_beta_track, grtrack,
                  layer_track[[1]],
                  layer_track[[2]],
                  layer_track[[3]],
                  layer_track[[4]],
                  layer_track[[5]]
                  ),
             from = start-1, to=stop+1, sizes =c(0.5,0.75,3,2,3,0.35,0.35,0.35,0.35,0.35),
            background.title = "white"
             )
  #  plotTracks(list(itrack, gtrack, raw_beta_track, norm_beta_track, manhattan_track, grtrack),
  #             from = start-1, to=stop+1, sizes =c(0.5,0.5,1,4,2,2) )
  dev.off() ####  stop writing
  
  
  ## adjacency plots
  col_fun = colorRamp2(c(0,fdrscaler-fdrscaler*0.9,fdrscaler,max(atomic_layer$scoring_values)), c("grey30", "grey60", "grey60", "red"))
  level = 2
  
  hc_points <- atomic_layer[which,]
  ggp_sig_adj <- ggplot(hc_points, aes(x=1:length(scoring_values), y=scoring_values)) +
    geom_segment(aes(xend=1:length(scoring_values), y=0, yend=scoring_values, alpha=0.5)) +
    geom_point(aes(color= scoring_values), size=0.7) +
    scale_color_gradient2(low="grey60", mid="grey30", high = "red", midpoint = fdrscaler)+
  xlab("CG index over region") + ylab("-log10(pval)")+
  ggtitle(paste("Adjacent_CG_plot_", chr,"_" ,start, "_", stop))+
  ylab("-log10pval") +
  theme_bw()+
  theme(
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_blank(), axis.title.x = element_blank(),
    text = element_text(size=12, color = "black"),
    axis.text = element_text(size=12, color = "black"),
    legend.position = "none"
  )
  #png(filename = filename,  width = 4, height = 6, units = "in", res = 450)
  
  plot(ggp_sig_adj)  
  #dev.off()
  
  
  
  df_betas <- cbind("index" = 1:length(which), locs[which,], "mean"= rowMeans2(B[which,g2]), "status"=levels(group)[1] )
  df_betas <- rbind(df_betas, cbind("index" = 1:length(which), locs[which,], "mean" = rowMeans2(B[which, g1]), "status"=levels(group)[2]))
                    
  ggp_beta_raw <- ggplot(df_betas)+
    geom_line(aes(x=index, y=mean, group=names, alpha=0.5))+
    geom_point(aes(x=index, y=mean, color=status), size=0.7)+
    scale_color_manual(values=color)+
  ylab("Beta") +
  theme_bw()+
  theme(
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_blank(), axis.title.x = element_blank(),
    text = element_text(size=12, color = "black"),
    axis.text = element_text(size=12, color = "black"),
    legend.position = "none"
  )
  
  plot(ggp_beta_raw)
  
    
  df_betas <- cbind("index" = 1:length(which), locs[which,], "mean"= rowMeans2(normB[which,g2]), "status"=levels(group)[1] )
  df_betas <- rbind(df_betas, cbind("index" = 1:length(which), locs[which,], "mean" = rowMeans2(normB[which, g1]), "status"=levels(group)[2]))
                    
  ggp_beta_norm <- ggplot(df_betas)+
    geom_line(aes(x=index, y=mean, group=names, alpha=0.5))+
    geom_point(aes(x=index, y=mean, color=status), size=0.7)+
    scale_color_manual(values=color)+
  ylab("Normalized Beta") +
  theme_bw()+
  theme(
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_blank(), axis.title.x = element_blank(),
    text = element_text(size=12, color = "black"),
    axis.text = element_text(size=12, color = "black"),
    legend.position = "none"
  )
  
  plot(ggp_beta_norm) 
  

    temp <- cbind("index" = 1:length(which), locs[which,])
    ggp_dmr_layers <- ggplot(temp)+
  #  geom_line(aes(x=index, y=dmr_layers )) +
  #  geom_point(aes(x=index, y=dmr_layers )) +
    geom_bar(aes(x=index, y=dmr_layers, color=dmr_layers), fill="grey70",  stat="identity")+
    scale_color_gradient(low="grey50", high = "grey50")+
    scale_y_continuous(breaks = c(0,1,2,3,4,5), labels = c("none","64_CG_layer","32_CG_layer","16_CG_layer","8_CG_layer","4_CG_layer"))+
    ylab("Layer") +
    xlab("CG index over region") +
    theme_bw()+
    theme(
      plot.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      text = element_text(size=12, color = "black"),
      axis.text = element_text(size=12, color = "black"),
      legend.position = "none"
    )
  plot(ggp_dmr_layers)
  filename=paste("/home/lbondhus/Desktop/PROJECTS/dmrscaler_real_data/figures/WEAVER/Adjacent_CG_plots_", chr,"_" ,start, "_", stop, ".png", sep = "")  
  png(filename = filename, width = 8, height = 6, units = "in", res = 320)

    grid.newpage()
  grid.draw(rbind(ggplotGrob(ggp_sig_adj), ggplotGrob(ggp_beta_raw), ggplotGrob(ggp_beta_norm), ggplotGrob(ggp_dmr_layers) ) )

  dev.off() 
}  

```



```{r hilbert curves}
library(circlize)

library(HilbertCurve)

col_fun = colorRamp2(c(0,fdrscaler*0.99,fdrscaler,max(atomic_layer$scoring_values)), c("grey30", "grey60", "red", "red"))
level = 6
## 4^level - 1 = # segments
for(i in 2){
  filename <- paste("/home/lbondhus/Desktop/PROJECTS/dmrscaler_real_data/figures/WEAVER/Hilbert_curves_chr", chr,"_highlight" ,start, "_", stop, ".png", sep = "")  
  png(filename = filename, width = 6, height = 6, units = "in", res = 320)
  
  hc_points <- atomic_layer[which(atomic_layer$chr==paste("chr", i, sep = "")),]
  
  hc_points <- atomic_layer[which(atomic_layer$chr==paste("chr", i, sep = "")),]
  hc_max <- nrow(hc_points)
  #np <- ceiling(hc_max / (4^level - 1))
  
#  hc_col <- rep("grey10", nrow(hc_points))
#  hc_col[which(hc_points$scoring_values>fdrscaler)] <- "red"
  hc_col <- col_fun(hc_points$scoring_values)
  hc_size <- hc_points$scoring_values / max(hc_points$scoring_values)
#  hc_size <- hc_points$scoring_values / fdrscaler
  hc <- HilbertCurve(s=1,e=hc_max, level = level, reference = F)
  
  hc_polygon(hc, IRanges(start=min(which(hc_points$pos >= HOX[5,]$start_pos)), end = max(which(hc_points$pos <= HOX[5,]$stop_pos))), gp = gpar(col = "blue")) 
  hc_points(hc, x1 = 1:nrow(hc_points), np = NULL, pch=15, size = unit(hc_size*2, "mm"),gp = gpar(col = hc_col, fill = hc_col))
  

  dev.off()
}

```

```{r bumphunter and dmrcate}
#g1<-pdat$Basename[which(trimws(pdat$Sex)=="female")][1:8]
#g2<-pdat$Basename[which(trimws(pdat$Sex)=="male")][1:8]
M <- getM(GRset.funnorm)
M <- M[,c(g1,g2)]
B_mod<-B[,c(g1,g2)]
                    ##   BUMPHUNTER  ##
design<-rep(-1,length(colnames(B_mod)))
design[which(is.element(colnames(B_mod),g1))]<-1
design<-cbind(rep(1,length(colnames(B_mod) ) ), design )

bumps_default <- bumphunter(B_mod,as.matrix(design),chr = locs$chr, pos=locs$pos, cutoff=0.1, B=250, smoothFunction=loessByCluster )

bumps_1Mb <- bumphunter(B_mod,as.matrix(design),chr = locs$chr, pos=locs$pos, cutoff=0.1, maxGap=1e6, B=250, smoothFunction=loessByCluster )
# 
# temp_names <- paste("called", bumps4$table$chr, bumps4$table$start, sep = "_")
# bh_called_grs <- GRanges(
#   seqnames = bumps4$table$chr,
#   ranges = IRanges(start = bumps4$table$start, end = bumps4$table$end, names = temp_names),
#   strand = rep("*", nrow(bumps4$table))
# )

                ## DMRcate  ##

M_mod<-M[,c(g1,g2)]
colnames(design) <- c("(Intercept)","(Intercept)")
myannotation <- cpg.annotate("array", M_mod, what="M", arraytype = "EPIC",analysis.type="differential", design=design, coef=2, fdr = 0.1)
dmrcate_dmrs_default <- dmrcate(myannotation, lambda=1000, C=2)
dmrcate_dmrs_default  <- extractRanges(dmrcate_dmrs_default , genome = "hg19")
dmrcate_dmrs_1M <- dmrcate(myannotation, lambda=1e6, C=2000)
dmrcate_dmrs_1M  <- extractRanges(dmrcate_dmrs_1M , genome = "hg19")
```

```{r make pretty table}
test<-built_layers[[5]]
test <- test[1:10,c(1:4,7)]

library("gt")

test_table <- gt(test, groupname_col = "chr") 
test_table <- tab_header(test_table, title = "test", subtitle = "subtitle")
test_table <- tab_spanner(test_table, label="position", columns = c("start_pos","stop_pos") )
test_table

results_table <- data.frame("method"=character(),
                            "autosome_num_dmrs"=numeric(), 
                            "autosome_avg_width"=numeric(),
                            "autosome_median_width"=numeric(),                            
                            "autosome_percent_total_width"=numeric())
for( i in 1:length(built_layers)){
  which_not <- which(built_layers[[i]]$chr != "chrX" & built_layers[[i]]$chr != "chrY")
  results_table <- rbind(results_table,
                         data.frame("method" = "dmrscaler", 
                         "autosome_num_dmrs" = nrow(built_layers[[i]][which_not,]),
                         "autosome_mean_width" = mean(built_layers[[i]]$stop_pos[which_not]-built_layers[[i]]$start_pos[which_not]),
                         "autosome_median_width"= median(built_layers[[i]]$stop_pos[which_not]-built_layers[[i]]$start_pos[which_not]),               
                         "autosome_percent_total_width" = -1 # add late
                         ))
  rownames(results_table)[i] <- paste("layer ",i,": ",layer_sizes[i] ," adjacent CpGs", sep = "")
}

### Add bumphunter results to table

which_not <- which(bumps_default$chr != "chrX" & bumps_default$chr != "chrY")
results_table <- rbind(results_table,
                         data.frame("method" = "bumphunter", 
                         "autosome_num_dmrs" = nrow(built_layers[[i]][which_not,]),
                         "autosome_mean_width" = mean(bumps_default$end[which_not]-bumps_default$start[which_not]),
                         "autosome_median_width"= median(bumps_default$end[which_not]-bumps_default$start[which_not]),                            
                         "autosome_percent_total_width" = -1 # add late
                         ))
rownames(results_table)[length(rownames(results_table))] <- "default: maxGap = 500 bp"

which_not <- which(bumps_1Mb$chr != "chrX" & bumps_1Mb$chr != "chrY")
results_table <- rbind(results_table,
                         data.frame("method" = "bumphunter", 
                         "autosome_num_dmrs" = nrow(built_layers[[i]][which_not,]),
                         "autosome_mean_width" = mean(bumps_1Mb$end[which_not]-bumps_1Mb$start[which_not]),
                         "autosome_median_width"= median(bumps_1Mb$end[which_not]-bumps_1Mb$start[which_not]),                            
                         "autosome_percent_total_width" = -1 # add late
                         ))
rownames(results_table)[length(rownames(results_table))] <- "maxGap = 1 Mb"


### Add combp results to table

which_not <- which(combp_results_1k_100$X.chrom != "chrX" & combp_results_1k_100$X.chrom != "chrY")
results_table <- rbind(results_table,
                         data.frame("method" = "combp", 
                         "autosome_num_dmrs" = nrow(built_layers[[i]][which_not,]),
                         "autosome_mean_width" = mean(combp_results_1k_100$end[which_not]-combp_results_1k_100$start[which_not]),
                         "autosome_median_width"= median(combp_results_1k_100$end[which_not]-combp_results_1k_100$start[which_not]),                            
                         "autosome_percent_total_width" = -1 # add late
                         ))
rownames(results_table)[length(rownames(results_table))] <- "dist = 1kb, step = 100 bp"

which_not <- which(combp_results_1M_5k$X.chrom != "chrX" & combp_results_1M_5k$X.chrom != "chrY")
results_table <- rbind(results_table,
                         data.frame("method" = "combp",
                         "autosome_num_dmrs" = nrow(built_layers[[i]][which_not,]),
                         "autosome_mean_width" = mean(combp_results_1M_5k$end[which_not]-combp_results_1M_5k$start[which_not]),
                         "autosome_median_width"= median(combp_results_1M_5k$end[which_not]-combp_results_1M_5k$start[which_not]),
                         "autosome_percent_total_width" = -1 # add late
                         ))
rownames(results_table)[length(rownames(results_table))] <- "dist = 1 Mb, step = 5 kb"

# which_not <- which(combp_results_100k_10k$X.chrom != "chrX" & combp_results_100k_10k$X.chrom != "chrY")
# results_table <- rbind(results_table,
#                          data.frame("method" = "combp", 
#                          "autosome_num_dmrs" = nrow(built_layers[[i]][which_not,]),
#                          "autosome_mean_width" = mean(combp_results_100k_10k$end[which_not]-combp_results_100k_10kstart[which_not]),
#                          "autosome_median_width"= median(combp_results_100k_10k$end[which_not]-combp_results_100k_10k$start[which_not]),                            
#                          "autosome_percent_total_width" = -1 # add late
#                          ))
# rownames(results_table)[length(rownames(results_table))] <- "dist = 100kb, step = 10 kb"



### add dmrcate results to table
widths <- dmrcate_dmrs_default$width
seqs <- dmrcate_dmrs_default$chr

which_not <- which(seqs != "chrX" & seqs != "chrY")
results_table <- rbind(results_table,
                         data.frame("method" = "dmrcate", 
                         "autosome_num_dmrs" = length(which_not),
                         "autosome_mean_width" = mean(widths[which_not]),
                         "autosome_median_width"= median(widths[which_not]),                            
                         "autosome_percent_total_width" = -1 # add late
                         ))
rownames(results_table)[length(rownames(results_table))] <- "default: lambda = 1 kb, C = 2 "

widths <- dmrcate_dmrs_1M$width
seqs <- dmrcate_dmrs_1M$chr

which_not <- which(seqs != "chrX" & seqs != "chrY")
results_table <- rbind(results_table,
                         data.frame("method" = "dmrcate", 
                         "autosome_num_dmrs" = length(which_not),
                         "autosome_mean_width" = mean(widths[which_not]),
                         "autosome_median_width"= median(widths[which_not]),                            
                         "autosome_percent_total_width" = -1 # add late
                         ))
rownames(results_table)[length(rownames(results_table))] <- "lambda = 1 Mb, C = 2000"





autosome_size <- 0
for(i in 1:22){
  chr <- paste("chr",i,sep = "")
  autosome_size <- autosome_size + max(locs$pos[which(locs$chr==chr)])
}

results_table$autosome_percent_total_width <- results_table$autosome_mean_width * results_table$autosome_num_dmrs / autosome_size

results_table$method_params <- rownames(results_table)

clean_results_table <- gt(results_table, groupname_col = "method", rowname_col = "method_params") 
clean_results_table <- tab_header(clean_results_table, title = "KAT6A Analysis DMR Summary Table")
clean_results_table <- tab_spanner(clean_results_table, label="Autosomes",
                                   columns = c("autosome_num_dmrs","autosome_mean_width","autosome_median_width", "autosome_percent_total_width") )
clean_results_table <- tab_stubhead(clean_results_table , "method")


clean_results_table <- cols_label(clean_results_table,
                                  autosome_num_dmrs = html(" #\n<br>DMRs "), autosome_mean_width = html(" mean\n<br>DMR width "),
                                  autosome_median_width = html(" median\n<br>DMR width "), autosome_percent_total_width = html(" % total\n<br>width " ))



clean_results_table <- fmt_number(clean_results_table, columns = vars( autosome_percent_total_width),
                                  n_sigfig = 2, scale_by = 100, pattern = "{x}%")

clean_results_table <- fmt_number(clean_results_table, columns = vars(autosome_mean_width),
                                  rows = (autosome_mean_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = vars(autosome_mean_width),
                                  rows = (autosome_mean_width > 1000) & (autosome_mean_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = vars(autosome_mean_width),
                                  rows = (autosome_mean_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")


clean_results_table <- fmt_number(clean_results_table, columns = vars(autosome_median_width),
                                  rows = (autosome_median_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = vars(autosome_median_width),
                                  rows = (autosome_median_width > 1000) & (autosome_median_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = vars(autosome_median_width),
                                  rows = (autosome_median_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")


clean_results_table

gtsave(clean_results_table, filename = paste(results_dir,"Analysis_DMR_summary_table.png",sep = ""))
 
# 
#  fmt_scientific(
#     columns = vars(num),
#     rows = num <= 500,
#     decimals = 1
#   )
#  fmt_number(
#     columns = vars(num),
#     rows = num > 500,
#     decimals = 1,
#     scale_by = 1/1000,
#     pattern = "{x}K"
#   )
```



```{r dmr gene analysis}
genes <- gtf_df[which(gtf_df$gene_biotype =="protein_coding" ),]
genes <- genes[which(genes$gene_source=="ensembl_havana"),]
genes <- genes[which(genes$type=="gene" ),]

gene_ranges <- GRanges(seqnames = genes$seqnames, IRanges(start = genes$start, width = genes$width))
dmr_ranges <- GRanges(seqnames = built_layers[[5]]$chr, IRanges(start = built_layers[[5]]$start_pos, end = built_layers[[5]]$stop_pos))

built_layers[[5]]$gene_overlap_count <- countOverlaps(dmr_ranges, gene_ranges)
table(countOverlaps(dmr_ranges, gene_ranges))

for(i in 1:length(dmr_ranges)){
  temp <- findOverlaps(dmr_ranges[i], gene_ranges)
  built_layers[[5]]$genes_overlapped[i] <- paste(genes[temp@to,]$gene_name, collapse = "," )
}

built_layers[[5]]$width <- built_layers[[5]]$stop_pos - built_layers[[5]]$start_pos

dmr_ranges[which(countOverlaps(dmr_ranges, gene_ranges) > 5)]
temp <- findOverlaps(dmr_ranges[which(countOverlaps(dmr_ranges, gene_ranges) > 5)], gene_ranges)
temp <- genes[temp@to,]
```


```{r funcs for kws overlap analysis}

frac_dbeta_same_dir <- function(gr12, B1, atom1, case1, B2, case2, atom2){
    num_sig <- numeric(length = length(gr12))
    frac_shared_sig <- numeric(length = length(gr12))
    frac_sig_same_dir <- numeric(length = length(gr12))
    for(i in 1:length(gr12)){
      #print(i)
      chr <- as.character(gr12@seqnames)[i]
      start <- start(gr12)[i]
      end <- end(gr12)[i]
      control1 <- (1:ncol(B1))[-case1]
      control2 <- (1:ncol(B2))[-case2]
      
      cgs <- which(atom1$chr == chr & atom1$pos <= end & atom1$pos >= start)
      
      ##hyper (caseBeta > controlBeta)
      if(length(cgs)==1){ 
        hyper1 <- mean(B1[cgs,case1]) > mean(B1[cgs,control1])
        hyper2 <- mean(B2[cgs,case2]) > mean(B2[cgs,control2]);
      }
      else{
        hyper1 <- ( rowMeans2(B1[cgs,case1]) > rowMeans2(B1[cgs,control1]) ) 
        hyper2 <- ( rowMeans2(B2[cgs,case2]) > rowMeans2(B2[cgs,control2]) ) 
      }
      sig1 <- (atom1$scoring_values[cgs] >1) ## 1 corresponds to FDR = 0.1 based on scoring func
      sig2 <- (atom2$scoring_values[cgs] >1) ## 1 corresponds to FDR = 0.1 based on scoring func
      num_sig[i] <- sum(sig1 | sig2)
      frac_shared_sig[i] <- sum(sig1 & sig2) / sum(sig1 | sig2)
      frac_sig_same_dir[i] <- sum((hyper1 == hyper2) & sig1 & sig2 )  / sum(sig1 & sig2)
      if(is.na(frac_sig_same_dir[i])){frac_sig_same_dir[i] = 0}
    }
    return( data.frame(num_sig = num_sig, frac_shared_sig = frac_shared_sig, frac_sig_same_dir = frac_sig_same_dir ))
}

frac_A_in_B <- function(gr1,gr2){
  which <- findOverlaps(gr1,gr2)@from
  frac_a_in_b <- numeric(length=length(gr1))
  for(i in 1:length(gr1)){
    if(is.element(i,which)){
      frac_a_in_b[i] <- sum(width(intersect(gr1[i],gr2)))/width(gr1[i])
    }
    else{
      frac_a_in_b[i] <- 0
    }
  }
  return(frac_a_in_b)
}
```

```{r kat6a-weaver-sotos overlap analysis}
genes <- gtf_df[which(gtf_df$gene_biotype =="protein_coding" ),]
genes <- genes[which(genes$gene_source=="ensembl_havana"),]
genes <- genes[which(genes$type=="gene" ),]
genes <- genes[,!(colSums(!is.na(genes))==0)]
gene_ranges <- GRanges(seqnames = genes$seqnames, IRanges(start = genes$start, width = genes$width))

#kws_dat_cp <- kws_dat
kws_dat <- kws_dat_cp

## overlap K-W, K-S, W-S, K-W-S
## feature overlap, bp overlap
## match layers 4-4, 8-8, 16-16, etc
## % same direction (hypo or hyper)
## genes overlapped
## OR+-CI cgs, OR+-CI gene sets, OR+-CI region
results_dir = "/home/lbondhus/Desktop/PROJECTS/dmrscaler_real_data/results/OVERLAPS/"
  
geneset_overlaps = data.frame(layer=numeric(), overlap_between = character(), gene = character())
range_overlaps <- data.frame(layer=numeric(), datasets=character(),
                             seqnames = character(), start = numeric(), width = numeric(),
                             overlapping_genes = character(), gene_overlap_count = numeric(),
                             num_cg_sig = numeric(), frac_shared_sig = numeric(), frac_sig_same_dir = numeric(),
                             frac_shared_kat6a = numeric(), frac_shared_sotos = numeric(),frac_shared_weaver = numeric() )  
for(layer_num in 1:length(kws_dat$KAT6A$built_layers)){
test <- foreach(layer_num=1:length(kws_dat$KAT6A$built_layers), .combine=rbind) %dopar% {
  geneset_overlaps = data.frame(layer=numeric(), overlap_between = character(), gene = character())
  range_overlaps <- data.frame(layer=numeric(), datasets=character(),
                             seqnames = character(), start = numeric(), width = numeric(),
                             overlapping_genes = character(), gene_overlap_count = numeric(),
                             num_cg_sig = numeric(), frac_shared_sig = numeric(), frac_sig_same_dir = numeric(),
                             frac_shared_kat6a = numeric(), frac_shared_sotos = numeric(),frac_shared_weaver = numeric() )  
  ## foreach
  
  print(layer_num)
  gr_k <- GRanges(seqnames = kws_dat$KAT6A$built_layers[[layer_num]]$chr,
                     IRanges(start = kws_dat$KAT6A$built_layers[[layer_num]]$start_pos, 
                             end = kws_dat$KAT6A$built_layers[[layer_num]]$stop_pos))
  gr_s <- GRanges(seqnames = kws_dat$Sotos$built_layers[[layer_num]]$chr,
                     IRanges(start = kws_dat$Sotos$built_layers[[layer_num]]$start_pos, 
                             end = kws_dat$Sotos$built_layers[[layer_num]]$stop_pos))
  gr_w <- GRanges(seqnames = kws_dat$Weaver$built_layers[[layer_num]]$chr,
                     IRanges(start = kws_dat$Weaver$built_layers[[layer_num]]$start_pos, 
                             end = kws_dat$Weaver$built_layers[[layer_num]]$stop_pos))
  for(i in 1:length(gr_k)){
    temp <- findOverlaps(gr_k[i], gene_ranges)
    gr_k$overlapping_genes[i] <- paste(genes[temp@to,]$gene_name, collapse = "," )
    gr_k$gene_overlap_count[i] <- countOverlaps(gr_k[i], gene_ranges)
  }
  for(i in 1:length(gr_s)){
    temp <- findOverlaps(gr_s[i], gene_ranges)
    gr_s$overlapping_genes[i] <- paste(genes[temp@to,]$gene_name, collapse = "," )
    gr_s$gene_overlap_count[i] <- countOverlaps(gr_s[i], gene_ranges)
  }
  for(i in 1:length(gr_w)){
    temp <- findOverlaps(gr_w[i], gene_ranges)
     gr_w$overlapping_genes[i] <- paste(genes[temp@to,]$gene_name, collapse = "," )
     gr_w$gene_overlap_count[i] <- countOverlaps(gr_w[i], gene_ranges)
  }
  temp <- frac_dbeta_same_dir( gr12 = gr_k,
                               B1 = kws_dat$KAT6A$B, atom1 = kws_dat$KAT6A$atomic_layer,
                               case1 = kws_dat$KAT6A$patient_index,
                               B2 = kws_dat$KAT6A$B, atom2 = kws_dat$KAT6A$atomic_layer,
                               case2 = kws_dat$KAT6A$patient_index)
  gr_k$num_cg_sig <- temp$num_sig
  gr_temp <- gr_k
  temp <- data.frame(layer=layer_num, datasets="KAT6A",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=NA, frac_sig_same_dir=NA,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w)
                     )
  range_overlaps <- rbind(range_overlaps,temp)
  
  temp <- frac_dbeta_same_dir( gr12 = gr_s,
                               B1 = kws_dat$Sotos$B, atom1 = kws_dat$Sotos$atomic_layer,
                               case1 = kws_dat$Sotos$patient_index,
                               B2 = kws_dat$Sotos$B, atom2 = kws_dat$Sotos$atomic_layer,
                               case2 = kws_dat$Sotos$patient_index)
  gr_s$num_cg_sig <- temp$num_sig
  gr_temp <- gr_s
  temp <- data.frame(layer=layer_num, datasets="Sotos",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=NA, frac_sig_same_dir=NA,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w))
  range_overlaps <- rbind(range_overlaps,temp)
  
  temp <- frac_dbeta_same_dir( gr12 = gr_w,
                               B1 = kws_dat$Weaver$B, atom1 = kws_dat$Weaver$atomic_layer,
                               case1 = kws_dat$Weaver$patient_index,
                               B2 = kws_dat$Weaver$B, atom2 = kws_dat$Weaver$atomic_layer,
                               case2 = kws_dat$Weaver$patient_index)
  gr_w$num_cg_sig <- temp$num_sig
  gr_temp <- gr_w
  temp <- data.frame(layer=layer_num, datasets="Weaver",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=NA, frac_sig_same_dir=NA,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w))
  range_overlaps <- rbind(range_overlaps,temp)
  
  
  gr_ks <- intersect(gr_k, gr_s)
  gr_kw <- intersect(gr_k, gr_w)
  gr_sw <- intersect(gr_s, gr_w)
  gr_ksw <- intersect(gr_ks,gr_w)


    
  for(i in 1:length(gr_ks)){
    temp <- findOverlaps(gr_ks[i],gene_ranges)
    gr_ks$overlapping_genes[i] <-  paste(genes[temp@to,]$gene_name, collapse = "," )
    gr_ks$gene_overlap_count[i] <-  countOverlaps(gr_ks[i], gene_ranges)
  }
  temp <- frac_dbeta_same_dir( gr12 = gr_ks,
                               B1 = kws_dat$KAT6A$B, atom1 = kws_dat$KAT6A$atomic_layer,
                               case1 = kws_dat$KAT6A$patient_index,
                               B2 = kws_dat$Sotos$B, atom2 = kws_dat$Sotos$atomic_layer,
                               case2 = kws_dat$Sotos$patient_index)
  gr_ks$num_cg_sig <- temp$num_sig
  gr_ks$frac_shared_sig <- temp$frac_shared_sig
  gr_ks$frac_sig_same_dir <- temp$frac_sig_same_dir
  gr_temp <- gr_ks
  temp <- data.frame(layer=layer_num, datasets="KAT6A-Sotos",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=gr_temp$frac_shared_sig, frac_sig_same_dir=gr_temp$frac_sig_same_dir,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w))
  range_overlaps <- rbind(range_overlaps,temp)
  
  
  for(i in 1:length(gr_kw)){
    temp <- findOverlaps(gr_kw[i],gene_ranges)
    gr_kw$overlapping_genes[i] <-  paste(genes[temp@to,]$gene_name, collapse = "," )
    gr_kw$gene_overlap_count[i] <-  countOverlaps(gr_kw[i], gene_ranges)
  }
  temp <- frac_dbeta_same_dir( gr12 = gr_kw,
                               B1 = kws_dat$KAT6A$B, atom1 = kws_dat$KAT6A$atomic_layer,
                               case1 = kws_dat$KAT6A$patient_index,
                               B2 = kws_dat$Weaver$B, atom2 = kws_dat$Weaver$atomic_layer,
                               case2 = kws_dat$Weaver$patient_index)
  gr_kw$num_cg_sig <- temp$num_sig
  gr_kw$frac_shared_sig <- temp$frac_shared_sig
  gr_kw$frac_sig_same_dir <- temp$frac_sig_same_dir
  gr_temp <- gr_kw
  temp <- data.frame(layer=layer_num, datasets="KAT6A-Weaver",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=gr_temp$frac_shared_sig, frac_sig_same_dir=gr_temp$frac_sig_same_dir,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w))
  range_overlaps <- rbind(range_overlaps,temp)
  
  for(i in 1:length(gr_sw)){
    temp <- findOverlaps(gr_sw[i],gene_ranges)
    gr_sw$overlapping_genes[i] <-  paste(genes[temp@to,]$gene_name, collapse = "," )
    gr_sw$gene_overlap_count[i] <-  countOverlaps(gr_sw[i], gene_ranges)
  }
  temp <- frac_dbeta_same_dir( gr12 = gr_sw,
                               B1 = kws_dat$Sotos$B, atom1 = kws_dat$Sotos$atomic_layer,
                               case1 = kws_dat$Sotos$patient_index,
                               B2 = kws_dat$Weaver$B, atom2 = kws_dat$Weaver$atomic_layer,
                               case2 = kws_dat$Weaver$patient_index)
  gr_sw$num_cg_sig <- temp$num_sig
  gr_sw$frac_shared_sig <- temp$frac_shared_sig
  gr_sw$frac_sig_same_dir <- temp$frac_sig_same_dir
  gr_temp <- gr_sw
  temp <- data.frame(layer=layer_num, datasets="Sotos-Weaver",
                     seqnames=as.character(seqnames(gr_temp)),start=start(gr_temp), width=width(gr_temp),
                     overlapping_genes=gr_temp$overlapping_genes, gene_overlap_count=gr_temp$gene_overlap_count,
                     num_cg_sig=gr_temp$num_cg_sig, frac_shared_sig=gr_temp$frac_shared_sig, frac_sig_same_dir=gr_temp$frac_sig_same_dir,
                     frac_shared_kat6a = frac_A_in_B(gr_temp,gr_k),
                     frac_shared_sotos = frac_A_in_B(gr_temp,gr_s),
                     frac_shared_weaver = frac_A_in_B(gr_temp,gr_w))
  range_overlaps <- rbind(range_overlaps,temp)
  
  ## gene set overlaps
  
  geneset_ks <- intersect(unique(genes[(findOverlaps(gr_k, gene_ranges)@to),]$gene_name),
                          unique(genes[(findOverlaps(gr_s, gene_ranges)@to),]$gene_name))
  geneset_kw <- intersect(unique(genes[(findOverlaps(gr_k, gene_ranges)@to),]$gene_name),
                          unique(genes[(findOverlaps(gr_w, gene_ranges)@to),]$gene_name))
  geneset_sw <- intersect(unique(genes[(findOverlaps(gr_s, gene_ranges)@to),]$gene_name),
                          unique(genes[(findOverlaps(gr_w, gene_ranges)@to),]$gene_name))
  geneset_ksw <-  intersect(intersect(unique(genes[(findOverlaps(gr_k, gene_ranges)@to),]$gene_name),
                                      unique(genes[(findOverlaps(gr_s, gene_ranges)@to),]$gene_name)),
                            unique(genes[(findOverlaps(gr_w, gene_ranges)@to),]$gene_name))

  geneset_overlaps <- rbind(geneset_overlaps, cbind(layer = layer_num, overlap_between = "KAT6A & Sotos", gene = geneset_ks) )
  geneset_overlaps <- rbind(geneset_overlaps, cbind(layer = layer_num, overlap_between = "KAT6A & Weaver", gene =  geneset_kw) )
  geneset_overlaps <- rbind(geneset_overlaps, cbind(layer = layer_num, overlap_between = "Sotos & Weaver", gene = geneset_sw) )
  geneset_overlaps <- rbind(geneset_overlaps, cbind(layer = layer_num, overlap_between = "KAT6A & Sotos & Weaver", gene = geneset_ksw) )
  
  


  range_overlaps
}

write.csv(test, file=paste(results_dir,"range_overlaps2.csv",sep = ""), row.names = F )
write.csv(geneset_overlaps, file=paste(results_dir,"geneset_overlaps2.csv",sep = ""), row.names = F )
```

```{r syndrome overlaps}

##calculate odds ratios
library("epitools")
library(ggplot2)
library(dplyr)
##

hist(atom_KAT6A$scoring_values)
hist(atom_Sotos$scoring_values)
hist(atom_Weaver$scoring_values)


cg_grs <- GRanges(seqnames = atomic_layer$chr,
                  ranges = IRanges( start = atomic_layer$pos, width = 1))

syndrome_cg_overlaps <- matrix(nrow = 7, ncol = 5)
rownames(syndrome_cg_overlaps) <- c("cg_k","cg_s","cg_w","cg_ks","cg_kw","cg_sw","cg_ksw")
colnames(syndrome_cg_overlaps) <- paste("L", 1:5, sep = "_")
syndrome_cg_overlaps_expect <- syndrome_cg_overlaps
syndrome_cg_overlaps_or <- syndrome_cg_overlaps
syndrome_cg_overlaps_or_ci <- syndrome_cg_overlaps
for(i in 1:5){
  ds <- "KAT6A"
  temp_k <- GRanges(seqnames = kws_dat[[ds]]$built_layers[[i]]$chr ,
                  ranges = IRanges( start = kws_dat[[ds]]$built_layers[[i]]$start_pos,
                                    end = kws_dat[[ds]]$built_layers[[i]]$stop_pos))
  ds <- "Sotos"
  temp_s <- GRanges(seqnames = kws_dat[[ds]]$built_layers[[i]]$chr ,
                  ranges = IRanges( start = kws_dat[[ds]]$built_layers[[i]]$start_pos,
                                    end = kws_dat[[ds]]$built_layers[[i]]$stop_pos))
  ds <- "Weaver"
  temp_w <- GRanges(seqnames = kws_dat[[ds]]$built_layers[[i]]$chr ,
                  ranges = IRanges( start = kws_dat[[ds]]$built_layers[[i]]$start_pos,
                                    end = kws_dat[[ds]]$built_layers[[i]]$stop_pos))
  
  cg_k <- unique(findOverlaps(cg_grs, temp_k)@from)
  cg_s <- unique(findOverlaps(cg_grs, temp_s)@from)
  cg_w <- unique(findOverlaps(cg_grs, temp_w)@from)
  cg_ks <- intersect(cg_k,cg_s)
  cg_kw <- intersect(cg_k,cg_w)
  cg_sw <- intersect(cg_s,cg_w)
  cg_ksw <- intersect(cg_k, cg_sw)
  syndrome_cg_overlaps["cg_k",i] <- length(cg_k)
  syndrome_cg_overlaps["cg_s",i] <- length(cg_s)
  syndrome_cg_overlaps["cg_w",i] <- length(cg_w)
  syndrome_cg_overlaps["cg_ks",i] <- length(cg_ks)
  syndrome_cg_overlaps["cg_kw",i] <- length(cg_kw)
  syndrome_cg_overlaps["cg_sw",i] <- length(cg_sw)
  syndrome_cg_overlaps["cg_ksw",i] <- length(cg_ksw)
  
  
  syndrome_cg_overlaps_expect["cg_ks",i] <- length(cg_k)*length(cg_s) / nrow(atomic_layer)
  syndrome_cg_overlaps_expect["cg_kw",i] <- length(cg_k)*length(cg_w) / nrow(atomic_layer)
  syndrome_cg_overlaps_expect["cg_sw",i] <- length(cg_s)*length(cg_w) / nrow(atomic_layer)
  syndrome_cg_overlaps_expect["cg_ksw",i] <- length(cg_kw)*length(cg_s) / nrow(atomic_layer)
  
  syn_or <- function(cg1, cg2, nt){
    cg12 <- length(intersect(cg1,cg2))
    cg_n1_y2 <- length(cg2) - cg12
    cg_n2_y1 <- length(cg1) - cg12
    cg_n12 <- nt - cg12 - cg_n1_y2 - cg_n2_y1
    return( round((cg12 * cg_n12) / (cg_n1_y2 * cg_n2_y1),2))
  }
  syn_or_CI95 <- function(cg1, cg2, nt){
    cg12 <- length(intersect(cg1,cg2))
    cg_n1_y2 <- length(cg2) - cg12
    cg_n2_y1 <- length(cg1) - cg12
    cg_n12 <- nt - cg12 - cg_n1_y2 - cg_n2_y1
    log_or <- log((cg12 * cg_n12) / (cg_n1_y2 * cg_n2_y1))
    se_log_oe <-  sqrt(1/cg12 + 1/cg_n1_y2 + 1/cg_n2_y1 + 1/cg_n12)
    left <-  log_or - 1.96*se_log_oe
    right <-  log_or + 1.96*se_log_oe
    return( paste(round(exp(left),2), "-",round(exp(right),2),sep = "")   )
  }
  
  syndrome_cg_overlaps_or["cg_ks",i] <- syn_or(cg_k,cg_s,nrow(atomic_layer))
  syndrome_cg_overlaps_or["cg_kw",i] <- syn_or(cg_k,cg_w,nrow(atomic_layer))
  syndrome_cg_overlaps_or["cg_sw",i] <- syn_or(cg_s,cg_w,nrow(atomic_layer))
  syndrome_cg_overlaps_or["cg_ksw",i] <- syn_or(cg_k,cg_sw,nrow(atomic_layer))
  
    syndrome_cg_overlaps_or_ci["cg_ks",i] <- syn_or_CI95(cg_k,cg_s,nrow(atomic_layer))
  syndrome_cg_overlaps_or_ci["cg_kw",i] <- syn_or_CI95(cg_k,cg_w,nrow(atomic_layer))
  syndrome_cg_overlaps_or_ci["cg_sw",i] <- syn_or_CI95(cg_s,cg_w,nrow(atomic_layer))
  syndrome_cg_overlaps_or_ci["cg_ksw",i] <- syn_or_CI95(cg_k,cg_sw,nrow(atomic_layer))
}


write.csv(syndrome_cg_overlaps,
          file = paste(results_dir,"syndrome_cg_overlaps.csv", sep = ""))
write.csv(syndrome_cg_overlaps_expect,
          file = paste(results_dir,"syndrome_cg_overlaps_expect.csv", sep = ""))
write.csv(syndrome_cg_overlaps/syndrome_cg_overlaps_expect,
          file = paste(results_dir,"syndrome_cg_overlaps_obs_over_exp.csv", sep = ""))
write.csv(syndrome_cg_overlaps_or,
          file = paste(results_dir,"syndrome_cg_overlaps_or.csv", sep = ""))
write.csv(syndrome_cg_overlaps_or_ci,
          file = paste(results_dir,"syndrome_cg_overlaps_or_ci.csv", sep = ""))


syndrome_cg_overlaps_or_ci
or_w_ci <- matrix(paste(syndrome_cg_overlaps_or, " (", syndrome_cg_overlaps_or_ci, ")", sep = ""), nrow=7,ncol=5)
#or_w_ci <- matrix(nrow = 7, ncol = 5)
rownames(or_w_ci) <- c("cg_k","cg_s","cg_w","cg_ks","cg_kw","cg_sw","cg_ksw")
colnames(or_w_ci) <- paste("L", 1:5, sep = "_")
write.csv(or_w_ci,
          file = paste(results_dir,"syndrome_cg_overlaps_or_w_ci.csv", sep = ""))

```



