---
title: "real_data_results_analysis"
output: html_document
---


```{r}
library(IRanges)
library(GenomicRanges)
library(data.table)
library(stringr)
library(foreach)
library(doParallel)
registerDoParallel(detectCores()-2)
library(Gviz)
library(ggplot2)
library(gridExtra)
```


```{r }
load("method_set_list.Rdata")
load("data_set_list.Rdata")

figures_dir <- "./results/figures/"
tables_dir <- "./results/tables/"
```

```{r add significance values to data_set_lists}
# for(i in 1:length(data_set_list)){
#   data_set <- data_set_list[[i]]
#   temp <- DMRscaler::run_MWW(data_set$g1, data_set$g2, data_set$B)
#   data_set_list[[i]]$locs$pval <- temp$p_val
# }
# save(data_set_list, file="data_set_list.Rdata")
```

```{r check method_set parameters }
for(i in 1:length(method_set_list)){
  print(method_set_list[[i]]$function_call )
}
```


```{r add granges for called dmrs}
filenames <- list.files(path = "./results/real_dataset_results/", pattern = "*result.csv", full.names = T)


data_set_results <- list()

for(data_set_id in 1:length(data_set_list)){
  for(method_id in 1:length(method_set_list)){
    filename <- filenames[grep(paste("data_set_",data_set_id,
                                     "__method_set_",method_id,
                                     sep=""), filenames)]
    if(length(filename)==0){next}
    result <- fread(filename)
    if(method_set_list[[method_id]]$method=="dmrscaler"){
      #result <- result[grep("64",result$layer),]
      result$pval_region <- result$pval_region_adj
    }
    
    locs <- data_set_list[[data_set_id]]$locs
    chr_seqlengths <- numeric(length=length(unique(locs$chr)))
    names(chr_seqlengths) <- unique(locs$chr)
    for(chr in unique(locs$chr)){
      chr_seqlengths[chr] <- max(locs$pos[which(locs$chr==chr)])+1
    }
    chr_seqinfo <- Seqinfo(names(chr_seqlengths), seqlengths = chr_seqlengths)
    
    data_set_results[[basename(filename)]]$grange <- GRanges(
      seqnames = result$chr,
      ranges = IRanges(start = result$start,
                       end = result$stop,
                       pval_region = result$pval_region),
      strand = rep("*",nrow(result)),
      seqinfo = chr_seqinfo
    )
    if(grepl("dmrscaler",names(method_set_list)[method_id] )){
      data_set_results[[basename(filename)]]$grange$layer <- result$layer
    }
    ### need to set a cutoff to region p-value for TP,FP, FN, etc
    which <- which(data_set_results[[basename(filename)]]$grange$pval_region < 0.05)
    data_set_results[[basename(filename)]]$grange <- data_set_results[[basename(filename)]]$grange[which]
    
    data_set_results[[basename(filename)]]$locs_grs <- GRanges(seqnames=locs$chr, ranges = IRanges(start=locs$pos, width=1))
    data_set_results[[basename(filename)]]$method <- method_set_list[[method_id]]$method
    data_set_results[[basename(filename)]]$method_id <- names(method_set_list)[method_id]
    data_set_results[[basename(filename)]]$data_set <- data_set_list[[data_set_id]]$data_set_name
    
    
  }
}


## save common formatted method results
if(TRUE){
  for(i in 1:length(data_set_results)){
    as.data.frame(data_set_results[[i]]$grange)
    filename <- paste(tables_dir,"dataset_",data_set_results[[i]]$data_set,
                      "__method_",data_set_results[[i]]$method_id, ".csv",sep="")
  write.csv(as.data.frame(data_set_results[[i]]$grange), file = filename)
  }
  
}

```

```{r write DMR width percentiles}
library("stringr")

quants_list <- list()
quants <- data.frame(percentile = numeric(), value = numeric(), method = character(), pars = character(), chrs = character())
ds_index <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", ds_index, sep = ""), names(data_set_results))
for(rs_index in result_set_ids){
  result_set <- data_set_results[[rs_index]] 
  if(grepl("dmrscaler", result_set$method)){
    if(grepl("_2", result_set$method_id)){
      for(l in 1:length(unique(result_set$grange$layer))){
          which <- which(seqnames(result_set$grange) == "chrX" &
                           result_set$grange$layer==unique(result_set$grange$layer)[l])
          temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
          quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                             method = "dmrscaler",
                                             pars = paste("dmrscaler",unique(result_set$grange$layer)[l]),
                                             chrs="chrX"))
          which <- which(seqnames(result_set$grange) != "chrX" & 
                           seqnames(result_set$grange) != "chrY"  &
                           result_set$grange$layer==unique(result_set$grange$layer)[l])
          temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
          quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                             method = "dmrscaler",
                                             pars =paste("dmrscaler",unique(result_set$grange$layer)[l]),
                                             chrs="Autosome"))
      }
    }
  } else {
    which <- which(seqnames(result_set$grange) == "chrX")
    temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
    quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                       method = result_set$method, pars = result_set$method_id,
                                       chrs="chrX"))
    which <- which(seqnames(result_set$grange) != "chrX" & seqnames(result_set$grange) != "chrY")
    temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
    quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                       method = result_set$method, pars = result_set$method_id,
                                       chrs="Autosome"))
  }
}
quants_list[[ names(data_set_list)[ds_index] ]] <- quants


data_set_ids <- grep("Sex", names(data_set_list), invert = T)
for(ds_index in data_set_ids){
  quants <- data.frame(percentile = numeric(), value = numeric(),
                       method = character(), pars = character(), chrs = character())
  result_set_ids <- grep(paste("data_set_", ds_index, sep = ""), names(data_set_results))
  for(rs_index in result_set_ids){
    result_set <- data_set_results[[rs_index]] 
    if(grepl("dmrscaler", result_set$method)){
      if(grepl("_2", result_set$method_id)){
        for(l in 1:length(unique(result_set$grange$layer))){
            which <- which(result_set$grange$layer==unique(result_set$grange$layer)[l])
            temp <- quantile(log10(end(result_set$grange[which]) - start(result_set$grange[which])), seq(0.01,1,by=0.01), type=1)
            quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                               method = "dmrscaler",
                                               pars = paste("dmrscaler",unique(result_set$grange$layer)[l])))
        }
      }
    } else {
      if(grepl("combp_2", result_set$method_id)){next}
      temp <- quantile(log10(end(result_set$grange) - start(result_set$grange)), seq(0.01,1,by=0.01), type=1)
      quants <- rbind(quants, data.frame(percentile = seq(0.01,1,by=0.01), value = temp,
                                         method = result_set$method, pars = result_set$method_id))
    }
  }
  quants_list[[ names(data_set_list)[ds_index] ]] <- quants
}


for(i in 1:length(quants_list)){
  quants <- quants_list[[i]]
  quants$pars <- as.factor(quants$pars)
  if(i == 2){
    quants$pars <- factor(quants$pars, levels=levels(quants$pars)[c(1:5,8,10,6,7,9)])
  } else {
    quants$pars <- factor(quants$pars, levels=levels(quants$pars)[c(1:6,9,11,7,8,10)])
  }
  gg<-ggplot(quants, aes(x=percentile, y=value, color=pars))+
    geom_line()+
    #ggtitle(paste(names(quants_list)[i]," Analysis:\ncalled DMR width percentiles",sep = ""))+
    ylab("-log10(Width)")+
    theme(text=element_text(size=14))+
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  if(grepl("Sex",names(quants_list)[i])){
    gg <- gg + facet_grid(rows = vars(chrs),cols = vars(method) )
  } else{
    gg <- gg + facet_grid(cols = vars(method) )
  }
  plot(gg)
  
  
  filename <- paste(figures_dir, names(quants_list)[i],"_width_percentile.png",sep = "") 
  ggsave(
    filename=filename,
    plot = gg,
    device = "png",
    path = NULL,
    scale = 1,
    width = 8,
    height = ifelse(grepl("Sex",filename), 5,3),
    units = c("in"),
    dpi = 200
  )

}


for(i in 1:length(quants_list)){
  filename <-  paste(tables_dir,"quantile_table_", names(quants_list)[i], ".csv", sep="")
  colnames(quants_list[[i]])[which(colnames(quants_list[[i]])=="value")] <- "log10_DMR_width"
  write.csv(file=filename, quants_list[[i]])
}

```

```{r}

gtf_df <- rtracklayer::import("../../STABLE_DATA/HUMAN_GENOME/Homo_sapiens.GRCh37.87.chr.gtf.gz")
gtf_df <- as.data.frame(gtf_df)
gtf_df$seqnames<-paste("chr", gtf_df$seqnames, sep = "")

genes<-as.data.frame(cbind(chromosome=gtf_df$seqnames, start=gtf_df$start, end=gtf_df$end ))
genes<-cbind(genes, width=as.numeric(as.character(gtf_df$width)), strand=gtf_df$strand, feature=gtf_df$type, gene=gtf_df$gene_id, exon=gtf_df$exon_id, transcript=gtf_df$transcript_id, symbol=gtf_df$gene_name)

genes$start<-as.numeric(as.character(genes$start))
genes$end<-as.numeric(as.character(genes$end))

```

```{r, fig.height=9}
data_set_regions_df <- data.frame("data_set"=c("Sex", "Sex", "KAT6A", "KAT6A", "Weaver","Sotos" ),
                                  "chr"=c("chrX","chr9" ,"chr17" ,"chr2","chr7","chr5"),
                                  "start"=c(71.45e6, 84.302e6 , 46.6e6 ,82e6 ,27.11e6 ,140.06e6 ),
                                  "stop"=c(71.55e6, 84.305e6 , 46.72e6, 85e6 , 27.25e6 , 141.00e6 ),
                                  "fig_height"=c(8,8,8,8,8,9.2),
                                  "fig_width"=c(6,4,8,8,8,8))
# 
data_set_regions_df <- data.frame("data_set"=c("KAT6A", "Weaver", "Sotos", "KAT6A", "Weaver","Sotos","KAT6A", "Weaver", "Sotos" ),
                                  "chr"=c("chr5","chr5" ,"chr5" ,"chr11","chr11","chr11","chr11","chr11","chr11"),
                                  "start"=c(140.65e6, 140.65e6, 140.65e6, 2.15e6, 2.15e6, 2.15e6, 2.180e6, 2.180e6, 2.180e6 ),
                                  "stop"=c(140.95e6, 140.95e6, 140.95e6, 2.24e6, 2.24e6, 2.24e6, 2.185e6, 2.185e6, 2.185e6 ),
                                  "fig_height"=c(8),
                                  "fig_width"=c(8))

for(region_id in 1:nrow(data_set_regions_df)){
  data_set <- data_set_list[[data_set_regions_df$data_set[region_id] ]] 
  g1 <- data_set$g1
  g2 <- data_set$g2
  
  group<-character(length = ncol(data_set$B[,c(g1,g2)]))
  group[which(is.element(colnames(data_set$B[,c(g1,g2)]),g1))] <- data_set$g1g2_labels$g1
  group[which(is.element(colnames(data_set$B[,c(g1,g2)]),g2))] <- data_set$g1g2_labels$g2

  group <- as.factor(group)
  color=c("blue", "seagreen3")
  linewide=c(2, 2 )
  
  chr <- data_set_regions_df$chr[region_id]
  start <- data_set_regions_df$start[region_id]
  stop <- data_set_regions_df$stop[region_id]
  
  
  genes_exons<-genes[which((genes$chr==chr) & (genes$start>=start) & (genes$end<=stop)),]
  genes_exons<-droplevels(genes_exons)
  genes_exons<-genes_exons[complete.cases(genes_exons),]
  which_gene_symbols <- c(grep("RPS4",genes_exons$symbol),
                          grep("PIN4",genes_exons$symbol),
                          grep("CITED",genes_exons$symbol),
                          grep("HDAC",genes_exons$symbol),
                          grep("ERCC",genes_exons$symbol),
                          grep("TLE1",genes_exons$symbol),
                          grep("HOX",genes_exons$symbol),
                          grep("PCD",genes_exons$symbol),
                          grep("EVX",genes_exons$symbol),
                          grep("RNU6",genes_exons$symbol),
                          grep("Y_RNA",genes_exons$symbol),
                          grep("MTND4P",genes_exons$symbol),
                          grep("FUNDC",genes_exons$symbol),
                          grep("ST6GALN",genes_exons$symbol),
                          grep("RP11-145",genes_exons$symbol),
                          grep("INS",genes_exons$symbol),
                          grep("IGF",genes_exons$symbol),
                          grep("TH",genes_exons$symbol)
                          )
  genes_exons <- genes_exons[which_gene_symbols,]
  
  gen<-"hg19"
  
  which<-which(data_set$locs$chr==chr & data_set$locs$pos>=start & data_set$locs$pos<=stop )
  B_sub<- data_set$B[which,c(g1,g2)]
  B_norm<-B_sub - rowMeans(B_sub[,g1])
  locs <- data_set$locs[which,]
  
  delta_means <- rowMeans(B_sub[,g1])-rowMeans(B_sub[,g2])
  
  
  irange<-IRanges(start=locs$pos, width=1)
  grange<-GRanges(seqnames=chr, ranges =  ranges(irange), mcols = B_sub )
  
  manhattan_range <- GRanges(seqnames=chr, ranges =  ranges(irange) )
  manhattan_track <- DataTrack(range = manhattan_range, data = -log10(locs$pval),
                               lwd=1.5, col=c("black"), type=c("p","g"), ylim=c(0,max(-log10(locs$pval))))

 
  raw_beta_track<-DataTrack(range=grange, data=t(B_sub), groups=group ,
                            genome=gen , name="Beta", col=color, type=c("a","g", "confint"),
                            lwd=linewide, fontcolor.title="grey20", col.axis="black",
                            fontcolor.legend="grey20", legend = TRUE, lwd.grid=0.8,
                            col.grid="grey65",v=0 )
  displayPars(raw_beta_track) <- list( cex.legend=1, cex.axis=1, cex.title=1 )

  norm_beta_track<-DataTrack(range=grange, data=delta_means,  genome=gen,
                             group ="delta", name="Delta Beta", col="black",
                             type=c("a", "g"), fontcolor.title="grey20",col.axis="black",
                             fontcolor.legend="grey20",legend = TRUE, baseline=0,
                             col.baseline="black", lty.baseline=3)
  displayPars(norm_beta_track) <- list( cex.legend=1, cex.axis=1, cex.title=1 , lwd.grid=1, col.grid="grey80",v=0,h=0)
  
  
  grtrack <- GeneRegionTrack(genes_exons, genome=gen, chromosome=chr,
                             name="Gene Model", transcriptAnnotation="symbol",
                             collapseTranscripts = "meta")
  displayPars(grtrack) <- list( background.panel="white", col="grey20", fontsize=25, 
                                fontsize.group=25, fill = "grey50", lwd=1, 
                                fontcolor.group="grey20", fontcolor.title="grey20",
                                stackHeight = 0.95, shape="box")
  if( length(which)==0 ){
    grtrack <- itrack
  }
  
  dmr_tracks <- list()
  
  data_set_id <- grep(data_set$data_set_name, names(data_set_list))
  result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
  for(rs_index in result_set_ids){
    result_set <- data_set_results[[rs_index]] 
    if(grepl("dmrscaler", result_set$method)){
      if(grepl("_2", result_set$method_id)){
        for(l in 1:length(unique(result_set$grange$layer))){
          which_in_layer <- which(result_set$grange$layer==unique(result_set$grange$layer)[l]
                                  & seqnames(result_set$grange)==chr )
          if(length(which_in_layer)==0){
            dmr_tracks[[length(dmr_tracks)+1]] <- AnnotationTrack()
          } else {
            dmr_tracks[[length(dmr_tracks)+1]] <- AnnotationTrack(start = start(result_set$grange[which_in_layer]) , 
                                                                  end =  end(result_set$grange[which_in_layer]),
                                                                  chromosome = chr,
                                                                  strand = "*", genome = "hg19",
                                                                  name=unique(result_set$grange$layer)[l])
          } 
          
          displayPars(dmr_tracks[[length(dmr_tracks)]]) <- list(cex.legend=0.7, cex.axis=0.7, cex.title=0.7, rotation.title=0, col="grey20", fill="grey70", fontcolor.title="grey20", stackHeight = 1, shape="box")
        }
      }
    } else {
      if(grepl("combp_2", result_set$method_id)){next}
      which_in_anno <- which(seqnames(result_set$grange)==chr)
      dmr_tracks[[length(dmr_tracks)+1]] <- AnnotationTrack(start = start(result_set$grange[which_in_anno]) , 
                                                              end =  end(result_set$grange[which_in_anno]),
                                                              chromosome = chr,
                                                              strand = "*", genome = "hg19",
                                                              name=result_set$method_id)
      displayPars(dmr_tracks[[length(dmr_tracks)]]) <- list(cex.legend=0.7, cex.axis=0.7, cex.title=0.7, rotation.title=0, col="grey20", fill="grey70", fontcolor.title="grey20", stackHeight = 1, shape="box")
      
    }
  }
  
  gtrack<-GenomeAxisTrack()
  displayPars(gtrack) <- list(cex=1, cex.id=1, fontcolor="grey20", col="grey20")
  itrack<-IdeogramTrack(genome=gen,chromosome = chr)
  displayPars(itrack) <- list(cex=1,cex.bands=1, fontcolor="grey20")
  track_list <- c(list(itrack, gtrack, manhattan_track, norm_beta_track, grtrack),dmr_tracks )
  #track_list <- c(list(itrack, gtrack, raw_beta_track, norm_beta_track, grtrack),dmr_tracks )
  
  filename <- paste(figures_dir, data_set$data_set_name, "_",chr,"_",start,"_",stop, ".png",sep="")
  png(file=filename, width = data_set_regions_df$fig_width[region_id], height = data_set_regions_df$fig_height[region_id], units = "in", res = 320)
  plotTracks(c(track_list),
             from = start-1, to=stop+1, # sizes =c(0.5,0.75,3,2,3,0.35,0.35,0.35,0.35,0.35),
             background.title = "white"
            )
  
  dev.off()
  
  
  
  
  gglist <- list()
  gglist[["signif_adj_plot"]] <- ggplot(locs, aes(x=1:length(pval), y=-log10(pval) )) +
    geom_segment(aes(xend=1:length(-log10(pval)), y=0, yend=-log10(pval), alpha=0.5)) +
    geom_point(aes(color= -log10(pval))) + 
    scale_color_gradient2(low="grey60", mid="grey30", high = "red", midpoint = -log10(0.05) )+
    xlab("CG index over region") + ylab("-log10(pval)")+
    ggtitle(paste("Adjacent_CG_plot_", chr,"_" ,start, "_", stop))+
    ylab("-log10pval") +
    scale_x_continuous(expand = c(0, 0))+
    theme_bw()+
    theme(
      plot.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.text.x = element_blank(), axis.title.x = element_blank(),
      text = element_text(size=12, color = "black"),
      axis.text = element_text(size=12, color = "black"),
      legend.position = "none"
    )

  #plot(gglist[["signif_adj_plot"]])  

  df_betas <- cbind("index" = 1:nrow(locs), locs, "mean"= rowMeans(B_sub[,g1]), "group"=data_set$g1g2_labels$g1 )
  df_betas <- rbind(df_betas, cbind("index" = 1:nrow(locs), locs, "mean" = rowMeans(B_sub[,g2]), "group"=data_set$g1g2_labels$g2))
                    
  gglist[["beta_raw_adj_plot"]] <- ggplot(df_betas)+
    geom_line(aes(x=index, y=mean, group=names, alpha=0.5))+
    geom_point(aes(x=index, y=mean, color=group))+
    scale_color_manual(values=color)+
    scale_x_continuous(expand = c(0, 0))+
  ylab("Beta") +
  theme_bw()+
  theme(
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_blank(), axis.title.x = element_blank(),
    text = element_text(size=12, color = "black"),
    axis.text = element_text(size=12, color = "black"),
    legend.position = "none"
  )
  
  #plot(gglist[["beta_raw_adj_plot"]])
    
  df_betas <- cbind("index" = 1:nrow(locs), locs, "mean"= rowMeans(B_norm[,g1]), "group"=data_set$g1g2_labels$g1 )
  df_betas <- rbind(df_betas, cbind("index" = 1:nrow(locs), locs, "mean" = rowMeans(B_norm[, g2]), "group"=data_set$g1g2_labels$g2))
  
  gglist[["beta_norm_adj_plot"]] <- ggplot(df_betas)+
    geom_line(aes(x=index, y=mean, group=names, alpha=0.5))+
    geom_point(aes(x=index, y=mean, color=group))+
    scale_color_manual(values=color)+
    scale_x_continuous(expand = c(0, 0))+
  ylab("Normalized Beta") +
  theme_bw()+
  theme(
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_blank(), axis.title.x = element_blank(),
    text = element_text(size=12, color = "black"),
    axis.text = element_text(size=12, color = "black"),
    legend.position = "none"
  )
  
  #plot(gglist[["beta_norm_adj_plot"]]) 
  
  
  
  ## is CG j in a DMR a level i
  in_dmr_list <- list()
  locs_gr <- GRanges(seqnames=locs$chr, IRanges(start=locs$pos, width=1))
  for(rs_index in result_set_ids){
    result_set <- data_set_results[[rs_index]]
    if(grepl("dmrscaler", result_set$method)){
      if(grepl("_2", result_set$method_id)){
        in_dmr_list[[result_set$method]]$locs <- locs
        in_dmr_list[[result_set$method]]$locs$index <- 1:nrow(locs)
        in_dmr_list[[result_set$method]]$locs$in_dmr <- 0 
        for(l in 1:length(unique(result_set$grange$layer))){
          which_in_layer <- which(result_set$grange$layer==unique(result_set$grange$layer)[l])
          in_dmr_list[[result_set$method]]$locs$in_dmr <- in_dmr_list[[result_set$method]]$locs$in_dmr + 
            countOverlaps(locs_gr, result_set$grange[which_in_layer])
        }
        in_dmr_list[[result_set$method]]$layer_names <- unique(result_set$grange$layer)
      }
    } else {
      if(grepl("combp_2",result_set$method_id)){next}
      in_dmr_list[[result_set$method_id]]$locs <- locs
        in_dmr_list[[result_set$method_id]]$locs$index <- 1:nrow(locs)
      in_dmr_list[[result_set$method_id]]$locs$in_dmr <- countOverlaps(locs_gr, result_set$grange)
    }
  }
  
  
  for(in_dmr_plot_id in 1:length(in_dmr_list)){
    plot_name <- paste("in_dmr_plot_", names(in_dmr_list)[in_dmr_plot_id], sep="") 
    gglist[[plot_name]] <- ggplot(in_dmr_list[[in_dmr_plot_id]]$locs)+
      geom_bar(aes(x=index, y=in_dmr, color=in_dmr), fill="grey70",  stat="identity")+
      scale_color_gradient(low="grey50", high = "grey50")+
      xlab("CG index over region") +
      theme_bw()+
      scale_x_continuous(expand = c(0, 0))+
      theme(
        plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        text = element_text(size=12, color = "black"),
        axis.text = element_text(size=12, color = "black"),
        legend.position = "none"
      )
    if(grepl("dmrscaler", names(in_dmr_list)[in_dmr_plot_id]) ){
      gglist[[plot_name]] <- gglist[[plot_name]] + 
        scale_y_continuous(breaks = seq(0,length(in_dmr_list$dmrscaler$layer_names),1) ,
                           labels= c("none", rev(in_dmr_list$dmrscaler$layer_names)))+
      ylab("Layer") 
    } else {
      gglist[[plot_name]] <- gglist[[plot_name]] + 
        scale_y_continuous(breaks = seq(0,1),
                           labels= c("0", "1"))+
        ylab(gsub("in_dmr_plot_","", plot_name))+
        theme(
          axis.title.y = element_text(angle = 0, hjust=1),
          axis.title.x=element_blank(),
          axis.text.x=element_blank(),
          axis.ticks.x=element_blank()
        )
    }
  }
  gtable <- ggplot_gtable(ggplot_build(gglist[[1]]))
  for(i in 2:length(gglist)){
    gtable <- rbind(gtable, ggplot_gtable(ggplot_build(gglist[[i]])) )
  }
  panels <- gtable$layout$t[grepl("panel", gtable$layout$name)]
  gtable$heights[panels] <- unit(c(rep(4,4), rep(1,6)),"null")
  
  filename <- paste(figures_dir, data_set$data_set_name, "_adj_cgs_",chr,"_",start,"_",stop, ".png",sep="")
  png(file=filename, width = data_set_regions_df$fig_width[region_id]+1, height = data_set_regions_df$fig_height[region_id], units = "in", res = 320)
  grid.newpage()
  grid.draw(gtable)
  dev.off()
  # filename=paste(figures_dir"Adjacent_CG_plots_", chr,"_" ,start, "_", stop, ".png", sep = "")  
  # png(filename = filename, width = 2.75, height = 4, units = "in", res = 320)
  

  # dev.off()
  
  
  
  
    
}


```

```{r}

```

```{r  make rare disease analysis results tables}

library("gt")

results_table_list <- list()

for(ds_id in grep("Sex", names(data_set_list), invert = T)){
  results_table <- data.frame("method"=character(),
                              "num_dmrs"=numeric(), 
                              "mean_width"=numeric(),
                              "median_width"=numeric(),
                              "percent_total_width"=numeric())
  result_set_ids <- grep(paste("data_set_", ds_id, sep = ""), names(data_set_results))
  for(rs_index in result_set_ids){
    result_set <- data_set_results[[rs_index]] 
    if(grepl("dmrscaler", result_set$method)){
      if(grepl("_2", result_set$method_id)){
        for(l in 1:length(unique(result_set$grange$layer))){
          which <- which(result_set$grange$layer == unique(result_set$grange$layer)[l])
          results_table <- rbind(results_table,
                                 data.frame("method" = result_set$method,
                                            "num_dmrs" = length(result_set$grange[which]),
                                            "mean_width" = mean(width(result_set$grange[which])),
                                            "median_width" = median(width(result_set$grange[which])), 
                                            "percent_total_width" = sum(width(result_set$grange[which])) / 
                                              sum(seqlengths(result_set$grange))
                           ))
          rownames(results_table)[nrow(results_table)] <- unique(result_set$grange$layer)[l] 
        }
      }
    }else{
      if(grepl("combp_2", result_set$method_id)){next}
      results_table <- rbind(results_table,
                             data.frame("method" = result_set$method,
                                        "num_dmrs" = length(result_set$grange),
                                        "mean_width" = mean(width(result_set$grange)),
                                        "median_width" = median(width(result_set$grange)), 
                                        "percent_total_width" = sum(width(result_set$grange)) / 
                                          sum(seqlengths(result_set$grange))
                       ))
      rownames(results_table)[nrow(results_table)] <- result_set$method_id
    }
  
  }
  
  results_table$method_params <- rownames(results_table)
  
  clean_results_table <- gt(results_table, groupname_col = "method", rowname_col = "method_params") 
  clean_results_table <- tab_header(clean_results_table, title = paste(names(data_set_list)[ds_id],"Analysis DMR Summary Table"))
  clean_results_table <- tab_stubhead(clean_results_table , "method")
  
  clean_results_table <- cols_label(clean_results_table,
                                    num_dmrs = html(" #\n<br>DMRs "),
                                    mean_width = html(" mean\n<br>DMR width "),
                                    median_width = html(" median\n<br>DMR width "),
                                    percent_total_width = html(" % total\n<br>width "))
  
  
  clean_results_table <- fmt_number(clean_results_table, columns = c(percent_total_width),
                                    n_sigfig = 2, scale_by = 100, pattern = "{x}%")
  clean_results_table <- fmt_number(clean_results_table, columns = c(mean_width),
                                    rows = (mean_width >= 1e6),
                                    decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
  clean_results_table <- fmt_number(clean_results_table, columns = c(mean_width),
                                    rows = (mean_width > 1000) & (mean_width < 1e6),
                                    decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
  clean_results_table <- fmt_number(clean_results_table, columns = c(mean_width),
                                    rows = (mean_width <= 1000),
                                    decimals = 0, pattern = "{x} bp")
  
  clean_results_table <- fmt_number(clean_results_table, columns = c(median_width),
                                    rows = (median_width >= 1e6),
                                    decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
  clean_results_table <- fmt_number(clean_results_table, columns = c(median_width),
                                    rows = (median_width > 1000) & (median_width < 1e6),
                                    decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
  clean_results_table <- fmt_number(clean_results_table, columns = c(median_width),
                                    rows = (median_width <= 1000),
                                    decimals = 0, pattern = "{x} bp")
  
  results_table_list[[names(data_set_list)[ds_id]]] <- clean_results_table
  
}


for(i in 1:length(results_table_list)){
  gtsave(results_table_list[[i]], filename = paste(tables_dir, names(results_table_list)[i], "_result_table.png",sep=""))
}

clean_results_table

```


```{r make XY sex analysis results table}

library("gt")

results_table <- data.frame("method"=character(),
                            "chrX_num_dmrs"=numeric(), 
                            "chrX_mean_width"=numeric(),
                            "chrX_median_width"=numeric(),
                            "chrX_percent_total_width"=numeric(),
                            "autosome_num_dmrs"=numeric(), 
                            "autosome_avg_width"=numeric(),
                            "autosome_median_width"=numeric(),                            
                            "autosome_percent_total_width"=numeric())


data_set_id <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
for(rs_index in result_set_ids){
  result_set <- data_set_results[[rs_index]] 
  if(grepl("dmrscaler", result_set$method)){
    if(grepl("_2", result_set$method_id)){
      for(l in 1:length(unique(result_set$grange$layer))){
        which_x <- which(result_set$grange@seqnames=="chrX" &
                           result_set$grange$layer == unique(result_set$grange$layer)[l])
        which_not_x <- which(result_set$grange@seqnames != "chrX" & result_set$grange@seqnames != "chrY" &
                           result_set$grange$layer == unique(result_set$grange$layer)[l])
        results_table <- rbind(results_table,
                               data.frame("method" = result_set$method,
                                          "chrX_num_dmrs" = length(which_x),
                                          "chrX_mean_width" = mean(width(result_set$grange[which_x,])),
                                          "chrX_median_width" = median(width(result_set$grange[which_x,])), 
                                          "chrX_percent_total_width" = sum(width(result_set$grange[which_x,])) / 
                                            seqlengths(result_set$grange)["chrX"],
                                          "autosome_num_dmrs" = length(which_not_x),
                                          "autosome_mean_width" = mean(width(result_set$grange[which_not_x,])),
                                          "autosome_median_width"=median(width(result_set$grange[which_not_x,])),
                                          "autosome_percent_total_width" = sum(width(result_set$grange[which_not_x,])) / 
                                            sum(seqlengths(result_set$grange)[!grepl("[XY]",names(seqlengths(result_set$grange)))])
                         ))
        rownames(results_table)[nrow(results_table)] <- unique(result_set$grange$layer)[l] 
      }
    }
  }else{
    if(grepl("combp_2", result_set$method_id)){next}
    which_x <- which(result_set$grange@seqnames=="chrX")
    which_not_x <- which(result_set$grange@seqnames != "chrX" & result_set$grange@seqnames != "chrY")
    results_table <- rbind(results_table,
                           data.frame("method" = result_set$method,
                                      "chrX_num_dmrs" = length(which_x),
                                      "chrX_mean_width" = mean(width(result_set$grange[which_x,])),
                                      "chrX_median_width" = median(width(result_set$grange[which_x,])), 
                                      "chrX_percent_total_width" = sum(width(result_set$grange[which_x,])) / 
                                        seqlengths(result_set$grange)["chrX"],
                                      "autosome_num_dmrs" = length(which_not_x),
                                      "autosome_mean_width" = mean(width(result_set$grange[which_not_x,])),
                                      "autosome_median_width"=median(width(result_set$grange[which_not_x,])),
                                      "autosome_percent_total_width" = sum(width(result_set$grange[which_not_x,])) / 
                                        sum(seqlengths(result_set$grange)[!grepl("[XY]",names(seqlengths(result_set$grange)))])
                     ))
    rownames(results_table)[nrow(results_table)] <- result_set$method_id
  }

}


results_table$method_params <- rownames(results_table)

clean_results_table <- gt(results_table, groupname_col = "method", rowname_col = "method_params") 
clean_results_table <- tab_header(clean_results_table, title = "Sex Analysis DMR Summary Table")
clean_results_table <- tab_spanner(clean_results_table, label="chrX",
                                   columns = c("chrX_num_dmrs","chrX_mean_width","chrX_median_width", "chrX_percent_total_width") )
clean_results_table <- tab_spanner(clean_results_table, label="Autosomes",
                                   columns = c("autosome_num_dmrs","autosome_mean_width","autosome_median_width", "autosome_percent_total_width") )
clean_results_table <- tab_stubhead(clean_results_table , "method")

clean_results_table <- cols_label(clean_results_table,
                                  chrX_num_dmrs = html(" #\n<br>DMRs "), chrX_mean_width = html(" mean\n<br>DMR width "),
                                  chrX_median_width = html(" median\n<br>DMR width "), chrX_percent_total_width = html(" % total\n<br>width "),
                                  autosome_num_dmrs = html(" #\n<br>DMRs "), autosome_mean_width = html(" mean\n<br>DMR width "),
                                  autosome_median_width = html(" median\n<br>DMR width "), autosome_percent_total_width = html(" % total\n<br>width " ))


clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_percent_total_width, autosome_percent_total_width),
                                  n_sigfig = 2, scale_by = 100, pattern = "{x}%")

clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_mean_width),
                                  rows = (autosome_mean_width > 1000) & (autosome_mean_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_mean_width),
                                  rows = (autosome_mean_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")

clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_mean_width),
                                  rows = (chrX_mean_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_mean_width),
                                  rows = (chrX_mean_width > 1000) & (chrX_mean_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_mean_width),
                                  rows = (chrX_mean_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")

clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_median_width),
                                  rows = (chrX_median_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_median_width),
                                  rows = (chrX_median_width > 1000) & (chrX_median_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(chrX_median_width),
                                  rows = (chrX_median_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")

clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_median_width),
                                  rows = (autosome_median_width >= 1e6),
                                  decimals = 2, scale_by = 1/1e6, pattern = "{x} Mb")
clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_median_width),
                                  rows = (autosome_median_width > 1000) & (autosome_median_width < 1e6),
                                  decimals = 2, scale_by = 1/1000, pattern = "{x} kb")
clean_results_table <- fmt_number(clean_results_table, columns = c(autosome_median_width),
                                  rows = (autosome_median_width <= 1000),
                                  decimals = 0, pattern = "{x} bp")


clean_results_table

#gtsave(clean_results_table, filename = paste( "Sex_Analysis_DMR_Summary_Table_2.png",sep=""))
gtsave(clean_results_table, filename = paste(tables_dir,"Sex_Analysis_DMR_Summary_Table_2.html",sep=""))

```

```{r make XY analysis circos network plot}

library(networkD3)
library(rjson)

data_set_id <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
method_set_id <- grep("dmrscaler_2", names(method_set_list))
method_set_ids <- grep(paste("method_set_", method_set_id, sep = ""), names(data_set_results))
result_set <- data_set_results[[ intersect(result_set_ids, method_set_ids ) ]] 



full_node_set <- list()
for(j in 1:length(unique(result_set$grange$layer))){
  #which <- which(built_layers[[j]]$chr =="chrX")
  #which <- 1:length(built_layers[[j]]$chr)
  gr_nodes <- list()
  which <- which(result_set$grange$layer == unique(result_set$grange$layer)[j]) #&
  #                 -log10(result_set$grange$pval_region) > 10) 
  for(i in 1:length(which)){
    #temp_name <- paste(as.character(seqnames(result_set$grange[ which[i],])), start(result_set$grange[which[i],]), end(result_set$grange[which[i],]), sep = "_")
    
    
    temp_name <- paste(format(start(result_set$grange[which[i],]),big.mark=","), format(end(result_set$grange[which[i],]),big.mark=","), sep = " - ")
    temp_child <- list("name"= temp_name, "alt_name"= temp_name, "grange"=result_set$grange[which[i],], "children"=list() )
    gr_nodes[[i]] <-  list("name"= temp_name, "alt_name"= temp_name, "grange"=result_set$grange[which[i],], "children"=list())
  }
  full_node_set[[j]]<-gr_nodes
}

temp_full_node_set <- full_node_set
#full_node_set <- temp_full_node_set

for(j in 2:length(full_node_set)){
  print(j)
  
  lower <- full_node_set[[j-1]]
  upper <- full_node_set[[j]]
  
  lower_gr <- lower[[1]]$grange
  for(i in 2:length(lower)){
    lower_gr <- c(lower_gr, lower[[i]]$grange)
  }
  upper_gr <- upper[[1]]$grange
  if(length(upper) >= 2){
    for(i in 2:length(upper)){
      upper_gr <- c(upper_gr, upper[[i]]$grange)
    }
  }
  
  hits <- GenomicRanges::findOverlaps(lower_gr,upper_gr,type = "within");hits
  for( i in 1:length(hits@from)){
    upper[[ hits@to[i] ]]$children <- c(upper[[ hits@to[i] ]]$children, list(lower[[ hits@from[i] ]]))
    if(upper[[ hits@to[i] ]]$name == lower[[ hits@from[i] ]]$name |
       upper[[ hits@to[i] ]]$alt_name == lower[[ hits@from[i] ]]$name  |
       upper[[ hits@to[i] ]]$name == lower[[ hits@from[i] ]]$alt_name
       ){  ### avoid redundant child-parent pairs
      upper[[ hits@to[i] ]]$name <- "..."
    }
  }
  full_node_set[[j]]<-upper
}

#dmr_tree <- full_node_set[[3]][[1]]
#dmr_tree <- list("name"="root","children"=full_node_set[[3]])
dmr_tree <- list("name"="root","children"=full_node_set[[length(full_node_set)]])

dmr_tree_clean <- list()

node <- dmr_tree



strip_grs <- function(node){
  node <- list("name" = node$name, "children" = node$children)
  if(length(node$children)==0){return(node)}
  for(i in 1:length(node$children)){
    node$children[[i]]<-strip_grs(node$children[[i]])
  }
  return(node)
}
dmr_tree_clean <- strip_grs(dmr_tree)


# library("webshot") ###  library to fix an obnoxious problem
# 
# temp <- radialNetwork(List = dmr_tree_clean, fontSize = 16, fontFamily = "bold", nodeStroke = "black", linkColour = "grey60", opacity = 1, margin=0, height = 1600, width = 1600)
# saveNetwork(temp,paste(figures_dir,"All_chrom_dmrs_radialNetwork2.html",sep = "") ,
#             selfcontained = T)
# webshot(paste(figures_dir,"All_chrom_dmrs_radialNetwork2.html",sep = ""),
#         paste(figures_dir,"All_all_layer_chrom_dmrs_radialNetwork2.png",sep = ""),
#         zoom = 2)
# 
# 

temp_tree <- dmr_tree_clean
temp_tree$children[c(1:22,24)] <- NULL
temp_tree$children[[1]]$children[[2]]$children[c(1:6,8:17)] <- NULL
#temp_tree$children[[1]]$children[[2]]$children[[1]]$children[c(seq(2,7,by=2),7)] <- NULL

temp_tree <- temp_tree$children[[1]]
temp <- diagonalNetwork(List = temp_tree, fontSize = 12, fontFamily = "bold", nodeStroke = "black", linkColour = "black", opacity = 1,  margin=0, height = 380, width = 280)
saveNetwork(temp,paste(figures_dir, "FINAL_X2_chrom_dmrs_diagonalNetwork.html", sep="") , selfcontained = T )
webshot(paste(figures_dir,"FINAL_X2_chrom_dmrs_diagonalNetwork.html", sep=""),
        paste(figures_dir,"FINAL_X2_chrom_dmrs_diagonalNetwork.png", sep=""),
        vheight = 380, vwidth = 280 , zoom =10 )


```



```{r make XY analysis DMR called width histograms}
library("gt")


data_set_id <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", data_set_id, sep = ""), names(data_set_results))
for(rs_index in result_set_ids){
  result_set <- data_set_results[[rs_index]] 
  if(grepl("dmrscaler", result_set$method)){
    if(grepl("_2", result_set$method_id)){
      for(l in 1:length(unique(result_set$grange$layer))){
        which_x <- which(result_set$grange@seqnames=="chrX" &
                           result_set$grange$layer == unique(result_set$grange$layer)[l])
        which_not_x <- which(result_set$grange@seqnames != "chrX" & result_set$grange@seqnames != "chrY" &
                           result_set$grange$layer == unique(result_set$grange$layer)[l])

        gglist <- list()
        gglist[["X_chromosome"]] <- ggplot(data.frame(widths=width(result_set$grange[which_x])), aes(x=log10(widths)) )+
          ggtitle(label = "X chromosome DMR widths",
                  subtitle = paste(unique(result_set$grange$layer)[l], sep=""))+
          geom_histogram(bins=50)+
          xlim(0,8.5)+
          theme(text=element_text(size=18))
        gglist[["Autosomes"]] <- ggplot(data.frame(widths=width(result_set$grange[which_not_x]) ), aes(x=log10(widths)) )+
          ggtitle(label = "Autosome DMR widths",
                  subtitle = paste(unique(result_set$grange$layer)[l], sep=""))+
          geom_histogram(bins=50)+
          xlim(0,8.5)+
          theme(text=element_text(size=18))
        filename=paste(figures_dir, result_set$data_set ,"_DMR_widths__",unique(result_set$grange$layer)[l], ".png", sep = "")
        png(filename = filename, width = 10, height = 6, units = "in", res = 150)
        grid.newpage()
        grid.arrange( gglist[["X_chromosome"]],  gglist[["Autosomes"]], nrow=1)
        dev.off()
      }
    }
  }else{
    gglist <- list()
    which_x <- which(result_set$grange@seqnames=="chrX")
    which_not_x <- which(result_set$grange@seqnames != "chrX" & result_set$grange@seqnames != "chrY")
    gglist <- list()
    gglist[["X_chromosome"]] <- ggplot(data.frame(widths=width(result_set$grange[which_x]) ), aes(x=log10(widths)) )+
      ggtitle(label = "X chromosome DMR widths",
              subtitle = paste(result_set$method_id, sep=""))+
      geom_histogram(bins=50)+
      xlim(0,8.5)+
      theme(text=element_text(size=18))
    gglist[["Autosomes"]] <- ggplot(data.frame(widths=width(result_set$grange[which_not_x]) ), aes(x=log10(widths)) )+
      ggtitle(label = "Autosome DMR widths",
              subtitle = paste(result_set$method_id, sep=""))+
      geom_histogram(bins=50)+
      xlim(0,8.5)+
      theme(text=element_text(size=18))
    filename=paste(figures_dir, result_set$data_set ,"_DMR_widths__",result_set$method_id, ".png", sep = "")
    png(filename = filename, width = 10, height = 6, units = "in", res = 150)
    grid.newpage()
    grid.arrange( gglist[["X_chromosome"]],  gglist[["Autosomes"]], nrow=1)
    dev.off()
  }
}

```








```{r funcs for kws overlap analysis}

frac_dbeta_same_dir <- function(gr12, B1, atom1, case1, B2, case2, atom2, sigcut){
    num_sig <- numeric(length = length(gr12))
    frac_shared_sig <- numeric(length = length(gr12))
    frac_sig_same_dir <- numeric(length = length(gr12))
    for(i in 1:length(gr12)){
      #print(i)
      chr <- as.character(gr12@seqnames)[i]
      start <- start(gr12)[i]
      end <- end(gr12)[i]
      control1 <- (1:ncol(B1))[-case1]
      control2 <- (1:ncol(B2))[-case2]
      
      cgs <- which(atom1$chr == chr & atom1$pos <= end & atom1$pos >= start)
      
      ##hyper (caseBeta > controlBeta)
      if(length(cgs)==1){ 
        hyper1 <- mean(B1[cgs,case1]) > mean(B1[cgs,control1])
        hyper2 <- mean(B2[cgs,case2]) > mean(B2[cgs,control2]);
      }
      else{
        hyper1 <- ( rowMeans(B1[cgs,case1]) > rowMeans(B1[cgs,control1]) ) 
        hyper2 <- ( rowMeans(B2[cgs,case2]) > rowMeans(B2[cgs,control2]) ) 
      }
      sig1 <- (atom1$pval[cgs] < sigcut) ## 1 corresponds to FDR = 0.1 based on scoring func
      sig2 <- (atom2$pval[cgs] < sigcut) ## 1 corresponds to FDR = 0.1 based on scoring func
      num_sig[i] <- sum(sig1 | sig2)
      frac_shared_sig[i] <- sum(sig1 & sig2) / sum(sig1 | sig2)
      frac_sig_same_dir[i] <- sum((hyper1 == hyper2) & sig1 & sig2 )  / sum(sig1 & sig2)
      if(is.na(frac_sig_same_dir[i])){frac_sig_same_dir[i] = 0}
    }
    return( data.frame(num_sig = num_sig, frac_shared_sig = frac_shared_sig, frac_sig_same_dir = frac_sig_same_dir ))
}

frac_A_in_B <- function(gr1,gr2){
  which <- findOverlaps(gr1,gr2)@from
  frac_a_in_b <- rep(0,length(gr1))
  
  temp_int_gr <- intersect(gr1, gr2)
  temp_ovs <- findOverlaps(gr1, temp_int_gr)
  
  froms <- temp_ovs@from
  tos <- temp_ovs@to
  
  gr1 <- as.data.frame(gr1)
  gr2 <- as.data.frame(gr2)
  temp_int_gr <- as.data.frame(temp_int_gr)
  for(i in 1:length(unique(froms))){
    frac_a_in_b[froms[i]] <- 
      sum(temp_int_gr$width[tos[which(froms==(unique(froms)[i]) )] ]) / 
      sum(gr1$width[froms[which(froms==unique(froms)[i]) ]]  )
      
  }
  
  return(frac_a_in_b)
}


syn_or <- function(cg1, cg2, nt, alpha_=0.05){
  df <- data.frame(or_estimate=1,or_ci="",fisher_p=1, alpha=alpha_)
  cg12 <- length(intersect(cg1,cg2))
  cg_n1_y2 <- length(cg2) - cg12
  cg_n2_y1 <- length(cg1) - cg12
  cg_n12 <- nt - cg12 - cg_n1_y2 - cg_n2_y1
  df$or_estimate <- round((cg12 * cg_n12) / (cg_n1_y2 * cg_n2_y1),2)
  
  log_or <- log((cg12 * cg_n12) / (cg_n1_y2 * cg_n2_y1))
  se_log_oe <-  sqrt(1/cg12 + 1/cg_n1_y2 + 1/cg_n2_y1 + 1/cg_n12)
  left <-  log_or - qnorm(1-alpha_/2)*se_log_oe
  right <-  log_or + qnorm(1-alpha_/2)*se_log_oe
  df$or_ci <- paste(round(exp(left),2), "-",round(exp(right),2),sep = "")
  
  df$fisher_p <- fisher.test(matrix(c(cg12,cg_n1_y2, cg_n2_y1, cg_n12),ncol=2))$p.value

  return(df)
}



```

```{r kat6a-weaver-sotos overlap analysis}
genes <- gtf_df[which(gtf_df$gene_biotype =="protein_coding" ),]
genes <- genes[which(genes$gene_source=="ensembl_havana"),]
genes <- genes[which(genes$type=="gene" ),]
genes <- genes[,!(colSums(!is.na(genes))==0)]
gene_ranges <- GRanges(seqnames = genes$seqnames, IRanges(start = genes$start, width = genes$width))
gene_ranges$gene_names <- genes$gene_name

if(file.exists("syndrome_overlap_list.Rdata")){
  load("syndrome_overlap_list.Rdata")

} else {
  geneset_overlaps = data.frame(layer=numeric(),
                                overlap_between = character(),
                                gene = character())
  range_overlaps <- data.frame(layer=numeric(), datasets=character(),
                               seqnames = character(), start = numeric(), width = numeric(),
                               overlapping_genes = character(),
                               gene_overlap_count = numeric(),
                               num_cg_sig = numeric(),
                               frac_shared_sig = numeric(),
                               frac_sig_same_dir = numeric(),
                               frac_shared_kat6a = numeric(),
                               frac_shared_sotos = numeric(),
                               frac_shared_weaver = numeric() )
  or_df_model <- data.frame(layer=numeric(), datasets=character(),
                            num_cg_ds1=numeric(),num_cg_ds2=numeric(),num_cg_ds12=numeric(),
                            or_estimate=numeric(), or_ci=character(),
                            fisher_p=numeric(), alpha=numeric())
  
  syndrome_overlap_list <- list()
  for(i in 1:length(method_set_list)){
    syndrome_overlap_list[[ names(method_set_list)[i] ]] <- list(geneset_overlaps = geneset_overlaps,
                                                                 range_overlaps = range_overlaps,
                                                                 or_table =or_df_model) 
  }
  
  for(dsr_index in 1:length(data_set_results)){
    print(dsr_index)
    temp_gr <- data_set_results[[dsr_index]]$grange
    temp_gr$overlapping_genes <- ""
    temp_gr$gene_overlap_count <- ""
    
    temp <- findOverlaps(temp_gr, gene_ranges)
    froms <- unique(temp@from)
    temp_genes <- gene_ranges$gene_names[temp@to]
    temp_gr <- as.data.frame(temp_gr)  ### NOTE: much faster if convert to df first -> grange ops are slow in loop
    
    for(i in 1:length(froms)){
      temp_gr$overlapping_genes[froms[i]] <- paste(temp_genes[which(temp@from==froms[i])], collapse = ",")
    }
    temp_gr <- GRanges(temp_gr)
    temp_gr$gene_overlap_count <- countOverlaps(temp_gr, gene_ranges)
    data_set_results[[dsr_index]]$grange <- temp_gr
  }
  
  
  
  data_set_ids <- grep("Sex", names(data_set_list), invert = T)
  for(ds_index_1 in data_set_ids){
    print(paste("ds_index_1:",ds_index_1))
    for(ds_index_2 in data_set_ids){
      print(paste("ds_index_2:",ds_index_2))
      result_set_ids <- grep(paste("data_set_", ds_index_1, sep = ""), names(data_set_results))
      for(rs_index_1 in result_set_ids){
        result_set_1 <- data_set_results[[rs_index_1]] 
        method_set <- gsub("data_set.*method_set_", "", names(data_set_results)[rs_index_1])
        method_set <- gsub("_result.csv","",method_set)
        rs_index_2 <- grep(paste("data_set_", ds_index_2,"__method_set_",
                                 method_set, sep = ""), names(data_set_results))
        if(length(rs_index_2)==0){next}
        
        which_dsr <- grep(paste("data_set_", grep("kat6a", names(data_set_list), ignore.case = T), 
                           "__method_set_", method_set, sep = ""), names(data_set_results))
        if(length(which_dsr)==0){gr_k <- GRanges()} else {
          gr_k <- data_set_results[[which_dsr]]$grange  
        }
        which_dsr <- grep(paste("data_set_", grep("sotos", names(data_set_list), ignore.case = T), 
                           "__method_set_", method_set, sep = ""), names(data_set_results))
        if(length(which_dsr)==0){gr_s <- GRanges()} else {
          gr_s <- data_set_results[[which_dsr]]$grange  
        }
        which_dsr <- grep(paste("data_set_", grep("weaver", names(data_set_list), ignore.case = T), 
                           "__method_set_", method_set, sep = ""), names(data_set_results))
        if(length(which_dsr)==0){gr_w <- GRanges()} else {
          gr_w <- data_set_results[[which_dsr]]$grange  
        }
        
        result_set_2 <- data_set_results[[rs_index_2]]
        
        if(grepl("dmrscaler", result_set_1$method)){
          if(grepl("_2", result_set_1$method_id)){
            for(l in 1:length(unique(result_set_1$grange$layer))){
              temp_gr_k <- gr_k[which(gr_k$layer==unique(result_set_1$grange$layer)[l])] 
              temp_gr_s <- gr_s[which(gr_s$layer==unique(result_set_1$grange$layer)[l])] 
              temp_gr_w <- gr_w[which(gr_w$layer==unique(result_set_1$grange$layer)[l])]
              
              
              print(paste("layer:",l))
              temp_geneset_overlaps <- syndrome_overlap_list[[result_set_1$method_id]]$geneset_overlaps
              which_1 <- which(result_set_1$grange$layer == unique(result_set_1$grange$layer)[l])
              which_2 <- which(result_set_2$grange$layer == unique(result_set_1$grange$layer)[l])
              temp_geneset_overlaps <- rbind(temp_geneset_overlaps,
                data.frame( layer =  unique(result_set_1$grange$layer)[l],
                   overlap_between = paste(sort(c(result_set_1$data_set,
                                                  result_set_2$data_set)),
                                           collapse="-"),
                   gene = c("",intersect(
                     unique(gene_ranges$gene_names[findOverlaps(result_set_1$grange[which_1], gene_ranges)@to]),
                     unique(gene_ranges$gene_names[findOverlaps(result_set_2$grange[which_2], gene_ranges)@to])
                   ))
                )
              )
              syndrome_overlap_list[[result_set_1$method_id]]$geneset_overlaps <- temp_geneset_overlaps
              
              
              
              temp_gr <- intersect(result_set_1$grange[which_1], result_set_2$grange[which_2])
              if(length(temp_gr)==0){
                syndrome_overlap_list[[result_set_1$method_id]]$range_overlaps <-
                  rbind(syndrome_overlap_list[[result_set_1$method_id]]$range_overlaps,
                        colnames(range_overlaps))
                next
              }
              B1inB2 <- is.element(rownames(data_set_list[[result_set_1$data_set]]$B),
                                   rownames(data_set_list[[result_set_2$data_set]]$B)
                                   )
              B2inB1 <- is.element(rownames(data_set_list[[result_set_2$data_set]]$B),
                                   rownames(data_set_list[[result_set_1$data_set]]$B)
                                   )
              temp_frac_dbsd <- frac_dbeta_same_dir(gr12 = temp_gr,
                                  B1 = data_set_list[[result_set_1$data_set]]$B[which(B1inB2),],
                                  atom1 = data_set_list[[result_set_1$data_set]]$locs[which(B1inB2),],
                                  case1 = which(is.element( colnames(data_set_list[[result_set_1$data_set]]$B),
                                                         data_set_list[[result_set_1$data_set]]$g2)),
                                  B2= data_set_list[[result_set_2$data_set]]$B[which(B2inB1),],
                                  atom2 = data_set_list[[result_set_2$data_set]]$locs[which(B2inB1),],
                                  case2 = which(is.element( colnames(data_set_list[[result_set_2$data_set]]$B),
                                                         data_set_list[[result_set_2$data_set]]$g2)),
                                  sigcut=0.05
                                  )
              temp_fo <- findOverlaps(temp_gr, gene_ranges)
              froms <- unique(temp_fo@from)
              temp_genes <- gene_ranges$gene_names[temp_fo@to]
              temp_gr <- as.data.frame(temp_gr)
              temp_gr$overlapping_genes <- ""
              temp_gr$overlap <- ""
              for(i in 1:length(froms)){
                temp_gr$overlapping_genes[froms[i]] <- paste(temp_genes[which(temp_fo@from==froms[i])],
                                                             collapse = ",")
              }
              temp_gr <- GRanges(temp_gr)
              temp_gr$gene_overlap_count <- countOverlaps(temp_gr, gene_ranges)
              
              temp_range_overlap <- syndrome_overlap_list[[result_set_1$method_id]]$range_overlaps
              temp_range_overlap <- rbind(temp_range_overlap,
                data.frame(layer=unique(result_set_1$grange$layer)[l],
                 datasets=paste(sort(c(result_set_1$data_set,
                                       result_set_2$data_set)),collapse="-"),
                 seqnames=as.character(seqnames(temp_gr)),start=start(temp_gr), width=width(temp_gr),
                 overlapping_genes=temp_gr$overlapping_genes, gene_overlap_count=temp_gr$gene_overlap_count,
                 num_cg_sig= temp_frac_dbsd$num_sig,
                 frac_shared_sig= temp_frac_dbsd$frac_shared_sig,
                 frac_sig_same_dir= temp_frac_dbsd$frac_sig_same_dir,
                 frac_shared_kat6a = frac_A_in_B(temp_gr,temp_gr_k),
                 frac_shared_sotos = frac_A_in_B(temp_gr,temp_gr_s),
                 frac_shared_weaver = frac_A_in_B(temp_gr,temp_gr_w)))
              
              syndrome_overlap_list[[result_set_1$method_id]]$range_overlaps <- temp_range_overlap
              
              
              cg_1 <- unique(findOverlaps(result_set_1$locs_grs[B1inB2], result_set_1$grange[which_1])@from)
              cg_2 <- unique(findOverlaps(result_set_2$locs_grs[B2inB1], result_set_2$grange[which_2])@from)
              cg_12 <- intersect(cg_1,cg_2)
              
              temp_or_table <- syndrome_overlap_list[[result_set_1$method_id]]$or_table
              temp_or_table <- rbind(temp_or_table,
                                     data.frame(cbind(layer=unique(result_set_1$grange$layer)[l],
                                                      datasets=paste(c(result_set_1$data_set,
                                                                            result_set_2$data_set),collapse="-"),
                                                      num_cg_ds1 = length(cg_1),num_cg_ds2 = length(cg_2),num_cg_ds12=length(cg_12),
                                                      syn_or(cg_1,cg_2,length(result_set_1$locs_grs[B1inB2]))
                                     )
                                     )
              )
              
              syndrome_overlap_list[[result_set_1$method_id]]$or_table <- temp_or_table
              
            }
          }
        } else {
          print(paste("method:",result_set_1$method_id))
          temp_geneset_overlaps <- syndrome_overlap_list[[result_set_1$method_id]]$geneset_overlaps
          temp_geneset_overlaps <- rbind(temp_geneset_overlaps,
            data.frame( layer =  result_set_1$method_id,
               overlap_between = paste(sort(c(result_set_1$data_set,
                                              result_set_2$data_set)),
                                       collapse="-"),
               gene = c("",intersect(
                 unique(gene_ranges$gene_names[findOverlaps(result_set_1$grange, gene_ranges)@to]),
                 unique(gene_ranges$gene_names[findOverlaps(result_set_2$grange, gene_ranges)@to])
               ))
            )
          )
          syndrome_overlap_list[[result_set_1$method_id]]$geneset_overlaps <- temp_geneset_overlaps
          
          
          
          temp_gr <- intersect(result_set_1$grange, result_set_2$grange)
          if(length(temp_gr)==0){
            syndrome_overlap_list[[result_set_1$method_id]]$range_overlaps <-
              rbind(syndrome_overlap_list[[result_set_1$method_id]]$range_overlaps,
                    colnames(range_overlaps))
            next
          }
          B1inB2 <- is.element(rownames(data_set_list[[result_set_1$data_set]]$B),
                                   rownames(data_set_list[[result_set_2$data_set]]$B)
                                   )
          B2inB1 <- is.element(rownames(data_set_list[[result_set_2$data_set]]$B),
                                   rownames(data_set_list[[result_set_1$data_set]]$B)
                                   )
          temp_frac_dbsd <- frac_dbeta_same_dir(gr12 = temp_gr,
                              B1 = data_set_list[[result_set_1$data_set]]$B[which(B1inB2),],
                              atom1 = data_set_list[[result_set_1$data_set]]$locs[which(B1inB2),],
                              case1 = which(is.element( colnames(data_set_list[[result_set_1$data_set]]$B),
                                                     data_set_list[[result_set_1$data_set]]$g2)),
                              B2= data_set_list[[result_set_2$data_set]]$B[which(B2inB1),],
                              atom2 = data_set_list[[result_set_2$data_set]]$locs[which(B2inB1),],
                              case2 = which(is.element( colnames(data_set_list[[result_set_2$data_set]]$B),
                                                     data_set_list[[result_set_2$data_set]]$g2)),
                              sigcut=0.05
                              )
          temp_fo <- findOverlaps(temp_gr, gene_ranges)
          froms <- unique(temp_fo@from)
          temp_genes <- gene_ranges$gene_names[temp_fo@to]
          temp_gr <- as.data.frame(temp_gr)
          temp_gr$overlapping_genes <- ""
          temp_gr$overlap <- ""
          for(i in 1:length(froms)){
            temp_gr$overlapping_genes[froms[i]] <- paste(temp_genes[which(temp_fo@from==froms[i])],
                                                         collapse = ",")
          }
          temp_gr <- GRanges(temp_gr)
          temp_gr$gene_overlap_count <- countOverlaps(temp_gr, gene_ranges)
          
          temp_range_overlap <- syndrome_overlap_list[[result_set_1$method_id]]$range_overlaps
              temp_range_overlap <- rbind(temp_range_overlap,
                data.frame(layer=result_set_1$method_id,
                     datasets=paste(sort(c(result_set_1$data_set,
                                           result_set_2$data_set)),collapse="-"),
                     seqnames=as.character(seqnames(temp_gr)),start=start(temp_gr), width=width(temp_gr),
                     overlapping_genes=temp_gr$overlapping_genes, gene_overlap_count=temp_gr$gene_overlap_count,
                     num_cg_sig= temp_frac_dbsd$num_sig,
                     frac_shared_sig= temp_frac_dbsd$frac_shared_sig,
                     frac_sig_same_dir= temp_frac_dbsd$frac_sig_same_dir,
                     frac_shared_kat6a = frac_A_in_B(temp_gr,gr_k),
                     frac_shared_sotos = frac_A_in_B(temp_gr,gr_s),
                     frac_shared_weaver = frac_A_in_B(temp_gr,gr_w)))
                  
          syndrome_overlap_list[[result_set_1$method_id]]$range_overlaps <- temp_range_overlap
          
          
          
          
          cg_1 <- unique(findOverlaps(result_set_1$locs_grs[B1inB2], result_set_1$grange)@from)
          cg_2 <- unique(findOverlaps(result_set_2$locs_grs[B2inB1], result_set_2$grange)@from)
          cg_12 <- intersect(cg_1,cg_2)
          
          temp_or_table <- syndrome_overlap_list[[result_set_1$method_id]]$or_table
          temp_or_table <- rbind(temp_or_table,
                                 data.frame(cbind(layer=result_set_1$method_id,
                                                  datasets=paste(c(result_set_1$data_set,
                                                                            result_set_2$data_set),collapse="-"),
                                                  num_cg_ds1 = length(cg_1),num_cg_ds2 = length(cg_2),num_cg_ds12=length(cg_12),
                                                  syn_or(cg_1,cg_2,length(result_set_1$locs_grs[B1inB2]))
                                 )
                                 )
          )
          
          syndrome_overlap_list[[result_set_1$method_id]]$or_table <- temp_or_table
          
          
          
        }
      }
    }
  }
  
  save(syndrome_overlap_list, file="syndrome_overlap_list.Rdata")

} 




data_set_ids <- grep("Sex", names(data_set_list), invert = T)
result_set_ids <- grep(paste("data_set_[", paste(data_set_ids,collapse = ""),"]", sep = ""), names(data_set_results))
method_sets <- gsub("data_set.*method_set_", "", names(data_set_results)[result_set_ids])
method_sets <- as.numeric(unique(gsub("_result.csv","",method_sets)))
for(method_set_id in method_sets){
  
  which_dsr <- grep(paste("data_set_", grep("kat6a", names(data_set_list), ignore.case = T), 
                     "__method_set_", method_set_id, sep = ""), names(data_set_results))
  if(length(which_dsr)==0){gr_k <- GRanges()} else {
    gr_k <- data_set_results[[which_dsr]]$grange
    locs_k <- data_set_results[[which_dsr]]$locs_grs
  }
  which_dsr <- grep(paste("data_set_", grep("sotos", names(data_set_list), ignore.case = T), 
                     "__method_set_", method_set_id, sep = ""), names(data_set_results))
  if(length(which_dsr)==0){gr_s <- GRanges()} else {
    gr_s <- data_set_results[[which_dsr]]$grange
    locs_s <- data_set_results[[which_dsr]]$locs_grs
  }
  which_dsr <- grep(paste("data_set_", grep("weaver", names(data_set_list), ignore.case = T), 
                     "__method_set_", method_set_id, sep = ""), names(data_set_results))
  if(length(which_dsr)==0){gr_w <- GRanges()} else {
    gr_w <- data_set_results[[which_dsr]]$grange
    locs_w <- data_set_results[[which_dsr]]$locs_grs
  }
  if(grepl("dmrscaler", names(method_set_list)[method_set_id])){
    if(grepl("_2", names(method_set_list)[method_set_id])){
      for(l in 1:length(unique(data_set_results[[2]]$grange$layer))){
        temp_gr_k <- gr_k[which(gr_k$layer==unique(data_set_results[[2]]$grange$layer)[l])] 
        temp_gr_s <- gr_s[which(gr_s$layer==unique(data_set_results[[2]]$grange$layer)[l])] 
        temp_gr_w <- gr_w[which(gr_w$layer==unique(data_set_results[[2]]$grange$layer)[l])]
        
            
        print(paste("layer:",l))
        temp_geneset_overlaps <- syndrome_overlap_list[[method_set_id]]$geneset_overlaps
        temp_geneset_overlaps <- rbind(temp_geneset_overlaps,
          data.frame( layer =  unique(temp_gr_k$layer),
             overlap_between = paste(sort(names(data_set_list)[data_set_ids]) ,collapse="-"),
             gene = c("",intersect(intersect(
               unique(gene_ranges$gene_names[findOverlaps(temp_gr_k, gene_ranges)@to]),
               unique(gene_ranges$gene_names[findOverlaps(temp_gr_s, gene_ranges)@to])
             ), unique(gene_ranges$gene_names[findOverlaps(temp_gr_w, gene_ranges)@to]))
          )
          )
        )
        syndrome_overlap_list[[method_set_id]]$geneset_overlaps <- temp_geneset_overlaps
            
        
        
        
        BKinBWS <- is.element(rownames(data_set_list$KAT6A$B),
                             rownames(data_set_list$Weaver$B)
                             )
        BWSinK <- is.element(rownames(data_set_list$Weaver$B),
                             rownames(data_set_list$KAT6A$B)
                             )
        
        
        cg_k <- unique(findOverlaps(locs_k[BKinBWS], temp_gr_k)@from)
        cg_w <- unique(findOverlaps(locs_w[BWSinK], temp_gr_w)@from)
        cg_s <- unique(findOverlaps(locs_s[BWSinK], temp_gr_s)@from)
        cg_kws <- intersect(intersect(cg_k,cg_w),cg_s)
        
        temp_or_table <- syndrome_overlap_list[[method_set_id]]$or_table
        temp_or_table <- rbind(temp_or_table,
                               data.frame(cbind(layer =  unique(temp_gr_k$layer),
                                                datasets = paste(sort(names(data_set_list)[data_set_ids]) ,collapse="-"),
                                                num_cg_ds1 = NA,num_cg_ds2 = NA,num_cg_ds12=length(cg_kws),
                                                or_estimate = NA, or_ci = NA, fisher_p = NA, alpha = NA
                               )
                               )
            )
            
            syndrome_overlap_list[[method_set_id]]$or_table <- temp_or_table
            
          }
        }
      } else {
        print(paste("method:", names(syndrome_overlap_list)[method_set_id]))
        temp_gr_k <- gr_k
        temp_gr_s <- gr_s
        temp_gr_w <- gr_w
        
        temp_geneset_overlaps <- syndrome_overlap_list[[method_set_id]]$geneset_overlaps
        temp_geneset_overlaps <- rbind(temp_geneset_overlaps,
          data.frame( layer =  names(syndrome_overlap_list)[method_set_id],
             overlap_between = paste(sort(names(data_set_list)[data_set_ids]) ,collapse="-"),
             gene = c("",intersect(intersect(
               unique(gene_ranges$gene_names[findOverlaps(temp_gr_k, gene_ranges)@to]),
               unique(gene_ranges$gene_names[findOverlaps(temp_gr_s, gene_ranges)@to])
             ), unique(gene_ranges$gene_names[findOverlaps(temp_gr_w, gene_ranges)@to]))
          )
          )
        )
        syndrome_overlap_list[[method_set_id]]$geneset_overlaps <- temp_geneset_overlaps
            
        
        
        
        BKinBWS <- is.element(rownames(data_set_list$KAT6A$B),
                             rownames(data_set_list$Weaver$B)
                             )
        BWSinK <- is.element(rownames(data_set_list$Weaver$B),
                             rownames(data_set_list$KAT6A$B)
                             )
        
        
        cg_k <- unique(findOverlaps(locs_k[BKinBWS], temp_gr_k)@from)
        cg_w <- unique(findOverlaps(locs_w[BWSinK], temp_gr_w)@from)
        cg_s <- unique(findOverlaps(locs_s[BWSinK], temp_gr_s)@from)
        cg_kws <- intersect(intersect(cg_k,cg_w),cg_s)
        
        temp_or_table <- syndrome_overlap_list[[method_set_id]]$or_table
        temp_or_table <- rbind(temp_or_table,
                               data.frame(cbind(layer =   names(syndrome_overlap_list)[method_set_id],
                                                datasets = paste(sort(names(data_set_list)[data_set_ids]) ,collapse="-"),
                                                num_cg_ds1 = NA,num_cg_ds2 = NA,num_cg_ds12=length(cg_kws),
                                                or_estimate = NA, or_ci = NA, fisher_p = NA, alpha = NA
                               )
                               )
            )
            
            syndrome_overlap_list[[method_set_id]]$or_table <- temp_or_table
            
      }
}


syndrome_overlap_list


for(i in 1:length(syndrome_overlap_list)){
  for(j in 1:length(syndrome_overlap_list[[i]])){
    write.csv(syndrome_overlap_list[[i]][[j]], file = paste(tables_dir,
                                                            names(syndrome_overlap_list)[i], "_",
                                                            names(syndrome_overlap_list[[i]])[j], ".csv", sep=""  ))
  }
}


```



```{r XY sex analysis escape from xci}
genes <- gtf_df[which(gtf_df$gene_biotype =="protein_coding" ),]
genes <- genes[which(genes$gene_source=="ensembl_havana"),]
genes <- genes[which(genes$type=="gene" ),]
genes <- genes[,!(colSums(!is.na(genes))==0)]
gene_ranges <- GRanges(seqnames = genes$seqnames, IRanges(start = genes$start, width = genes$width))
gene_ranges$gene_names <- genes$gene_name

xci_escape_or_table  <- data.frame(layer=numeric(), datasets=character(),
                                   or_estimate=numeric(), or_ci=character(),
                                   fisher_p=numeric(), alpha=numeric())


balaton_xci_escape <- fread("balaton_xci_escape.csv")
balaton_xci_escape <- balaton_xci_escape[which(is.element(balaton_xci_escape$gene_name, genes$gene_name)),]
balaton_xci_escape_genes <- genes[which(is.element(genes$gene_name, balaton_xci_escape$gene_name ) & genes$seqnames=="chrX"),]      
balaton_xci_escape_genes <- merge(balaton_xci_escape_genes, balaton_xci_escape)
balaton_xci_escape_genes_gr <- GRanges(balaton_xci_escape_genes)+1000
strand(balaton_xci_escape_genes_gr) <- "*"

ds_index <- grep("Sex", names(data_set_list))
result_set_ids <- grep(paste("data_set_", ds_index, sep = ""), names(data_set_results))

for(rs_index in result_set_ids){
  result_set <- data_set_results[[rs_index]] 
  method_set_id <- gsub("data_set.*method_set_", "", names(data_set_results)[rs_index])
  method_set_id <- gsub("_result.csv","",method_set_id)
  
  if(grepl("dmrscaler", result_set$method)){
      if(grepl("_2", result_set$method_id)){
        for(l in 1:length(unique(result_set$grange$layer))){
          
          print(paste("layer:",l))
          which_in_layer <- which(result_set$grange$layer == unique(result_set$grange$layer)[l])
          
          called_gr <- result_set$grange[which_in_layer]
          called_gr_gaps <- gaps(called_gr)
          
          locs_gr <- result_set$locs_grs
           
          
          which_xi <- which(is.element(balaton_xci_escape_genes_gr$balaton_consensus_calls, c("S", "Mostly S")) )
          cgs_xi_gr <- intersect(locs_gr,balaton_xci_escape_genes_gr[which_xi])
          which_xi_escape <- which(is.element(balaton_xci_escape_genes_gr$balaton_consensus_calls, c("E", "Mostly E")) )
          cgs_xi_escape_gr <- intersect(locs_gr,balaton_xci_escape_genes_gr[which_xi_escape])
          
          
          ft <- fisher.test(matrix(c(length(intersect(cgs_xi_gr, called_gr)),
                               length(intersect(cgs_xi_escape_gr, called_gr)),
                               length(intersect(cgs_xi_gr, called_gr_gaps)),
                               length(intersect(cgs_xi_escape_gr, called_gr_gaps))), ncol=2) )
          ft$p.value
          xci_escape_or_table <- rbind(xci_escape_or_table,
                                       data.frame(
                                         layer=unique(result_set$grange$layer)[l],
                                         datasets="Sex",
                                         or_estimate = ft$estimate,
                                         or_ci = paste(round(ft$conf.int[1],2), "-",round(ft$conf.int[2],2),sep = ""),
                                         fisher_p=ft$p.value,
                                         alpha= 1-attr(ft$conf.int, "conf.level") 
                                       )
          )
        }
      }
  } else {
    
    called_gr <- result_set$grange
    called_gr_gaps <- gaps(called_gr)
    
    locs_gr <- result_set$locs_grs
     
    which_xi <- which(is.element(balaton_xci_escape_genes_gr$balaton_consensus_calls, c("S", "Mostly S")) )
    cgs_xi_gr <- intersect(locs_gr,balaton_xci_escape_genes_gr[which_xi])
    which_xi_escape <- which(is.element(balaton_xci_escape_genes_gr$balaton_consensus_calls, c("E", "Mostly E")) )
    cgs_xi_escape_gr <- intersect(locs_gr,balaton_xci_escape_genes_gr[which_xi_escape])
    
    
    ft <- fisher.test(matrix(c(length(intersect(cgs_xi_gr, called_gr)),
                         length(intersect(cgs_xi_escape_gr, called_gr)),
                         length(intersect(cgs_xi_gr, called_gr_gaps)),
                         length(intersect(cgs_xi_escape_gr, called_gr_gaps))), ncol=2) )
    ft$p.value
    xci_escape_or_table <- rbind(xci_escape_or_table,
                                 data.frame(
                                   layer=result_set$method_id,
                                   datasets="Sex",
                                   or_estimate = ft$estimate,
                                   or_ci = paste(round(ft$conf.int[1],2), "-",round(ft$conf.int[2],2),sep = ""),
                                   fisher_p=ft$p.value,
                                   alpha= 1-attr(ft$conf.int, "conf.level") 
                                 )
    )

  }
  
}


xci_escape_or_table$OR <- paste(signif(xci_escape_or_table$or_estimate,3)," (", xci_escape_or_table$or_ci, ")" , sep="")
filename <- paste(tables_dir, "xci_escape_or_table.csv")
write.csv(xci_escape_or_table, file = filename)



```







